{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis Template\n",
    "WIP Analysis template, adapted from Margaux's notebook from 07/12/2023."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "### PARAMETERS:\n",
    "EGFP_SLOPE = 91900.03\n",
    "EGFP_SLOPE_CONC_UNITS = 'nM' #RFU/nM\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "#enables autoreloding of modules\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import os\n",
    "from copy import deepcopy\n",
    "\n",
    "import skimage\n",
    "import math\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import scipy.stats\n",
    "from pprint import pprint\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.linear_model import LinearRegression\n",
    "from scipy.optimize import curve_fit\n",
    "\n",
    "from pathlib import Path \n",
    "from htbam_analysis.htbam_db_api import LocalHtbamDBAPI\n",
    "\n",
    "#Import Kinetics Package for line fitting:\n",
    "import kinetics\n",
    "\n",
    "#Configuration settings for pandas and seaborn\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_rows', None)\n",
    "pd.set_option('display.float_format', '{:.2f}'.format)\n",
    "sns.set(style='ticks', context='paper', font_scale=1.2, rc={\"lines.linewidth\": 1.2})\n",
    "\n",
    "#enable inline plotting of matplotlib figures\n",
    "%matplotlib inline\n",
    "\n",
    "#set the figure format to SVG\n",
    "%config InlineBackend.figure_format = 'svg'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## 1. Connect DB Api"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: parsing concentration from string\n",
      "Warning: parsing concentration from string\n",
      "Warning: parsing concentration from string\n",
      "Warning: parsing concentration from string\n",
      "Warning: parsing concentration from string\n",
      "Warning: parsing concentration from string\n",
      "Warning: parsing concentration from string\n",
      "Warning: parsing concentration from string\n",
      "Warning: parsing concentration from string\n",
      "Warning: parsing concentration from string\n",
      "Warning: parsing concentration from string\n"
     ]
    }
   ],
   "source": [
    "from htbam_analysis.htbam_db_api import LocalHtbamDBAPI\n",
    "\n",
    "root = '../test/test_data/'\n",
    "db_conn = LocalHtbamDBAPI(standard_curve_data_path= root + 'd2_5_StandardSeries_Analysis.csv.bz2', standard_name=\"NADPH std curve\", standard_type=\"NADPH\", standard_units=\"uM\",\n",
    "                         kinetic_data_path= root+ 'd2_TitrationSeries_Analysis.csv.bz2', kinetic_name=\"ADP kinetics curve\", kinetic_type=\"ADP\", kinetic_units=\"uM\")\n",
    "\n",
    "### TODO: Still want the following data in the database:\n",
    "# date (and time?) collected\n",
    "# operator name\n",
    "# Add additional descriptors\n",
    "# substrate_name\n",
    "# setup(?) and device_num\n",
    "# width/height\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Format of DB:\n",
    "DB\n",
    "- **chamber_metadata**\n",
    "    - '1,1'\n",
    "        - 'id'                  'organism_ADK'\n",
    "        - 'radius_chamber'      35.0\n",
    "        - 'x_center_chamber'    31.0\n",
    "        - 'y_center_chamber'    43.0\n",
    "        - 'xslice'              '(6758, 6805)'\n",
    "        - 'yslice'              '(6704, 6804)'\n",
    "    - ...\n",
    "    - '32,56' ...\n",
    "- **runs**\n",
    "    - 'standard_0'\n",
    "        - 'name'            'NAPDH_std_curve'\n",
    "        - 'type'            'NAPDH'\n",
    "        - 'conc_unit'       'uM'\n",
    "        - 'assays':\n",
    "            - 0:\n",
    "                - 'conc'\n",
    "                - 'time_s'\n",
    "                - 'chambers'\n",
    "                    - '1,1'\n",
    "                    - ...\n",
    "                    - '32,56' ...\n",
    "            - ...\n",
    "            - 6: ...\n",
    "        - 'analyses':\n",
    "            - 'linear_regression':\n",
    "                - 'chambers':\n",
    "                    - '1,1':\n",
    "                        - slope:\n",
    "                        - intercept:\n",
    "                        - r_value:\n",
    "                        - p_value: \n",
    "                        - std_err:\n",
    "                    - ...\n",
    "                    - '32,56'...\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "chamber_metadata: \n",
      "\t{\n",
      "\t1,1: \n",
      "\t\t{\n",
      "\t\tid: M_caps_ADK\n",
      "\t\tx_center_chamber: 35.0\n",
      "\t\ty_center_chamber: 51.0\n",
      "\t\tradius_chamber: 35.0\n",
      "\t\txslice: (395, 495)\n",
      "\t\t...\n",
      "\t\t}\n",
      "\t1,2: \n",
      "\t\t{\n",
      "\t\tid: P_syri_ADK\n",
      "\t\tx_center_chamber: 35.0\n",
      "\t\ty_center_chamber: 51.0\n",
      "\t\tradius_chamber: 34.0\n",
      "\t\txslice: (510, 610)\n",
      "\t\t...\n",
      "\t\t}\n",
      "\t1,3: \n",
      "\t\t{\n",
      "\t\tid: L_reut_ADK\n",
      "\t\tx_center_chamber: 37.0\n",
      "\t\ty_center_chamber: 51.0\n",
      "\t\tradius_chamber: 37.0\n",
      "\t\txslice: (626, 726)\n",
      "\t\t...\n",
      "\t\t}\n",
      "\t1,4: \n",
      "\t\t{\n",
      "\t\tid: G_ther_ADK\n",
      "\t\tx_center_chamber: 37.0\n",
      "\t\ty_center_chamber: 49.0\n",
      "\t\tradius_chamber: 36.0\n",
      "\t\txslice: (742, 842)\n",
      "\t\t...\n",
      "\t\t}\n",
      "\t1,5: \n",
      "\t\t{\n",
      "\t\tid: bmADK\n",
      "\t\tx_center_chamber: 35.0\n",
      "\t\ty_center_chamber: 49.0\n",
      "\t\tradius_chamber: 34.0\n",
      "\t\txslice: (858, 958)\n",
      "\t\t...\n",
      "\t\t}\n",
      "\t...\n",
      "\t}\n",
      "runs: \n",
      "\t{\n",
      "\tstandard_0: \n",
      "\t\t{\n",
      "\t\tname: NADPH std curve\n",
      "\t\ttype: NADPH\n",
      "\t\tconc_unit: uM\n",
      "\t\tassays: \n",
      "\t\t\t{\n",
      "\t\t\t0: \n",
      "\t\t\t\t{\n",
      "\t\t\t\tconc: 3.90625\n",
      "\t\t\t\ttime_s: [0]\n",
      "\t\t\t\tchambers: \n",
      "\t\t\t\t\t{\n",
      "\t\t\t\t\t1,1: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [188096.0]\n",
      "\t\t\t\t\t\tstd_chamber: [78.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,10: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [228483.0]\n",
      "\t\t\t\t\t\tstd_chamber: [99.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,11: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [224214.0]\n",
      "\t\t\t\t\t\tstd_chamber: [84.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,12: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [196095.0]\n",
      "\t\t\t\t\t\tstd_chamber: [80.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,13: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [234481.0]\n",
      "\t\t\t\t\t\tstd_chamber: [253.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t...\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\t\t\t1: \n",
      "\t\t\t\t{\n",
      "\t\t\t\tconc: 7.8125\n",
      "\t\t\t\ttime_s: [0]\n",
      "\t\t\t\tchambers: \n",
      "\t\t\t\t\t{\n",
      "\t\t\t\t\t1,1: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [240692.0]\n",
      "\t\t\t\t\t\tstd_chamber: [88.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,10: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [259851.0]\n",
      "\t\t\t\t\t\tstd_chamber: [105.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,11: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [267881.0]\n",
      "\t\t\t\t\t\tstd_chamber: [94.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,12: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [229483.0]\n",
      "\t\t\t\t\t\tstd_chamber: [85.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,13: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [266559.0]\n",
      "\t\t\t\t\t\tstd_chamber: [257.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t...\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\t\t\t2: \n",
      "\t\t\t\t{\n",
      "\t\t\t\tconc: 15.625\n",
      "\t\t\t\ttime_s: [0]\n",
      "\t\t\t\tchambers: \n",
      "\t\t\t\t\t{\n",
      "\t\t\t\t\t1,1: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [330890.0]\n",
      "\t\t\t\t\t\tstd_chamber: [104.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,10: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [358876.0]\n",
      "\t\t\t\t\t\tstd_chamber: [119.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,11: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [367405.0]\n",
      "\t\t\t\t\t\tstd_chamber: [106.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,12: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [349277.0]\n",
      "\t\t\t\t\t\tstd_chamber: [103.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,13: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [368497.0]\n",
      "\t\t\t\t\t\tstd_chamber: [264.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t...\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\t\t\t3: \n",
      "\t\t\t\t{\n",
      "\t\t\t\tconc: 31.25\n",
      "\t\t\t\ttime_s: [0]\n",
      "\t\t\t\tchambers: \n",
      "\t\t\t\t\t{\n",
      "\t\t\t\t\t1,1: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [644967.0]\n",
      "\t\t\t\t\t\tstd_chamber: [141.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,10: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [643179.0]\n",
      "\t\t\t\t\t\tstd_chamber: [147.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,11: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [681549.0]\n",
      "\t\t\t\t\t\tstd_chamber: [138.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,12: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [656808.0]\n",
      "\t\t\t\t\t\tstd_chamber: [134.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,13: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [655309.0]\n",
      "\t\t\t\t\t\tstd_chamber: [272.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t...\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\t\t\t4: \n",
      "\t\t\t\t{\n",
      "\t\t\t\tconc: 125.0\n",
      "\t\t\t\ttime_s: [0]\n",
      "\t\t\t\tchambers: \n",
      "\t\t\t\t\t{\n",
      "\t\t\t\t\t1,1: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [2990428.0]\n",
      "\t\t\t\t\t\tstd_chamber: [327.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,10: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [2682125.0]\n",
      "\t\t\t\t\t\tstd_chamber: [289.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,11: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [2997113.0]\n",
      "\t\t\t\t\t\tstd_chamber: [323.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,12: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [3050228.0]\n",
      "\t\t\t\t\t\tstd_chamber: [293.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,13: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [2964056.0]\n",
      "\t\t\t\t\t\tstd_chamber: [399.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t...\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\t\t\t...\n",
      "\t\t\t}\n",
      "\t\t}\n",
      "\tkinetics_0: \n",
      "\t\t{\n",
      "\t\tname: ADP kinetics curve\n",
      "\t\ttype: ADP\n",
      "\t\tconc_unit: uM\n",
      "\t\tassays: \n",
      "\t\t\t{\n",
      "\t\t\t0: \n",
      "\t\t\t\t{\n",
      "\t\t\t\tconc: 1000.0\n",
      "\t\t\t\ttime_s: [0, 107, 1257, 143, 1557, 179, 1857, 214, 250, 286, 322, 357, 36, 417, 477, 537, 597, 657, 71, 957]\n",
      "\t\t\t\tchambers: \n",
      "\t\t\t\t\t{\n",
      "\t\t\t\t\t1,1: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [5706550.0, 6111047.0, 11687580.0, 6252476.0, 12771839.0, 6496490.0, 13601648.0, 6560664.0, 6807164.0, 6904110.0, 7100560.0, 7285440.0, 5779893.0, 7742776.0, 8097891.0, 8364556.0, 8657411.0, 8870927.0, 5918860.0, 10368417.0]\n",
      "\t\t\t\t\t\tstd_chamber: [387.0, 478.0, 1495.0, 513.0, 1694.0, 548.0, 1861.0, 583.0, 627.0, 658.0, 696.0, 724.0, 409.0, 796.0, 853.0, 909.0, 969.0, 1015.0, 441.0, 1271.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,10: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [993859.0, 6815357.0, 21599775.0, 8271935.0, 22307698.0, 9475365.0, 22550946.0, 10476613.0, 11408539.0, 12161098.0, 12919157.0, 13566534.0, 2989408.0, 14696183.0, 15726882.0, 16494785.0, 17213104.0, 17816508.0, 5088504.0, 20274702.0]\n",
      "\t\t\t\t\t\tstd_chamber: [219.0, 1210.0, 3669.0, 1461.0, 3772.0, 1669.0, 3801.0, 1843.0, 2000.0, 2136.0, 2265.0, 2370.0, 546.0, 2558.0, 2727.0, 2859.0, 2975.0, 3078.0, 908.0, 3469.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,11: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [1019894.0, 1283020.0, 5306931.0, 1424235.0, 6137206.0, 1558000.0, 6822833.0, 1673322.0, 1811411.0, 1913603.0, 2042143.0, 2168285.0, 1088892.0, 2451438.0, 2723241.0, 2939916.0, 3129633.0, 3308864.0, 1170345.0, 4381145.0]\n",
      "\t\t\t\t\t\tstd_chamber: [211.0, 261.0, 845.0, 283.0, 961.0, 301.0, 1071.0, 322.0, 342.0, 361.0, 378.0, 399.0, 222.0, 438.0, 473.0, 502.0, 534.0, 563.0, 241.0, 714.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,12: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [1018086.0, 5533008.0, 22021641.0, 6805264.0, 23610365.0, 7888695.0, 24491342.0, 8849341.0, 9713049.0, 10436485.0, 11202118.0, 11878670.0, 2472329.0, 13062224.0, 14126802.0, 14966136.0, 15785480.0, 16511326.0, 4099259.0, 19723086.0]\n",
      "\t\t\t\t\t\tstd_chamber: [217.0, 1003.0, 3773.0, 1225.0, 4027.0, 1413.0, 4162.0, 1577.0, 1730.0, 1860.0, 1986.0, 2105.0, 465.0, 2302.0, 2478.0, 2621.0, 2759.0, 2880.0, 748.0, 3411.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,13: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [893434.0, 1562758.0, 8276068.0, 1854220.0, 9517212.0, 2115900.0, 10547386.0, 2365359.0, 2602063.0, 2817998.0, 3057931.0, 3296254.0, 1076345.0, 3749090.0, 4181894.0, 4546259.0, 4891835.0, 5217944.0, 1289226.0, 6855831.0]\n",
      "\t\t\t\t\t\tstd_chamber: [321.0, 409.0, 1345.0, 441.0, 1525.0, 480.0, 1676.0, 514.0, 545.0, 581.0, 613.0, 647.0, 339.0, 705.0, 766.0, 816.0, 866.0, 915.0, 367.0, 1147.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t...\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\t\t\t1: \n",
      "\t\t\t\t{\n",
      "\t\t\t\tconc: 125.0\n",
      "\t\t\t\ttime_s: [0, 107, 1258, 143, 1558, 179, 1858, 214, 250, 286, 322, 358, 36, 418, 478, 538, 598, 658, 71, 958]\n",
      "\t\t\t\tchambers: \n",
      "\t\t\t\t\t{\n",
      "\t\t\t\t\t1,1: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [8136071.0, 8779072.0, 12297277.0, 8973313.0, 12711744.0, 9135669.0, 12874143.0, 9337468.0, 9460129.0, 9571357.0, 9656313.0, 9815460.0, 8274315.0, 10101405.0, 10303423.0, 10462064.0, 10687599.0, 10734019.0, 8517578.0, 11716124.0]\n",
      "\t\t\t\t\t\tstd_chamber: [487.0, 511.0, 755.0, 519.0, 795.0, 531.0, 801.0, 539.0, 552.0, 565.0, 576.0, 584.0, 488.0, 604.0, 615.0, 635.0, 649.0, 658.0, 498.0, 715.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,10: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [625410.0, 3474955.0, 6217674.0, 3918941.0, 6377513.0, 4242168.0, 6363578.0, 4512960.0, 4710216.0, 4847984.0, 4966483.0, 5048604.0, 1780525.0, 5245305.0, 5388851.0, 5491937.0, 5573753.0, 5632528.0, 2794202.0, 6016738.0]\n",
      "\t\t\t\t\t\tstd_chamber: [195.0, 418.0, 639.0, 456.0, 647.0, 484.0, 643.0, 510.0, 528.0, 541.0, 551.0, 562.0, 277.0, 578.0, 588.0, 597.0, 601.0, 607.0, 353.0, 629.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,11: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [540378.0, 738863.0, 2930814.0, 807846.0, 3332516.0, 866794.0, 3533592.0, 955709.0, 1045780.0, 1095217.0, 1154148.0, 1224701.0, 592045.0, 1363423.0, 1517568.0, 1633090.0, 1759191.0, 1836781.0, 680949.0, 2467240.0]\n",
      "\t\t\t\t\t\tstd_chamber: [132.0, 156.0, 345.0, 165.0, 376.0, 170.0, 396.0, 181.0, 191.0, 195.0, 199.0, 207.0, 140.0, 221.0, 236.0, 244.0, 254.0, 261.0, 152.0, 310.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,12: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [567603.0, 3128404.0, 6835637.0, 3622077.0, 7062621.0, 4005722.0, 7068715.0, 4325448.0, 4580633.0, 4786534.0, 4959712.0, 5110547.0, 1525622.0, 5365631.0, 5581630.0, 5739273.0, 5883106.0, 5966593.0, 2458335.0, 6557399.0]\n",
      "\t\t\t\t\t\tstd_chamber: [137.0, 369.0, 685.0, 416.0, 695.0, 455.0, 701.0, 485.0, 509.0, 529.0, 543.0, 561.0, 225.0, 585.0, 599.0, 614.0, 626.0, 631.0, 305.0, 669.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,13: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [414467.0, 950454.0, 4466705.0, 1143795.0, 4952392.0, 1311376.0, 5244840.0, 1483921.0, 1649943.0, 1766141.0, 1906750.0, 2032047.0, 562480.0, 2279292.0, 2516475.0, 2703785.0, 2876810.0, 3023237.0, 763219.0, 3857929.0]\n",
      "\t\t\t\t\t\tstd_chamber: [257.0, 296.0, 515.0, 311.0, 538.0, 330.0, 561.0, 334.0, 347.0, 354.0, 364.0, 372.0, 273.0, 380.0, 401.0, 416.0, 423.0, 428.0, 286.0, 481.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t...\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\t\t\t2: \n",
      "\t\t\t\t{\n",
      "\t\t\t\tconc: 15.625\n",
      "\t\t\t\ttime_s: [0, 107, 1256, 142, 1556, 178, 1856, 214, 250, 285, 321, 35, 356, 416, 476, 536, 596, 656, 71, 956]\n",
      "\t\t\t\tchambers: \n",
      "\t\t\t\t\t{\n",
      "\t\t\t\t\t1,1: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [468947.0, 495223.0, 933946.0, 523283.0, 912354.0, 544724.0, 917781.0, 570212.0, 579292.0, 572060.0, 600666.0, 456552.0, 610655.0, 642864.0, 636502.0, 647804.0, 661358.0, 702630.0, 458014.0, 800365.0]\n",
      "\t\t\t\t\t\tstd_chamber: [135.0, 142.0, 206.0, 148.0, 205.0, 150.0, 202.0, 157.0, 157.0, 156.0, 160.0, 134.0, 165.0, 169.0, 166.0, 172.0, 172.0, 178.0, 137.0, 192.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,10: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [490584.0, 753449.0, 996616.0, 780848.0, 956578.0, 815317.0, 967917.0, 832076.0, 824184.0, 821190.0, 830193.0, 599113.0, 802926.0, 829600.0, 815588.0, 813541.0, 821437.0, 849516.0, 679183.0, 894883.0]\n",
      "\t\t\t\t\t\tstd_chamber: [143.0, 181.0, 206.0, 187.0, 203.0, 190.0, 208.0, 192.0, 192.0, 193.0, 194.0, 159.0, 193.0, 195.0, 191.0, 191.0, 193.0, 196.0, 171.0, 201.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,11: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [535340.0, 485538.0, 818524.0, 496728.0, 801014.0, 499421.0, 842316.0, 521427.0, 508488.0, 489754.0, 515960.0, 513026.0, 488340.0, 517563.0, 509119.0, 539875.0, 532066.0, 580712.0, 490005.0, 663461.0]\n",
      "\t\t\t\t\t\tstd_chamber: [136.0, 133.0, 174.0, 133.0, 172.0, 134.0, 175.0, 138.0, 135.0, 134.0, 137.0, 134.0, 136.0, 141.0, 136.0, 141.0, 141.0, 149.0, 132.0, 158.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,12: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [470655.0, 721218.0, 1080245.0, 775635.0, 1014988.0, 801196.0, 1028059.0, 835863.0, 835737.0, 829286.0, 836905.0, 570745.0, 833564.0, 854152.0, 836863.0, 852024.0, 862002.0, 883963.0, 647772.0, 944915.0]\n",
      "\t\t\t\t\t\tstd_chamber: [136.0, 172.0, 212.0, 180.0, 205.0, 184.0, 205.0, 188.0, 188.0, 188.0, 189.0, 150.0, 189.0, 191.0, 192.0, 192.0, 192.0, 197.0, 164.0, 200.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,13: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [478166.0, 489205.0, 947494.0, 510785.0, 925715.0, 545181.0, 966055.0, 563899.0, 570733.0, 570826.0, 600189.0, 470683.0, 568195.0, 620881.0, 617719.0, 642148.0, 669995.0, 691957.0, 468226.0, 815515.0]\n",
      "\t\t\t\t\t\tstd_chamber: [285.0, 282.0, 324.0, 281.0, 320.0, 294.0, 321.0, 292.0, 287.0, 292.0, 291.0, 280.0, 286.0, 295.0, 296.0, 294.0, 298.0, 296.0, 278.0, 313.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t...\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\t\t\t3: \n",
      "\t\t\t\t{\n",
      "\t\t\t\tconc: 2000.0\n",
      "\t\t\t\ttime_s: [0, 107, 1257, 143, 1557, 179, 1857, 214, 250, 286, 322, 357, 36, 417, 477, 537, 597, 657, 72, 957]\n",
      "\t\t\t\tchambers: \n",
      "\t\t\t\t\t{\n",
      "\t\t\t\t\t1,1: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [5627998.0, 5920694.0, 11383704.0, 6090270.0, 12552552.0, 6207816.0, 13624902.0, 6405748.0, 6614038.0, 6754358.0, 6936485.0, 7094913.0, 5655647.0, 7415492.0, 7796819.0, 8094819.0, 8369920.0, 8615189.0, 5786569.0, 10173032.0]\n",
      "\t\t\t\t\t\tstd_chamber: [417.0, 514.0, 1603.0, 546.0, 1819.0, 584.0, 2014.0, 625.0, 664.0, 698.0, 743.0, 775.0, 447.0, 836.0, 903.0, 966.0, 1018.0, 1079.0, 477.0, 1354.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,10: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [1363446.0, 8447498.0, 22000192.0, 10365329.0, 22147464.0, 11983564.0, 22222185.0, 13361742.0, 14569287.0, 15590652.0, 16573526.0, 17380636.0, 3565965.0, 18594608.0, 19547522.0, 20176072.0, 20646472.0, 20943412.0, 6224505.0, 21898281.0]\n",
      "\t\t\t\t\t\tstd_chamber: [275.0, 1477.0, 3752.0, 1801.0, 3759.0, 2087.0, 3759.0, 2316.0, 2531.0, 2715.0, 2872.0, 3012.0, 638.0, 3215.0, 3374.0, 3476.0, 3547.0, 3604.0, 1089.0, 3743.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,11: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [1503895.0, 1802314.0, 6234549.0, 1955100.0, 7216987.0, 2090102.0, 8109026.0, 2225588.0, 2369030.0, 2491170.0, 2652715.0, 2781230.0, 1541312.0, 3068866.0, 3345926.0, 3568084.0, 3815430.0, 4051724.0, 1678663.0, 5292879.0]\n",
      "\t\t\t\t\t\tstd_chamber: [278.0, 335.0, 1016.0, 357.0, 1166.0, 381.0, 1303.0, 406.0, 430.0, 449.0, 474.0, 494.0, 290.0, 532.0, 576.0, 612.0, 649.0, 688.0, 313.0, 870.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,12: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [1457764.0, 6958205.0, 24322089.0, 8642840.0, 24652175.0, 10088833.0, 24838996.0, 11393315.0, 12571546.0, 13664576.0, 14693028.0, 15598688.0, 3047405.0, 17116710.0, 18418610.0, 19535541.0, 20471010.0, 21219041.0, 5105053.0, 23729057.0]\n",
      "\t\t\t\t\t\tstd_chamber: [289.0, 1234.0, 4179.0, 1526.0, 4221.0, 1777.0, 4231.0, 2005.0, 2213.0, 2408.0, 2580.0, 2739.0, 559.0, 2997.0, 3218.0, 3401.0, 3559.0, 3688.0, 912.0, 4092.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,13: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [1386957.0, 2093055.0, 9623520.0, 2387479.0, 11142535.0, 2683778.0, 12518506.0, 2947950.0, 3201093.0, 3468491.0, 3754681.0, 3999817.0, 1509319.0, 4489910.0, 4909632.0, 5344920.0, 5758692.0, 6144362.0, 1817168.0, 8086768.0]\n",
      "\t\t\t\t\t\tstd_chamber: [376.0, 472.0, 1571.0, 509.0, 1792.0, 554.0, 1996.0, 592.0, 630.0, 666.0, 709.0, 746.0, 393.0, 819.0, 874.0, 934.0, 998.0, 1055.0, 434.0, 1341.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t...\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\t\t\t4: \n",
      "\t\t\t\t{\n",
      "\t\t\t\tconc: 250.0\n",
      "\t\t\t\ttime_s: [0, 107, 1258, 143, 1558, 179, 1858, 215, 250, 286, 322, 358, 36, 418, 478, 538, 598, 658, 71, 958]\n",
      "\t\t\t\tchambers: \n",
      "\t\t\t\t\t{\n",
      "\t\t\t\t\t1,1: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [7535584.0, 8382561.0, 14100299.0, 8760094.0, 14987080.0, 9017251.0, 15508256.0, 9292434.0, 9562967.0, 9802961.0, 10015168.0, 10258792.0, 7734090.0, 10656319.0, 11049251.0, 11329217.0, 11528167.0, 11800331.0, 8072487.0, 13333677.0]\n",
      "\t\t\t\t\t\tstd_chamber: [441.0, 483.0, 1002.0, 511.0, 1086.0, 530.0, 1137.0, 550.0, 576.0, 598.0, 619.0, 638.0, 449.0, 675.0, 711.0, 738.0, 763.0, 784.0, 465.0, 927.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,10: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [880995.0, 5370831.0, 11811674.0, 6296887.0, 12157938.0, 6981369.0, 12257941.0, 7521239.0, 7995276.0, 8345707.0, 8723488.0, 8995632.0, 2621311.0, 9471998.0, 9826605.0, 10097197.0, 10342822.0, 10527784.0, 4215896.0, 11501721.0]\n",
      "\t\t\t\t\t\tstd_chamber: [214.0, 578.0, 1158.0, 664.0, 1183.0, 729.0, 1182.0, 785.0, 821.0, 865.0, 893.0, 923.0, 336.0, 965.0, 998.0, 1022.0, 1042.0, 1063.0, 469.0, 1138.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,11: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [775713.0, 1071546.0, 4143507.0, 1204379.0, 4734995.0, 1319881.0, 5129872.0, 1431101.0, 1568418.0, 1651551.0, 1791274.0, 1883931.0, 853886.0, 2093719.0, 2287495.0, 2465715.0, 2632796.0, 2746208.0, 980488.0, 3622744.0]\n",
      "\t\t\t\t\t\tstd_chamber: [153.0, 184.0, 457.0, 198.0, 504.0, 210.0, 543.0, 221.0, 231.0, 242.0, 250.0, 259.0, 162.0, 278.0, 293.0, 310.0, 322.0, 335.0, 176.0, 403.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,12: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [850858.0, 4675163.0, 12579265.0, 5541853.0, 13171457.0, 6241412.0, 13397844.0, 6832114.0, 7367405.0, 7779092.0, 8202876.0, 8534325.0, 2243673.0, 9115483.0, 9607137.0, 9979384.0, 10299599.0, 10591028.0, 3594942.0, 11989015.0]\n",
      "\t\t\t\t\t\tstd_chamber: [157.0, 504.0, 1230.0, 583.0, 1275.0, 657.0, 1293.0, 711.0, 761.0, 803.0, 841.0, 878.0, 275.0, 929.0, 974.0, 1006.0, 1036.0, 1068.0, 402.0, 1179.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,13: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [618191.0, 1413115.0, 6585653.0, 1685668.0, 7410296.0, 1953105.0, 7971561.0, 2184447.0, 2461198.0, 2669471.0, 2875424.0, 3065539.0, 838047.0, 3453987.0, 3761130.0, 4034591.0, 4274868.0, 4526082.0, 1117964.0, 5789835.0]\n",
      "\t\t\t\t\t\tstd_chamber: [273.0, 329.0, 668.0, 340.0, 716.0, 364.0, 760.0, 373.0, 390.0, 403.0, 420.0, 428.0, 289.0, 455.0, 470.0, 492.0, 514.0, 528.0, 311.0, 602.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t...\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\t\t\t...\n",
      "\t\t\t}\n",
      "\t\t}\n",
      "\t}\n",
      "button_quant: \n",
      "\t{\n",
      "\t1,1: \n",
      "\t\t{\n",
      "\t\tsummed_button_Button_Quant: 15041\n",
      "\t\tsummed_button_BGsub_Button_Quant: 7088\n",
      "\t\tstd_button_Button_Quant: 21\n",
      "\t\t}\n",
      "\t1,2: \n",
      "\t\t{\n",
      "\t\tsummed_button_Button_Quant: 21514\n",
      "\t\tsummed_button_BGsub_Button_Quant: 6570\n",
      "\t\tstd_button_Button_Quant: 17\n",
      "\t\t}\n",
      "\t1,3: \n",
      "\t\t{\n",
      "\t\tsummed_button_Button_Quant: 152498\n",
      "\t\tsummed_button_BGsub_Button_Quant: 122704\n",
      "\t\tstd_button_Button_Quant: 163\n",
      "\t\t}\n",
      "\t1,4: \n",
      "\t\t{\n",
      "\t\tsummed_button_Button_Quant: 276149\n",
      "\t\tsummed_button_BGsub_Button_Quant: 239605\n",
      "\t\tstd_button_Button_Quant: 296\n",
      "\t\t}\n",
      "\t1,5: \n",
      "\t\t{\n",
      "\t\tsummed_button_Button_Quant: 1085272\n",
      "\t\tsummed_button_BGsub_Button_Quant: 1051853\n",
      "\t\tstd_button_Button_Quant: 384\n",
      "\t\t}\n",
      "\t...\n",
      "\t}\n",
      "}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(str(db_conn))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## 4. Standards"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each chamber, we've taken values at a set of concentrations. We need to perform a linear regression __for each chamber__ to relate the luminance of each chamber to its substrate concentration."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "#N.F. Sometime soon, let's change the HTBAM API so that kinetics/standards are in a unified format. \n",
    "# In both assays, we record information at several concentrations (substrate for kinetics, product for standard)\n",
    "# In standard, we only measure once, while for kinetics, we measure over time. We could have the fluorescence value be a list, of length 1 for standard.\n",
    "\n",
    "def assay_data_to_array(db_obj, run_name):\n",
    "    '''This function takes as input an HTBAM Database object.\n",
    "    For each kinetics run, we have \n",
    "    It returns 3 numpy arrays:\n",
    "    chamber_ids: an array of the chamber ids (in the format '1,1' ... '32,56')\n",
    "        shape: (n_chambers,)\n",
    "    luminance_data: an array of the luminance data for each chamber\n",
    "        shape: (n_time_points, n_chambers, n_assays)\n",
    "    conc_data: an array of the concentration data for each chamber.\n",
    "        shape: (n_assays,)\n",
    "    time_data: an array of the time data for each time point.\n",
    "        shape: (n_time_points, n_assays)\n",
    "    '''\n",
    "    \n",
    "    chamber_idxs = np.array(list(db_obj._json_dict['chamber_metadata'].keys()))\n",
    "    luminance_data = None\n",
    "    time_data = None\n",
    "    conc_data = np.array([])\n",
    "\n",
    "    #Each assay may have recorded a different # of time points.\n",
    "    #First, we'll just check what the max # of time points is:\n",
    "    max_time_points = 0\n",
    "    for assay in db_obj._json_dict[\"runs\"][run_name]['assays'].keys():\n",
    "        current_assay_time_points = len(np.array(db_obj._json_dict[\"runs\"][run_name]['assays'][assay]['time_s']))\n",
    "        if current_assay_time_points > max_time_points:\n",
    "            max_time_points = current_assay_time_points\n",
    "\n",
    "    for assay in db_obj._json_dict[\"runs\"][run_name]['assays'].keys():\n",
    "        \n",
    "        #to make things easier later, we'll be sorting the datapoints by time value.\n",
    "        #Get time data:\n",
    "        #collect from DB\n",
    "        current_time_array = np.array(db_obj._json_dict[\"runs\"][run_name]['assays'][assay]['time_s'])\n",
    "        current_time_array = current_time_array.astype(float) #so we can pad with NaNs\n",
    "        #pad the array with NaNs if there are fewer time points than the max\n",
    "        current_time_array = np.pad(current_time_array, (0, max_time_points - len(current_time_array)), 'constant', constant_values=np.nan)\n",
    "        #sort, and capture sorting idxs:\n",
    "        sorting_idxs = np.argsort(current_time_array)\n",
    "        current_time_array = current_time_array[sorting_idxs]\n",
    "        current_time_array = np.expand_dims(current_time_array, axis=1)\n",
    "        #add to our dataset\n",
    "        if time_data is None:\n",
    "            time_data = current_time_array\n",
    "        else:\n",
    "            time_data = np.concatenate([time_data, current_time_array], axis=1)\n",
    "\n",
    "        #Get luminance data:\n",
    "        current_luminance_array = None\n",
    "        for chamber_idx in chamber_idxs:\n",
    "            #collect from DB\n",
    "            current_chamber_array = np.array(db_obj._json_dict[\"runs\"][run_name]['assays'][assay]['chambers'][chamber_idx]['sum_chamber'])\n",
    "            #pad the array with NaNs if there are fewer time points than the max\n",
    "            current_chamber_array = np.pad(current_chamber_array, (0, max_time_points - len(current_chamber_array)), 'constant', constant_values=np.nan)\n",
    "            #sort by time:\n",
    "            current_chamber_array = current_chamber_array[sorting_idxs]\n",
    "            #add a dimension at the end:\n",
    "            current_chamber_array = np.expand_dims(current_chamber_array, axis=1)\n",
    "\n",
    "            if current_luminance_array is None:\n",
    "                current_luminance_array = current_chamber_array\n",
    "            else:\n",
    "                current_luminance_array = np.concatenate([current_luminance_array, current_chamber_array], axis=1)\n",
    "        #add a dimension at the end:\n",
    "        current_luminance_array = np.expand_dims(current_luminance_array, axis=2)\n",
    "        #add to our dataset\n",
    "        if luminance_data is None:\n",
    "            luminance_data = current_luminance_array\n",
    "        else:\n",
    "            luminance_data = np.concatenate([luminance_data, current_luminance_array], axis=2)\n",
    "        \n",
    "        #Get concentration data:\n",
    "        #collect from DB\n",
    "        current_conc = db_obj._json_dict[\"runs\"][run_name]['assays'][assay]['conc']\n",
    "        conc_data = np.append(conc_data, current_conc)\n",
    "\n",
    "    #sort once more, by conc_data:\n",
    "    sorting_idxs = np.argsort(conc_data)\n",
    "    conc_data = conc_data[sorting_idxs]\n",
    "\n",
    "    #sort luminance data by conc_data:\n",
    "    luminance_data = luminance_data[:,:,sorting_idxs]\n",
    "    \n",
    "    return chamber_idxs, luminance_data, conc_data, time_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ### SANITY CHECK\n",
    "# chamber_idxs, luminance_data, conc_data, time_data = assay_data_to_array(db_conn, 'kinetics_0')\n",
    "# print(conc_data.shape) #expecting (11,)\n",
    "# print(luminance_data.shape) #expecting (20, 1792, 11)\n",
    "# print(time_data.shape) #expecting (20, 11)\n",
    "# print(chamber_idxs.shape) #expecting (1792,)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "TODO: fix code up to here, now that I unified the formats between kinetics and standard curves. It doesn't like the dimensionality of everything...\n",
    "actually - fixed?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1, 1792, 7)\n"
     ]
    }
   ],
   "source": [
    "#since we're just doing a linear fit to all the data here, we don't need the kinetics package.\n",
    "\n",
    "#Here, we collect our data into numpy arrays\n",
    "chamber_idxs, luminance_data, conc_data, _ = assay_data_to_array(db_conn, 'standard_0')\n",
    "\n",
    "#shape should be (n_assays, n_chambers)\n",
    "# 6 x (32*56=1792)\n",
    "print(luminance_data.shape)\n",
    "\n",
    "#Perform linear regression for each chamber:\n",
    "#and store immediately in 'database' object\n",
    "if 'analyses' not in db_conn._json_dict['runs']['standard_0'].keys():\n",
    "    db_conn._json_dict['runs']['standard_0']['analyses'] = {}\n",
    "\n",
    "db_conn._json_dict['runs']['standard_0']['analyses']['linear_regression'] = {'chambers': {}} #initialize the dictionary\n",
    "for i in range(len(chamber_idxs)):\n",
    "    slope, intercept, r_value, p_value, std_err = scipy.stats.linregress(conc_data, luminance_data[:,i])\n",
    "    db_conn._json_dict['runs']['standard_0']['analyses']['linear_regression']['chambers'][chamber_idxs[i]] = {'slope': slope, 'intercept': intercept, 'r_value': r_value, 'r2':r_value**2, 'p_value': p_value, 'std_err': std_err}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "chamber_metadata: \n",
      "\t{\n",
      "\t1,1: \n",
      "\t\t{\n",
      "\t\tid: M_caps_ADK\n",
      "\t\tx_center_chamber: 35.0\n",
      "\t\ty_center_chamber: 51.0\n",
      "\t\tradius_chamber: 35.0\n",
      "\t\txslice: (395, 495)\n",
      "\t\t...\n",
      "\t\t}\n",
      "\t1,2: \n",
      "\t\t{\n",
      "\t\tid: P_syri_ADK\n",
      "\t\tx_center_chamber: 35.0\n",
      "\t\ty_center_chamber: 51.0\n",
      "\t\tradius_chamber: 34.0\n",
      "\t\txslice: (510, 610)\n",
      "\t\t...\n",
      "\t\t}\n",
      "\t1,3: \n",
      "\t\t{\n",
      "\t\tid: L_reut_ADK\n",
      "\t\tx_center_chamber: 37.0\n",
      "\t\ty_center_chamber: 51.0\n",
      "\t\tradius_chamber: 37.0\n",
      "\t\txslice: (626, 726)\n",
      "\t\t...\n",
      "\t\t}\n",
      "\t1,4: \n",
      "\t\t{\n",
      "\t\tid: G_ther_ADK\n",
      "\t\tx_center_chamber: 37.0\n",
      "\t\ty_center_chamber: 49.0\n",
      "\t\tradius_chamber: 36.0\n",
      "\t\txslice: (742, 842)\n",
      "\t\t...\n",
      "\t\t}\n",
      "\t1,5: \n",
      "\t\t{\n",
      "\t\tid: bmADK\n",
      "\t\tx_center_chamber: 35.0\n",
      "\t\ty_center_chamber: 49.0\n",
      "\t\tradius_chamber: 34.0\n",
      "\t\txslice: (858, 958)\n",
      "\t\t...\n",
      "\t\t}\n",
      "\t...\n",
      "\t}\n",
      "runs: \n",
      "\t{\n",
      "\tstandard_0: \n",
      "\t\t{\n",
      "\t\tname: NADPH std curve\n",
      "\t\ttype: NADPH\n",
      "\t\tconc_unit: uM\n",
      "\t\tassays: \n",
      "\t\t\t{\n",
      "\t\t\t0: \n",
      "\t\t\t\t{\n",
      "\t\t\t\tconc: 3.90625\n",
      "\t\t\t\ttime_s: [0]\n",
      "\t\t\t\tchambers: \n",
      "\t\t\t\t\t{\n",
      "\t\t\t\t\t1,1: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [188096.0]\n",
      "\t\t\t\t\t\tstd_chamber: [78.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,10: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [228483.0]\n",
      "\t\t\t\t\t\tstd_chamber: [99.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,11: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [224214.0]\n",
      "\t\t\t\t\t\tstd_chamber: [84.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,12: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [196095.0]\n",
      "\t\t\t\t\t\tstd_chamber: [80.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,13: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [234481.0]\n",
      "\t\t\t\t\t\tstd_chamber: [253.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t...\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\t\t\t1: \n",
      "\t\t\t\t{\n",
      "\t\t\t\tconc: 7.8125\n",
      "\t\t\t\ttime_s: [0]\n",
      "\t\t\t\tchambers: \n",
      "\t\t\t\t\t{\n",
      "\t\t\t\t\t1,1: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [240692.0]\n",
      "\t\t\t\t\t\tstd_chamber: [88.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,10: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [259851.0]\n",
      "\t\t\t\t\t\tstd_chamber: [105.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,11: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [267881.0]\n",
      "\t\t\t\t\t\tstd_chamber: [94.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,12: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [229483.0]\n",
      "\t\t\t\t\t\tstd_chamber: [85.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,13: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [266559.0]\n",
      "\t\t\t\t\t\tstd_chamber: [257.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t...\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\t\t\t2: \n",
      "\t\t\t\t{\n",
      "\t\t\t\tconc: 15.625\n",
      "\t\t\t\ttime_s: [0]\n",
      "\t\t\t\tchambers: \n",
      "\t\t\t\t\t{\n",
      "\t\t\t\t\t1,1: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [330890.0]\n",
      "\t\t\t\t\t\tstd_chamber: [104.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,10: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [358876.0]\n",
      "\t\t\t\t\t\tstd_chamber: [119.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,11: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [367405.0]\n",
      "\t\t\t\t\t\tstd_chamber: [106.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,12: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [349277.0]\n",
      "\t\t\t\t\t\tstd_chamber: [103.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,13: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [368497.0]\n",
      "\t\t\t\t\t\tstd_chamber: [264.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t...\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\t\t\t3: \n",
      "\t\t\t\t{\n",
      "\t\t\t\tconc: 31.25\n",
      "\t\t\t\ttime_s: [0]\n",
      "\t\t\t\tchambers: \n",
      "\t\t\t\t\t{\n",
      "\t\t\t\t\t1,1: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [644967.0]\n",
      "\t\t\t\t\t\tstd_chamber: [141.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,10: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [643179.0]\n",
      "\t\t\t\t\t\tstd_chamber: [147.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,11: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [681549.0]\n",
      "\t\t\t\t\t\tstd_chamber: [138.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,12: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [656808.0]\n",
      "\t\t\t\t\t\tstd_chamber: [134.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,13: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [655309.0]\n",
      "\t\t\t\t\t\tstd_chamber: [272.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t...\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\t\t\t4: \n",
      "\t\t\t\t{\n",
      "\t\t\t\tconc: 125.0\n",
      "\t\t\t\ttime_s: [0]\n",
      "\t\t\t\tchambers: \n",
      "\t\t\t\t\t{\n",
      "\t\t\t\t\t1,1: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [2990428.0]\n",
      "\t\t\t\t\t\tstd_chamber: [327.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,10: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [2682125.0]\n",
      "\t\t\t\t\t\tstd_chamber: [289.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,11: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [2997113.0]\n",
      "\t\t\t\t\t\tstd_chamber: [323.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,12: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [3050228.0]\n",
      "\t\t\t\t\t\tstd_chamber: [293.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,13: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [2964056.0]\n",
      "\t\t\t\t\t\tstd_chamber: [399.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t...\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\t\t\t...\n",
      "\t\t\t}\n",
      "\t\tanalyses: \n",
      "\t\t\t{\n",
      "\t\t\tlinear_regression: \n",
      "\t\t\t\t{\n",
      "\t\t\t\tchambers: \n",
      "\t\t\t\t\t{\n",
      "\t\t\t\t\t1,1: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tslope: 22733.577962000814\n",
      "\t\t\t\t\t\tintercept: 43975.0999340429\n",
      "\t\t\t\t\t\tr_value: 0.9998427526217066\n",
      "\t\t\t\t\t\tr2: 0.9996855299701513\n",
      "\t\t\t\t\t\tp_value: 5.954918030206768e-10\n",
      "\t\t\t\t\t\t...\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,2: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tslope: 23678.734586242845\n",
      "\t\t\t\t\t\tintercept: 36383.9117678348\n",
      "\t\t\t\t\t\tr_value: 0.9998331684188061\n",
      "\t\t\t\t\t\tr2: 0.9996663646703887\n",
      "\t\t\t\t\t\tp_value: 6.904157486009038e-10\n",
      "\t\t\t\t\t\t...\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,3: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tslope: 25638.041855449883\n",
      "\t\t\t\t\t\tintercept: -12384.234069488011\n",
      "\t\t\t\t\t\tr_value: 0.9998221003666415\n",
      "\t\t\t\t\t\tr2: 0.9996442323815626\n",
      "\t\t\t\t\t\tp_value: 8.106811626021826e-10\n",
      "\t\t\t\t\t\t...\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,4: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tslope: 25345.88627407928\n",
      "\t\t\t\t\t\tintercept: 11886.140901256353\n",
      "\t\t\t\t\t\tr_value: 0.9998277064263551\n",
      "\t\t\t\t\t\tr2: 0.9996554425377857\n",
      "\t\t\t\t\t\tp_value: 7.483184467077394e-10\n",
      "\t\t\t\t\t\t...\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,5: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tslope: 23595.500598880877\n",
      "\t\t\t\t\t\tintercept: 8589.301823365036\n",
      "\t\t\t\t\t\tr_value: 0.9998446234446491\n",
      "\t\t\t\t\t\tr2: 0.9996892710311721\n",
      "\t\t\t\t\t\tp_value: 5.779382148695999e-10\n",
      "\t\t\t\t\t\t...\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t...\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\t\t\t}\n",
      "\t\t}\n",
      "\tkinetics_0: \n",
      "\t\t{\n",
      "\t\tname: ADP kinetics curve\n",
      "\t\ttype: ADP\n",
      "\t\tconc_unit: uM\n",
      "\t\tassays: \n",
      "\t\t\t{\n",
      "\t\t\t0: \n",
      "\t\t\t\t{\n",
      "\t\t\t\tconc: 1000.0\n",
      "\t\t\t\ttime_s: [0, 107, 1257, 143, 1557, 179, 1857, 214, 250, 286, 322, 357, 36, 417, 477, 537, 597, 657, 71, 957]\n",
      "\t\t\t\tchambers: \n",
      "\t\t\t\t\t{\n",
      "\t\t\t\t\t1,1: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [5706550.0, 6111047.0, 11687580.0, 6252476.0, 12771839.0, 6496490.0, 13601648.0, 6560664.0, 6807164.0, 6904110.0, 7100560.0, 7285440.0, 5779893.0, 7742776.0, 8097891.0, 8364556.0, 8657411.0, 8870927.0, 5918860.0, 10368417.0]\n",
      "\t\t\t\t\t\tstd_chamber: [387.0, 478.0, 1495.0, 513.0, 1694.0, 548.0, 1861.0, 583.0, 627.0, 658.0, 696.0, 724.0, 409.0, 796.0, 853.0, 909.0, 969.0, 1015.0, 441.0, 1271.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,10: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [993859.0, 6815357.0, 21599775.0, 8271935.0, 22307698.0, 9475365.0, 22550946.0, 10476613.0, 11408539.0, 12161098.0, 12919157.0, 13566534.0, 2989408.0, 14696183.0, 15726882.0, 16494785.0, 17213104.0, 17816508.0, 5088504.0, 20274702.0]\n",
      "\t\t\t\t\t\tstd_chamber: [219.0, 1210.0, 3669.0, 1461.0, 3772.0, 1669.0, 3801.0, 1843.0, 2000.0, 2136.0, 2265.0, 2370.0, 546.0, 2558.0, 2727.0, 2859.0, 2975.0, 3078.0, 908.0, 3469.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,11: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [1019894.0, 1283020.0, 5306931.0, 1424235.0, 6137206.0, 1558000.0, 6822833.0, 1673322.0, 1811411.0, 1913603.0, 2042143.0, 2168285.0, 1088892.0, 2451438.0, 2723241.0, 2939916.0, 3129633.0, 3308864.0, 1170345.0, 4381145.0]\n",
      "\t\t\t\t\t\tstd_chamber: [211.0, 261.0, 845.0, 283.0, 961.0, 301.0, 1071.0, 322.0, 342.0, 361.0, 378.0, 399.0, 222.0, 438.0, 473.0, 502.0, 534.0, 563.0, 241.0, 714.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,12: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [1018086.0, 5533008.0, 22021641.0, 6805264.0, 23610365.0, 7888695.0, 24491342.0, 8849341.0, 9713049.0, 10436485.0, 11202118.0, 11878670.0, 2472329.0, 13062224.0, 14126802.0, 14966136.0, 15785480.0, 16511326.0, 4099259.0, 19723086.0]\n",
      "\t\t\t\t\t\tstd_chamber: [217.0, 1003.0, 3773.0, 1225.0, 4027.0, 1413.0, 4162.0, 1577.0, 1730.0, 1860.0, 1986.0, 2105.0, 465.0, 2302.0, 2478.0, 2621.0, 2759.0, 2880.0, 748.0, 3411.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,13: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [893434.0, 1562758.0, 8276068.0, 1854220.0, 9517212.0, 2115900.0, 10547386.0, 2365359.0, 2602063.0, 2817998.0, 3057931.0, 3296254.0, 1076345.0, 3749090.0, 4181894.0, 4546259.0, 4891835.0, 5217944.0, 1289226.0, 6855831.0]\n",
      "\t\t\t\t\t\tstd_chamber: [321.0, 409.0, 1345.0, 441.0, 1525.0, 480.0, 1676.0, 514.0, 545.0, 581.0, 613.0, 647.0, 339.0, 705.0, 766.0, 816.0, 866.0, 915.0, 367.0, 1147.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t...\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\t\t\t1: \n",
      "\t\t\t\t{\n",
      "\t\t\t\tconc: 125.0\n",
      "\t\t\t\ttime_s: [0, 107, 1258, 143, 1558, 179, 1858, 214, 250, 286, 322, 358, 36, 418, 478, 538, 598, 658, 71, 958]\n",
      "\t\t\t\tchambers: \n",
      "\t\t\t\t\t{\n",
      "\t\t\t\t\t1,1: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [8136071.0, 8779072.0, 12297277.0, 8973313.0, 12711744.0, 9135669.0, 12874143.0, 9337468.0, 9460129.0, 9571357.0, 9656313.0, 9815460.0, 8274315.0, 10101405.0, 10303423.0, 10462064.0, 10687599.0, 10734019.0, 8517578.0, 11716124.0]\n",
      "\t\t\t\t\t\tstd_chamber: [487.0, 511.0, 755.0, 519.0, 795.0, 531.0, 801.0, 539.0, 552.0, 565.0, 576.0, 584.0, 488.0, 604.0, 615.0, 635.0, 649.0, 658.0, 498.0, 715.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,10: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [625410.0, 3474955.0, 6217674.0, 3918941.0, 6377513.0, 4242168.0, 6363578.0, 4512960.0, 4710216.0, 4847984.0, 4966483.0, 5048604.0, 1780525.0, 5245305.0, 5388851.0, 5491937.0, 5573753.0, 5632528.0, 2794202.0, 6016738.0]\n",
      "\t\t\t\t\t\tstd_chamber: [195.0, 418.0, 639.0, 456.0, 647.0, 484.0, 643.0, 510.0, 528.0, 541.0, 551.0, 562.0, 277.0, 578.0, 588.0, 597.0, 601.0, 607.0, 353.0, 629.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,11: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [540378.0, 738863.0, 2930814.0, 807846.0, 3332516.0, 866794.0, 3533592.0, 955709.0, 1045780.0, 1095217.0, 1154148.0, 1224701.0, 592045.0, 1363423.0, 1517568.0, 1633090.0, 1759191.0, 1836781.0, 680949.0, 2467240.0]\n",
      "\t\t\t\t\t\tstd_chamber: [132.0, 156.0, 345.0, 165.0, 376.0, 170.0, 396.0, 181.0, 191.0, 195.0, 199.0, 207.0, 140.0, 221.0, 236.0, 244.0, 254.0, 261.0, 152.0, 310.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,12: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [567603.0, 3128404.0, 6835637.0, 3622077.0, 7062621.0, 4005722.0, 7068715.0, 4325448.0, 4580633.0, 4786534.0, 4959712.0, 5110547.0, 1525622.0, 5365631.0, 5581630.0, 5739273.0, 5883106.0, 5966593.0, 2458335.0, 6557399.0]\n",
      "\t\t\t\t\t\tstd_chamber: [137.0, 369.0, 685.0, 416.0, 695.0, 455.0, 701.0, 485.0, 509.0, 529.0, 543.0, 561.0, 225.0, 585.0, 599.0, 614.0, 626.0, 631.0, 305.0, 669.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,13: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [414467.0, 950454.0, 4466705.0, 1143795.0, 4952392.0, 1311376.0, 5244840.0, 1483921.0, 1649943.0, 1766141.0, 1906750.0, 2032047.0, 562480.0, 2279292.0, 2516475.0, 2703785.0, 2876810.0, 3023237.0, 763219.0, 3857929.0]\n",
      "\t\t\t\t\t\tstd_chamber: [257.0, 296.0, 515.0, 311.0, 538.0, 330.0, 561.0, 334.0, 347.0, 354.0, 364.0, 372.0, 273.0, 380.0, 401.0, 416.0, 423.0, 428.0, 286.0, 481.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t...\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\t\t\t2: \n",
      "\t\t\t\t{\n",
      "\t\t\t\tconc: 15.625\n",
      "\t\t\t\ttime_s: [0, 107, 1256, 142, 1556, 178, 1856, 214, 250, 285, 321, 35, 356, 416, 476, 536, 596, 656, 71, 956]\n",
      "\t\t\t\tchambers: \n",
      "\t\t\t\t\t{\n",
      "\t\t\t\t\t1,1: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [468947.0, 495223.0, 933946.0, 523283.0, 912354.0, 544724.0, 917781.0, 570212.0, 579292.0, 572060.0, 600666.0, 456552.0, 610655.0, 642864.0, 636502.0, 647804.0, 661358.0, 702630.0, 458014.0, 800365.0]\n",
      "\t\t\t\t\t\tstd_chamber: [135.0, 142.0, 206.0, 148.0, 205.0, 150.0, 202.0, 157.0, 157.0, 156.0, 160.0, 134.0, 165.0, 169.0, 166.0, 172.0, 172.0, 178.0, 137.0, 192.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,10: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [490584.0, 753449.0, 996616.0, 780848.0, 956578.0, 815317.0, 967917.0, 832076.0, 824184.0, 821190.0, 830193.0, 599113.0, 802926.0, 829600.0, 815588.0, 813541.0, 821437.0, 849516.0, 679183.0, 894883.0]\n",
      "\t\t\t\t\t\tstd_chamber: [143.0, 181.0, 206.0, 187.0, 203.0, 190.0, 208.0, 192.0, 192.0, 193.0, 194.0, 159.0, 193.0, 195.0, 191.0, 191.0, 193.0, 196.0, 171.0, 201.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,11: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [535340.0, 485538.0, 818524.0, 496728.0, 801014.0, 499421.0, 842316.0, 521427.0, 508488.0, 489754.0, 515960.0, 513026.0, 488340.0, 517563.0, 509119.0, 539875.0, 532066.0, 580712.0, 490005.0, 663461.0]\n",
      "\t\t\t\t\t\tstd_chamber: [136.0, 133.0, 174.0, 133.0, 172.0, 134.0, 175.0, 138.0, 135.0, 134.0, 137.0, 134.0, 136.0, 141.0, 136.0, 141.0, 141.0, 149.0, 132.0, 158.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,12: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [470655.0, 721218.0, 1080245.0, 775635.0, 1014988.0, 801196.0, 1028059.0, 835863.0, 835737.0, 829286.0, 836905.0, 570745.0, 833564.0, 854152.0, 836863.0, 852024.0, 862002.0, 883963.0, 647772.0, 944915.0]\n",
      "\t\t\t\t\t\tstd_chamber: [136.0, 172.0, 212.0, 180.0, 205.0, 184.0, 205.0, 188.0, 188.0, 188.0, 189.0, 150.0, 189.0, 191.0, 192.0, 192.0, 192.0, 197.0, 164.0, 200.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,13: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [478166.0, 489205.0, 947494.0, 510785.0, 925715.0, 545181.0, 966055.0, 563899.0, 570733.0, 570826.0, 600189.0, 470683.0, 568195.0, 620881.0, 617719.0, 642148.0, 669995.0, 691957.0, 468226.0, 815515.0]\n",
      "\t\t\t\t\t\tstd_chamber: [285.0, 282.0, 324.0, 281.0, 320.0, 294.0, 321.0, 292.0, 287.0, 292.0, 291.0, 280.0, 286.0, 295.0, 296.0, 294.0, 298.0, 296.0, 278.0, 313.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t...\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\t\t\t3: \n",
      "\t\t\t\t{\n",
      "\t\t\t\tconc: 2000.0\n",
      "\t\t\t\ttime_s: [0, 107, 1257, 143, 1557, 179, 1857, 214, 250, 286, 322, 357, 36, 417, 477, 537, 597, 657, 72, 957]\n",
      "\t\t\t\tchambers: \n",
      "\t\t\t\t\t{\n",
      "\t\t\t\t\t1,1: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [5627998.0, 5920694.0, 11383704.0, 6090270.0, 12552552.0, 6207816.0, 13624902.0, 6405748.0, 6614038.0, 6754358.0, 6936485.0, 7094913.0, 5655647.0, 7415492.0, 7796819.0, 8094819.0, 8369920.0, 8615189.0, 5786569.0, 10173032.0]\n",
      "\t\t\t\t\t\tstd_chamber: [417.0, 514.0, 1603.0, 546.0, 1819.0, 584.0, 2014.0, 625.0, 664.0, 698.0, 743.0, 775.0, 447.0, 836.0, 903.0, 966.0, 1018.0, 1079.0, 477.0, 1354.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,10: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [1363446.0, 8447498.0, 22000192.0, 10365329.0, 22147464.0, 11983564.0, 22222185.0, 13361742.0, 14569287.0, 15590652.0, 16573526.0, 17380636.0, 3565965.0, 18594608.0, 19547522.0, 20176072.0, 20646472.0, 20943412.0, 6224505.0, 21898281.0]\n",
      "\t\t\t\t\t\tstd_chamber: [275.0, 1477.0, 3752.0, 1801.0, 3759.0, 2087.0, 3759.0, 2316.0, 2531.0, 2715.0, 2872.0, 3012.0, 638.0, 3215.0, 3374.0, 3476.0, 3547.0, 3604.0, 1089.0, 3743.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,11: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [1503895.0, 1802314.0, 6234549.0, 1955100.0, 7216987.0, 2090102.0, 8109026.0, 2225588.0, 2369030.0, 2491170.0, 2652715.0, 2781230.0, 1541312.0, 3068866.0, 3345926.0, 3568084.0, 3815430.0, 4051724.0, 1678663.0, 5292879.0]\n",
      "\t\t\t\t\t\tstd_chamber: [278.0, 335.0, 1016.0, 357.0, 1166.0, 381.0, 1303.0, 406.0, 430.0, 449.0, 474.0, 494.0, 290.0, 532.0, 576.0, 612.0, 649.0, 688.0, 313.0, 870.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,12: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [1457764.0, 6958205.0, 24322089.0, 8642840.0, 24652175.0, 10088833.0, 24838996.0, 11393315.0, 12571546.0, 13664576.0, 14693028.0, 15598688.0, 3047405.0, 17116710.0, 18418610.0, 19535541.0, 20471010.0, 21219041.0, 5105053.0, 23729057.0]\n",
      "\t\t\t\t\t\tstd_chamber: [289.0, 1234.0, 4179.0, 1526.0, 4221.0, 1777.0, 4231.0, 2005.0, 2213.0, 2408.0, 2580.0, 2739.0, 559.0, 2997.0, 3218.0, 3401.0, 3559.0, 3688.0, 912.0, 4092.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,13: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [1386957.0, 2093055.0, 9623520.0, 2387479.0, 11142535.0, 2683778.0, 12518506.0, 2947950.0, 3201093.0, 3468491.0, 3754681.0, 3999817.0, 1509319.0, 4489910.0, 4909632.0, 5344920.0, 5758692.0, 6144362.0, 1817168.0, 8086768.0]\n",
      "\t\t\t\t\t\tstd_chamber: [376.0, 472.0, 1571.0, 509.0, 1792.0, 554.0, 1996.0, 592.0, 630.0, 666.0, 709.0, 746.0, 393.0, 819.0, 874.0, 934.0, 998.0, 1055.0, 434.0, 1341.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t...\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\t\t\t4: \n",
      "\t\t\t\t{\n",
      "\t\t\t\tconc: 250.0\n",
      "\t\t\t\ttime_s: [0, 107, 1258, 143, 1558, 179, 1858, 215, 250, 286, 322, 358, 36, 418, 478, 538, 598, 658, 71, 958]\n",
      "\t\t\t\tchambers: \n",
      "\t\t\t\t\t{\n",
      "\t\t\t\t\t1,1: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [7535584.0, 8382561.0, 14100299.0, 8760094.0, 14987080.0, 9017251.0, 15508256.0, 9292434.0, 9562967.0, 9802961.0, 10015168.0, 10258792.0, 7734090.0, 10656319.0, 11049251.0, 11329217.0, 11528167.0, 11800331.0, 8072487.0, 13333677.0]\n",
      "\t\t\t\t\t\tstd_chamber: [441.0, 483.0, 1002.0, 511.0, 1086.0, 530.0, 1137.0, 550.0, 576.0, 598.0, 619.0, 638.0, 449.0, 675.0, 711.0, 738.0, 763.0, 784.0, 465.0, 927.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,10: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [880995.0, 5370831.0, 11811674.0, 6296887.0, 12157938.0, 6981369.0, 12257941.0, 7521239.0, 7995276.0, 8345707.0, 8723488.0, 8995632.0, 2621311.0, 9471998.0, 9826605.0, 10097197.0, 10342822.0, 10527784.0, 4215896.0, 11501721.0]\n",
      "\t\t\t\t\t\tstd_chamber: [214.0, 578.0, 1158.0, 664.0, 1183.0, 729.0, 1182.0, 785.0, 821.0, 865.0, 893.0, 923.0, 336.0, 965.0, 998.0, 1022.0, 1042.0, 1063.0, 469.0, 1138.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,11: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [775713.0, 1071546.0, 4143507.0, 1204379.0, 4734995.0, 1319881.0, 5129872.0, 1431101.0, 1568418.0, 1651551.0, 1791274.0, 1883931.0, 853886.0, 2093719.0, 2287495.0, 2465715.0, 2632796.0, 2746208.0, 980488.0, 3622744.0]\n",
      "\t\t\t\t\t\tstd_chamber: [153.0, 184.0, 457.0, 198.0, 504.0, 210.0, 543.0, 221.0, 231.0, 242.0, 250.0, 259.0, 162.0, 278.0, 293.0, 310.0, 322.0, 335.0, 176.0, 403.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,12: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [850858.0, 4675163.0, 12579265.0, 5541853.0, 13171457.0, 6241412.0, 13397844.0, 6832114.0, 7367405.0, 7779092.0, 8202876.0, 8534325.0, 2243673.0, 9115483.0, 9607137.0, 9979384.0, 10299599.0, 10591028.0, 3594942.0, 11989015.0]\n",
      "\t\t\t\t\t\tstd_chamber: [157.0, 504.0, 1230.0, 583.0, 1275.0, 657.0, 1293.0, 711.0, 761.0, 803.0, 841.0, 878.0, 275.0, 929.0, 974.0, 1006.0, 1036.0, 1068.0, 402.0, 1179.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t1,13: \n",
      "\t\t\t\t\t\t{\n",
      "\t\t\t\t\t\tsum_chamber: [618191.0, 1413115.0, 6585653.0, 1685668.0, 7410296.0, 1953105.0, 7971561.0, 2184447.0, 2461198.0, 2669471.0, 2875424.0, 3065539.0, 838047.0, 3453987.0, 3761130.0, 4034591.0, 4274868.0, 4526082.0, 1117964.0, 5789835.0]\n",
      "\t\t\t\t\t\tstd_chamber: [273.0, 329.0, 668.0, 340.0, 716.0, 364.0, 760.0, 373.0, 390.0, 403.0, 420.0, 428.0, 289.0, 455.0, 470.0, 492.0, 514.0, 528.0, 311.0, 602.0]\n",
      "\t\t\t\t\t\t}\n",
      "\t\t\t\t\t...\n",
      "\t\t\t\t\t}\n",
      "\t\t\t\t}\n",
      "\t\t\t...\n",
      "\t\t\t}\n",
      "\t\t}\n",
      "\t}\n",
      "button_quant: \n",
      "\t{\n",
      "\t1,1: \n",
      "\t\t{\n",
      "\t\tsummed_button_Button_Quant: 15041\n",
      "\t\tsummed_button_BGsub_Button_Quant: 7088\n",
      "\t\tstd_button_Button_Quant: 21\n",
      "\t\t}\n",
      "\t1,2: \n",
      "\t\t{\n",
      "\t\tsummed_button_Button_Quant: 21514\n",
      "\t\tsummed_button_BGsub_Button_Quant: 6570\n",
      "\t\tstd_button_Button_Quant: 17\n",
      "\t\t}\n",
      "\t1,3: \n",
      "\t\t{\n",
      "\t\tsummed_button_Button_Quant: 152498\n",
      "\t\tsummed_button_BGsub_Button_Quant: 122704\n",
      "\t\tstd_button_Button_Quant: 163\n",
      "\t\t}\n",
      "\t1,4: \n",
      "\t\t{\n",
      "\t\tsummed_button_Button_Quant: 276149\n",
      "\t\tsummed_button_BGsub_Button_Quant: 239605\n",
      "\t\tstd_button_Button_Quant: 296\n",
      "\t\t}\n",
      "\t1,5: \n",
      "\t\t{\n",
      "\t\tsummed_button_Button_Quant: 1085272\n",
      "\t\tsummed_button_BGsub_Button_Quant: 1051853\n",
      "\t\tstd_button_Button_Quant: 384\n",
      "\t\t}\n",
      "\t...\n",
      "\t}\n",
      "}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(db_conn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for i in range(len(chamber_idxs)):\n",
    "#     #slope, intercept, r_value, p_value, std_err = scipy.stats.linregress(conc_data[:,i], luminance_data[:,i])\n",
    "#     print(db_conn._json_dict['runs']['standard_0']['analyses']['linear_regression']['chambers'][chamber_idxs[i]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_curves_with_var(db_obj, run_name, analysis_name, analysis_var, title=None):\n",
    "    ''' This function creates a Dash visualization of a chip, based on a certain Run (run_name)\n",
    "        Each well will be colored by a variable that is specified by analysis_var, from the analysis specified by analysis_name.\n",
    "        For example, to color the chip based on the slope of the linear regression, you would set analysis_var='slope' and analysis_name='linear_regression'\n",
    "    '''\n",
    "    from dash import Dash, dcc, html, Input, Output, no_update\n",
    "    import plotly.graph_objs as go\n",
    "    import base64\n",
    "    import tempfile\n",
    "\n",
    "    #get chamber ids:\n",
    "    chamber_idxs, luminance_data, conc_data, time_data = assay_data_to_array(db_conn, 'standard_0')\n",
    "    chamber_ids = np.array([db_obj._json_dict['chamber_metadata'][chamber_idx]['id'] for chamber_idx in chamber_idxs])\n",
    "    \n",
    "    #get slope from the analysis:\n",
    "    slope = np.array([db_obj._json_dict['runs'][run_name]['analyses'][analysis_name]['chambers'][chamber_idx]['slope'] for chamber_idx in chamber_idxs])\n",
    "    intercept = np.array([db_obj._json_dict['runs'][run_name]['analyses'][analysis_name]['chambers'][chamber_idx]['intercept'] for chamber_idx in chamber_idxs])\n",
    "    #get our special var, specified by 'analysis_var', to be used for coloring:\n",
    "    coloring_var = np.array([db_obj._json_dict['runs'][run_name]['analyses'][analysis_name]['chambers'][chamber_idx][analysis_var] for chamber_idx in chamber_idxs])\n",
    "\n",
    "    # Make the image array\n",
    "    #N.B. eventually, store width/height in DB and reference!\n",
    "    img_array = np.zeros([56,32])\n",
    "\n",
    "    for i, chamber_id in enumerate(chamber_idxs):\n",
    "        x = int(chamber_id.split(',')[0])\n",
    "        y = int(chamber_id.split(',')[1])\n",
    "        img_array[y-1,x-1] = coloring_var[i]\n",
    "    \n",
    "    #generate title\n",
    "    if title is None:\n",
    "        title = analysis_var + ' by chamber, for ' + run_name\n",
    "\n",
    "    #Create the figure\n",
    "    layout = go.Layout()\n",
    "    fig = go.Figure(layout=layout, data=go.Heatmap(z=img_array, colorscale='Viridis'))\n",
    "    fig.update_layout(title=title, \n",
    "                        yaxis=dict(scaleanchor=\"x\", scaleratio=1, autorange='reversed'), \n",
    "                        xaxis=dict(scaleratio=1),\n",
    "                        plot_bgcolor='rgba(0,0,0,0)',\n",
    "                        width=600, height=600,\n",
    "                        hovermode='x')\n",
    "    fig.update_xaxes(showticklabels=False)\n",
    "    fig.update_yaxes(showticklabels=False)\n",
    "\n",
    "    #create dash app:\n",
    "    app = Dash(__name__)\n",
    "    app.layout = html.Div([\n",
    "        dcc.Graph(id=\"graph\", figure=fig, clear_on_unhover=True),\n",
    "        dcc.Tooltip(id=\"graph-tooltip\"),\n",
    "    ])\n",
    "\n",
    "    @app.callback(\n",
    "        Output(\"graph-tooltip\", \"show\"),\n",
    "        Output(\"graph-tooltip\", \"bbox\"),\n",
    "        Output(\"graph-tooltip\", \"children\"),\n",
    "        Input(\"graph\", \"hoverData\"),\n",
    "    )\n",
    "    def display_hover(hoverData):\n",
    "        if hoverData is None:\n",
    "            return False, no_update, no_update\n",
    "\n",
    "        # demo only shows the first point, but other points may also be available\n",
    "        pt = hoverData[\"points\"][0]\n",
    "        bbox = pt[\"bbox\"]\n",
    "\n",
    "        #get the data for the point:\n",
    "        \n",
    "        data_name = str(pt['x']+1) + ',' + str(pt['y']+1)\n",
    "        data_id = list(chamber_idxs).index(data_name)\n",
    "        x_data = conc_data\n",
    "        y_data = luminance_data[:,data_id]\n",
    "        m = slope[data_id]\n",
    "        b = intercept[data_id]\n",
    "        chamber_name = chamber_ids[pt['x']+pt['y']*32]\n",
    "\n",
    "        #make a simple matplotlib plot\n",
    "        fig, ax = plt.subplots()\n",
    "        ax.scatter(x_data, y_data)\n",
    "        if not (np.isnan(m) or np.isnan(b)):\n",
    "            #return False, no_update, no_update\n",
    "            ax.plot(x_data, m*np.array(x_data) + b)\n",
    "        #reduce whitespace on margins of graph:\n",
    "        fig.subplots_adjust(left=0.1, bottom=0.1, right=0.9, top=0.9, wspace=0, hspace=0)\n",
    "        #save the figure as a temp file:\n",
    "        tempfile_name = tempfile.NamedTemporaryFile().name+'.png'\n",
    "        plt.savefig(tempfile_name)\n",
    "        plt.close()\n",
    "\n",
    "        # #read in temp file as base64 encoded string:\n",
    "        with open(tempfile_name, \"rb\") as image_file:\n",
    "            img_src = \"data:image/png;base64,\" + str(base64.b64encode(image_file.read()).decode(\"utf-8\"))\n",
    "        \n",
    "        children = [\n",
    "            html.Div(children=[\n",
    "                #no space after header:\n",
    "                html.H3('{},{}:  {}'.format(pt['x'], pt['y'], chamber_name), style={\"color\": 'black', \"fontFamily\":\"Arial\", \"textAlign\": \"center\", \"marginBottom\": \"0px\"}),\n",
    "                #add the image with reduced whitespace:\n",
    "                html.Img(src=img_src, style={\"width\": \"100%\"}),\n",
    "            ],\n",
    "            style={'width': '400px', 'white-space': 'none'})\n",
    "        ]\n",
    "\n",
    "        return True, bbox, children\n",
    "    \n",
    "    app.run_server()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"100%\"\n",
       "            height=\"650\"\n",
       "            src=\"http://127.0.0.1:8050/\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "            \n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x7f324bfaae30>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_curves_with_var(db_conn, 'standard_0', 'linear_regression', 'slope')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The following is old code, which I have here to reference as I fix the rest:\n",
    "\n",
    "# #takes a DataFrame df as input and performs a linear regression analysis on the data in the DataFrame\n",
    "# def fit_standard_linear(df):\n",
    "    \n",
    "#     # Make a copy of the input DataFrame and sort it based on 'concentration_uM' column.\n",
    "#     dfcopy = df.copy().sort_values('concentration_uM')\n",
    "    \n",
    "#     # Perform linear regression on the 'concentration_uM' and 'sum_chamber' columns of the DataFrame.\n",
    "#     slope, intercept, r_value, p_value, std_err = scipy.stats.linregress(df.concentration_uM, df.sum_chamber)\n",
    "\n",
    "#     # Create a new pandas Series to store the results of the linear regression analysis.\n",
    "#     # Define specific index names to store each result.\n",
    "#     fitted_series = pd.Series(data = [df.iloc[0].id,\n",
    "#                       dfcopy.iloc[0].x, dfcopy.iloc[0].y, \n",
    "#                       dfcopy.concentration_uM.values.tolist(), \n",
    "#                       dfcopy.sum_chamber.values.tolist(),\n",
    "#                       dfcopy.std_chamber.values.tolist(),\n",
    "#                      slope, intercept, r_value**2, p_value, std_err],\n",
    "#               index = ['id', 'x', 'y', 'std_concs', 'std_summed_intensity', 'std_stddev_intensity', 'std_slope', \n",
    "#                        'std_intercept', 'std_r2', 'std_pvalue', 'std_stderr'])\n",
    "#     return fitted_series\n",
    "\n",
    "\n",
    "# # Fit linear regression by chamber\n",
    "# # Group the DataFrame 'standard_data' by the values in the 'indices' (i.e. chamber coord) column and apply the 'fit_standard_linear' function to each group. \n",
    "# # The result is a Series containing the output of linear regression analysis for each chamber\n",
    "\n",
    "# fitted_standards_filtered = standard_data.groupby('indices').apply(fit_standard_linear)\n",
    "\n",
    "# # Add additional descriptors\n",
    "# fitted_standards_filtered['std_slope_units'] = 'RFU/uM'\n",
    "# fitted_standards_filtered['substrate_name'] = 'ADP'\n",
    "# fitted_standards_filtered['experimental_day'] = '20230628'\n",
    "# fitted_standards_filtered['setup'] = '1'\n",
    "# fitted_standards_filtered['device'] = 'd2'\n",
    "# fitted_standards_filtered['assay_type'] = 'NADPH'\n",
    "# fitted_standards_filtered['position'] = [(int(r['y']),int(r['x'])) for i,r in fitted_standards_filtered.iterrows()]\n",
    "\n",
    "# fitted_standards_filtered = fitted_standards_filtered.sort_values(by = 'position')\n",
    "# fitted_standards_filtered.head(2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "#NF: completely replace\n",
    "# Is this useful? I don't think so.\n",
    "# plot how standard curve parameters vary based on location in the device\n",
    "\n",
    "# def plot_std_curve_parameters(df, filename = 'standard_curve_parameters'):\n",
    "#     # plot how slope and r2 vary accross the chip\n",
    "#     fig, ax = plt.subplots(nrows=2, ncols=3, figsize=(20, 10))\n",
    "#     plt.xticks(fontsize=6)\n",
    "#     plt.yticks(fontsize=6)\n",
    "    \n",
    "#     # Flatten the axes array for easy indexing\n",
    "#     ax = ax.flatten()\n",
    "    \n",
    "#     ax[0].scatter(df['x'], df['std_r2'], color = 'b')\n",
    "#     ax[0].set_xlabel('x-index', fontsize = 12)\n",
    "#     ax[0].set_ylabel('r2', fontsize = 12)\n",
    "    \n",
    "#     ax[1].scatter(df['y'], df['std_r2'], color = 'b')\n",
    "#     ax[1].set_xlabel('y-index', fontsize = 12)\n",
    "#     ax[1].set_ylabel('r2', fontsize = 12)\n",
    "    \n",
    "#     ax[2].hist(df['std_r2'])\n",
    "#     ax[2].set_xlabel('r2', fontsize = 12)\n",
    "#     ax[2].set_ylabel('count', fontsize = 12)\n",
    "    \n",
    "#     ax[3].scatter(df['x'], df['std_slope'], color = 'b')\n",
    "#     ax[3].set_xlabel('x-index', fontsize = 12)\n",
    "#     ax[3].set_ylabel('slope', fontsize = 12)\n",
    "    \n",
    "#     ax[4].scatter(df['y'], df['std_slope'], color = 'b')\n",
    "#     ax[4].set_xlabel('y-index', fontsize = 12)\n",
    "#     ax[4].set_ylabel('slope', fontsize = 12)\n",
    "    \n",
    "#     ax[5].hist(df['std_slope'])\n",
    "#     ax[5].set_xlabel('slope', fontsize = 12)\n",
    "#     ax[5].set_ylabel('count', fontsize = 12)\n",
    "    \n",
    "#     plt.tight_layout()\n",
    "#     plt.rcParams['pdf.fonttype'] = 42\n",
    "#     plt.savefig('{}.pdf'.format(filename), format='pdf')\n",
    "    \n",
    "\n",
    "# plot_std_curve_parameters(fitted_standards_filtered, 'NADPH_standard_curve_parameters')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # save final filtered standard curve\n",
    "# fitted_standards_filtered.to_csv(new_root+'/2_fitted_standards_filtered.csv')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. Fit initial rates from processed kinetic data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Clean and merge dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #NF: Need to completely rework how we track variables like substrate, concentrations with experimental data.\n",
    "\n",
    "# # define the substrate concentrations and corresponding file descriptions used in the kinetics experiment\n",
    "# # concentrations = [15.625,\n",
    "# #                  31.125,\n",
    "# #                  62.5,\n",
    "# #                  125,\n",
    "# #                  250,\n",
    "# #                  500,\n",
    "# #                  1000,\n",
    "# #                  2000,\n",
    "# #                  4000,\n",
    "# #                  8000,\n",
    "# #                  16000]\n",
    "\n",
    "# # kinetic_descriptions = ['7_8125uM_ADP',\n",
    "# #                         '15_625uM_ADP',\n",
    "# #                         '31_125uM_ADP',\n",
    "# #                         '62.5uM_ADP',\n",
    "# #                         '125uM_ADP',\n",
    "# #                         '250uM_ADP',\n",
    "# #                         '500uM_ADP',\n",
    "# #                         '1000uMADP',\n",
    "# #                         '2000uM_ADP',\n",
    "# #                         '4000uM_ADP',\n",
    "# #                         '8000uM_ADP']\n",
    "                        \n",
    "# substrate_conc_mapper = {d:c for (d,c) in zip(kinetic_descriptions, concentrations)}\n",
    "\n",
    "# # for filtering if a particular susbtrate concentration should not be processed\n",
    "# # mmfit_mask_mapper = {'7_8125uM_ADP': True, \n",
    "# #                      '15_625uM_ADP': True, \n",
    "# #                      '31_125uM_ADP': True, \n",
    "# #                      '62.5uM_ADP': True, \n",
    "# #                      '125uM_ADP': True, \n",
    "# #                      '250uM_ADP': True, \n",
    "# #                      '500uM_ADP': True,\n",
    "# #                      '1000uMADP': True,\n",
    "# #                      '2000uM_ADP': True,\n",
    "# #                      '4000uM_ADP': True,\n",
    "# #                      '8000uM_ADP': True}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #NF where are we generating this? we need to specify the file path.\n",
    "\n",
    "# #read in the kinetic data (chamber intensities vs time) processed above\n",
    "# full_kinetic_data = pd.read_csv(root+'/kinetics/d2_TitrationSeries_Analysis.csv.bz2')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #NF: Can we arrange our data such that we don't need to change format halfway through?\n",
    "\n",
    "# # takes a tidy DataFrame of chamber intensity values and timepoint as input and compresses these values \n",
    "# # into a single row by returning a series where intensity values and timepoints are each a list\n",
    "\n",
    "# def squeeze_kinetics(df):\n",
    "#     # Create a copy of the input DataFrame and sort it based on the 'time_s' column.\n",
    "#     dfcopy = df.copy().sort_values('time_s')\n",
    "\n",
    "#     # Extract relevant data for the first row of the sorted DataFrame.\n",
    "#     # These values are assumed to be the same for all rows in the group.\n",
    "#     chamber_id = dfcopy.iloc[0]['id']\n",
    "#     substrate_conc = dfcopy.iloc[0]['substrate_conc']\n",
    "#     mm_tofit_mask = dfcopy.iloc[0]['mm_tofit_mask']\n",
    "\n",
    "#     # Extract specific column data as lists.\n",
    "#     median_chamber = dfcopy.median_chamber.values.tolist()\n",
    "#     sum_chamber = dfcopy.sum_chamber.values.tolist()\n",
    "#     std_chamber = dfcopy.std_chamber.values.tolist()\n",
    "#     button_quant_raw = dfcopy.summed_button_Button_Quant.values.tolist()\n",
    "#     button_quant_bgsub = dfcopy.summed_button_BGsub_Button_Quant.values.tolist()\n",
    "#     times = dfcopy.time_s.values.tolist()\n",
    "    \n",
    "#     # Create a new pandas Series to store the extracted data.\n",
    "#     squeezed_series = pd.Series(data = [chamber_id, mm_tofit_mask, substrate_conc, times, median_chamber, \n",
    "#                                         sum_chamber, std_chamber, button_quant_raw, button_quant_bgsub],\n",
    "#                                index = ['id', 'mm_tofit_mask','substrate_conc', 'time_s', 'chamber_median', \n",
    "#                                         'chamber_sum', 'chamber_stddev', 'button_sum', 'button_sum_bgsub'])\n",
    "#     return squeezed_series\n",
    "\n",
    "# #NF: We do a similar thing above for standard curve, but slightly different. Must standardize (or remove)\n",
    "\n",
    "# # Add additional descriptors\n",
    "# full_kinetic_data['indices'] = full_kinetic_data.x.astype('str') + ',' + full_kinetic_data.y.astype('str')\n",
    "# full_kinetic_data['substrate_conc'] = full_kinetic_data.series_index.map(substrate_conc_mapper)\n",
    "# full_kinetic_data['mm_tofit_mask'] = full_kinetic_data.series_index.map(mmfit_mask_mapper)\n",
    "\n",
    "# # Create 'untidy' data by chamber position (\"indices\") and substrate concentration (\"series_index\")\n",
    "# squeezed_kinetic_data = full_kinetic_data.groupby(['indices', 'series_index']).apply(squeeze_kinetics).reset_index().set_index('indices')\n",
    "# squeezed_kinetic_data.head(2)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We will keep 1730 out of 1792 chambers.\n"
     ]
    }
   ],
   "source": [
    "#Here, we determine which chambers we want to discard.\n",
    "#We do this by choosing a Pearson r^2 value threshold, and discarding any chambers that have a lower r^2 value than that threshold.\n",
    "r2_threshold = 0.999\n",
    "\n",
    "chamber_idxs = np.array(list(db_conn._json_dict['chamber_metadata'].keys()))\n",
    "r2_values = np.array([db_conn._json_dict['runs']['standard_0']['analyses']['linear_regression']['chambers'][chamber_idx]['r2'] for chamber_idx in chamber_idxs])\n",
    "\n",
    "#find the indices of the chambers that we want to keep:\n",
    "chamber_idxs_to_keep = chamber_idxs[r2_values > r2_threshold]\n",
    "\n",
    "print('We will keep {} out of {} chambers.'.format(len(chamber_idxs_to_keep), len(chamber_idxs)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"100%\"\n",
       "            height=\"650\"\n",
       "            src=\"http://127.0.0.1:8050/\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "            \n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x7f325958ba00>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#NF: I should abstract this, since I'm using the same code to plot the data. Only difference is which data / what representation.\n",
    "\n",
    "#Sanity check by plotting:\n",
    "def plot_chip_by_idx(chamber_idx_to_keep, title=None):\n",
    "    ''' This function creates a Dash visualization of a chip, based on a certain Run (run_name)\n",
    "        Each well will be colored by a variable that is specified by analysis_var, from the analysis specified by analysis_name.\n",
    "        For example, to color the chip based on the slope of the linear regression, you would set analysis_var='slope' and analysis_name='linear_regression'\n",
    "    '''\n",
    "    from dash import Dash, dcc, html, Input, Output, no_update\n",
    "    import plotly.graph_objs as go\n",
    "    import base64\n",
    "    import tempfile\n",
    "\n",
    "    # Make the image array\n",
    "    #N.B. eventually, store width/height in DB and reference!\n",
    "    img_array = np.zeros([56,32])\n",
    "\n",
    "    for i, chamber_id in enumerate(chamber_idx_to_keep):\n",
    "        x = int(chamber_id.split(',')[0])\n",
    "        y = int(chamber_id.split(',')[1])\n",
    "        img_array[y-1,x-1] = 1 \n",
    "    \n",
    "    #generate title\n",
    "    if title is None:\n",
    "        title = ''\n",
    "\n",
    "    #Create the figure\n",
    "    layout = go.Layout()\n",
    "    fig = go.Figure(layout=layout, data=go.Heatmap(z=img_array, colorscale='Viridis'))\n",
    "    #center title in fig\n",
    "    fig.update_layout(title=title,\n",
    "                        title_x=0.5, \n",
    "                        yaxis=dict(scaleanchor=\"x\", scaleratio=1, autorange='reversed'), \n",
    "                        xaxis=dict(scaleratio=1),\n",
    "                        plot_bgcolor='rgba(0,0,0,0)',\n",
    "                        width=600, height=600,\n",
    "                        hovermode='x')\n",
    "    fig.update_xaxes(showticklabels=False)\n",
    "    fig.update_yaxes(showticklabels=False)\n",
    "\n",
    "    #create dash app:\n",
    "    app = Dash(__name__)\n",
    "    app.layout = html.Div([\n",
    "        dcc.Graph(id=\"graph\", figure=fig, clear_on_unhover=True),\n",
    "        dcc.Tooltip(id=\"graph-tooltip\"),\n",
    "    ])\n",
    "\n",
    "    app.run_server()\n",
    "\n",
    "plot_chip_by_idx(chamber_idxs_to_keep, title='Chambers to keep for r2 of {}'.format(r2_threshold))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Join kinetics and standard curve data\n",
    "# std_columns_tokeep = ['x', 'y','std_concs', 'std_summed_intensity', 'std_stddev_intensity', 'std_slope', \n",
    "#                       'std_intercept', 'std_r2', 'std_pvalue', 'std_stderr', 'std_slope_units', 'substrate_name', 'experimental_day', 'setup', 'device', 'assay_type', 'position']\n",
    "# kinetics_and_standards = squeezed_kinetic_data.join(fitted_standards_filtered[std_columns_tokeep])\n",
    "\n",
    "# # keep only data from chambers with standard curves with linear regression r2 values >0.98 (i.e., we don't process any data from chambers with poor standard curves)\n",
    "# kinetics_and_standards = kinetics_and_standards[kinetics_and_standards['std_r2'] > 0.98]\n",
    "# kinetics_and_standards.sort_values(by = 'std_slope').head(2)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For this next part, we'll be generating the predicted product concentration for each well, for each time, and (I think) for each concentration. That's a 3 or 4D array.\n",
    "I think we just do this by (something like) dividing by the slope for that chamber and adding the intercept - but it looks like below, they weren't adding the intercept.\n",
    "I'm too tired and will have to think about this next part tomorrow.\n",
    "\n",
    "Morning: Our standard curve relates product fluorescence to product concentration. (slope = RFU/conc) (intercept = RFU)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "chamber_idxs, luminance_data, conc_data, time_data = assay_data_to_array(db_conn, 'kinetics_0')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20, 1792, 11)"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "luminance_data.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each well contains time_series data. We have 32*56 wells (1792), and then N assays with different concentrations.\n",
    "So we have a np array with dimensions (time x wells x assays)\n",
    "For a standard, this will be (1, 1792, # concentrations)\n",
    "For a kinetics run, this will be (~20, 1792, # concentrations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "chamber_idxs, luminance_data, conc_data, time_data = assay_data_to_array(db_conn, 'kinetics_0')\n",
    "\n",
    "slopes = np.array([db_conn._json_dict['runs']['standard_0']['analyses']['linear_regression']['chambers'][chamber_idx]['slope'] for chamber_idx in chamber_idxs])\n",
    "\n",
    "#calculate product concentration by dividing every chamber intensity by the slope of the standard curve for that chamber\n",
    "product_concentration = luminance_data / slopes[np.newaxis, :, np.newaxis]\n",
    "\n",
    "#mask chambers that aren't in chamber_idxs_to_keep, by overwriting with NaNs:\n",
    "#product_concentration = deepcopy(product_concentration)\n",
    "for i, chamber_idx in enumerate(chamber_idxs):\n",
    "    if chamber_idx not in chamber_idxs_to_keep:\n",
    "        product_concentration[:,i,:] = np.nan\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #find idxs we are not keeping:\n",
    "# idxs_to_drop = []\n",
    "# for i, chamber_idx in enumerate(chamber_idxs):\n",
    "#     if chamber_idx not in chamber_idxs_to_keep:\n",
    "#         idxs_to_drop.append(i)\n",
    "# print(idxs_to_drop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# #NF where are we generating the CSVs? We need to specify filenames whenever we generate files.\n",
    "\n",
    "# # clean dataframe and merge button quant data\n",
    "# cleaned_kinetics_df = kinetics_and_standards.drop(\n",
    "#     columns=[\n",
    "#         \"x\",\n",
    "#         \"y\",\n",
    "#         \"chamber_median\",\n",
    "#         \"chamber_stddev\",\n",
    "#         \"std_concs\",\n",
    "#         \"std_summed_intensity\",\n",
    "#         \"std_stddev_intensity\",\n",
    "#         \"std_pvalue\",\n",
    "#         \"std_stderr\",\n",
    "#         \"std_slope_units\",\n",
    "#         \"experimental_day\",\n",
    "#         \"setup\",\n",
    "#         \"device\",\n",
    "#         'button_sum',\n",
    "#         'button_sum_bgsub'\n",
    "\n",
    "#     ]\n",
    "# )\n",
    "# df_button = pd.read_csv(root+'/20230712_adk_orthologs_egfp_summary.csv')\n",
    "\n",
    "# # add a position value that is a tuple of ints that can be sorted on numerically, unlike the \"indicies\" column, which is a string\n",
    "# df_button['position'] = [(int(r['y']),int(r['x'])) for i,r in df_button.iterrows()]\n",
    "\n",
    "# # merge kinetics and button quant data\n",
    "# cleaned_kinetics_df = pd.merge(cleaned_kinetics_df, df_button, on='position', how='left')\n",
    "\n",
    "# # clean merged dataframe of unnecessary columns etc\n",
    "# cleaned_kinetics_df = cleaned_kinetics_df.drop(\n",
    "#     columns=[\n",
    "#         \"x\",\n",
    "#         \"y\",\n",
    "#         \"median_button\",\n",
    "#         \"std_button\",\n",
    "#         \"x_button_center\",\n",
    "#         \"y_button_center\",\n",
    "#         \"radius_button_disk\",\n",
    "#         \"median_button_annulus\",\n",
    "#         \"std_button_annulus_localBG\",\n",
    "#         \"outer_radius_button_annulus\",\n",
    "#         'summed_button_annulus_normed',\n",
    "#         'inner_radius_button_annulus',\n",
    "#         \"xslice\",\n",
    "#         \"yslice\",\n",
    "#         \"id_y\"\n",
    "\n",
    "#     ]\n",
    "# )\n",
    "# cleaned_kinetics_df = cleaned_kinetics_df.rename(columns={'id_x': 'id'})\n",
    "# cleaned_kinetics_df.sort_values(by = 'std_r2').head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make numpy array of all button_quants with[ subtracted backgrounds:\n",
    "button_quant_no_background = [] #we will soon turn this into a numpy array\n",
    "for chamber_idx in chamber_idxs:\n",
    "    next_button_quant = db_conn._json_dict['button_quant'][chamber_idx]['summed_button_BGsub_Button_Quant']\n",
    "    button_quant_no_background.append(next_button_quant)\n",
    "button_quant_no_background = np.array(button_quant_no_background)\n",
    "\n",
    "# use eGFP standard curve to convert between button quant and eGFP concentration\n",
    "enzyme_concentration = button_quant_no_background / EGFP_SLOPE\n",
    "#is this low?\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### Fit initial rates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# class KineticDataException(Exception):\n",
    "#     pass\n",
    "\n",
    "# def divide_chunks(l, n):\n",
    "#     # Function to split a list 'l' into 'n' roughly equal-sized chunks.\n",
    "#     # The function yields each chunk as a separate list.\n",
    "#     # It ensures that no chunk is larger than the original list.\n",
    "#     n = math.ceil(len(l) / float(n))\n",
    "#     for i in range(0, len(l), n):\n",
    "#         yield l[i:i + n]\n",
    "\n",
    "# def mm_model(x, v_max, Km):\n",
    "#     # Michaelis-Menten equation to model enzyme kinetics.\n",
    "#     # 'x': substrate concentration, 'v_max': maximum reaction rate, 'Km': Michaelis constant.\n",
    "#     return v_max * x / (x + Km)\n",
    "\n",
    "\n",
    "# def compute_initial_reaction_slope(time_arr: np.array, \n",
    "#                                    signal_arr: np.array,\n",
    "#                                   min_included_percent: int = 2): \n",
    "#     \"\"\"\n",
    "#     Determines best linear fit to initial slope of data, by fitting regressions to \n",
    "#     all percentiles of the data--greater than some defined minimum--anchored at the origin.\n",
    "    \n",
    "#     Args:\n",
    "#         time_arr (np.array): array of assay read times\n",
    "#         signal_arr (np.array): array of kinetic signal readouts\n",
    "#         min_included_percent (int) = 5: minimum percent of data to be included \n",
    "    \n",
    "#     Returns:\n",
    "#         slope, intercept, score ((float, float, float)): fit parameters of best fit\n",
    "        \n",
    "#     \"\"\"\n",
    "#     # Need to triage further for different definitions of minimum\n",
    "#     MIN_INCLUDED_DATA_POINTS = 3\n",
    "    \n",
    "#     perc_concs = list(divide_chunks(signal_arr, 100))\n",
    "#     perc_times = list(divide_chunks(time_arr, 100))\n",
    " \n",
    "#     scores = []\n",
    "#     slopes = [] \n",
    "#     intercepts = []\n",
    "    \n",
    "#     min_inclusion = max(len(perc_times) * min_included_percent // 100, MIN_INCLUDED_DATA_POINTS)\n",
    "    \n",
    "#     for i in range(len(perc_times), min_inclusion, -1):\n",
    "#         curr_times = np.concatenate(perc_times[:i])\n",
    "#         curr_concs = np.concatenate(perc_concs[:i])\n",
    "#         reg = LinearRegression().fit(np.array(curr_times).reshape(-1,1), curr_concs)\n",
    "#         curr_score = reg.score(np.array(curr_times).reshape(-1,1), curr_concs)\n",
    "#         scores.append(curr_score)\n",
    "#         slopes.append(reg.coef_)\n",
    "#         intercepts.append(reg.intercept_)\n",
    "    \n",
    "#     if len(scores) == 0 and min_included_percent == 100:\n",
    "#         reg = LinearRegression().fit(np.array(perc_times).reshape(-1,1), perc_concs)\n",
    "#         curr_score = reg.score(np.array(perc_times).reshape(-1,1), perc_concs)\n",
    "#         return reg.coef_.item(), reg.intercept_.item(), curr_score\n",
    "\n",
    "#     # The fit with the highest R-squared value is selected as the best fit.\n",
    "#     max_r2_idx = np.argmax(scores)\n",
    "    \n",
    "#     if scores[max_r2_idx] < 0.9:\n",
    "#         return np.array([np.nan]), np.array([np.nan]), np.array([np.nan])\n",
    "    \n",
    "#     return slopes[max_r2_idx], intercepts[max_r2_idx], scores[max_r2_idx]\n",
    "\n",
    "\n",
    "# def get_scoop_filter_index(times: list, data: list):\n",
    "#     # Find the index to filter data to remove the initial scooping artifact in kinetic measurements.\n",
    "#     # The scooping artifact is the region where the curve decreases at the beginning of the reaction.\n",
    "#     rate_of_change = np.array(data[1:]) - np.array(data[:-1])\n",
    "#     try:\n",
    "#         last_negative_index = np.where(rate_of_change < 0)[0][-1] + 1\n",
    "#     except IndexError:\n",
    "#         last_negative_index = 0\n",
    "#     duration = max(times) - min(times)\n",
    "#     max_time_cutoff = np.where(np.array(times) < (duration / 3))[0][-1]\n",
    "\n",
    "#     return min(last_negative_index, max_time_cutoff)\n",
    "\n",
    "   \n",
    "# def fit_kinetics_linear(row):\n",
    "#     # Fit linear models to the initial slope of the kinetic data for each row (reaction) in the dataset.\n",
    "#     # The function uses 'compute_initial_reaction_slope' to compute the initial slope fit,\n",
    "#     # and 'get_scoop_filter_index' to determine the index to filter out the scooping artifact.\n",
    "#     # It returns slope, intercept, and R-squared values for both the entire reaction and the filtered part.\n",
    "#     try:\n",
    "#         slope, intercept, score = compute_initial_reaction_slope(row['time_s'],row['chamber_product_concs'])\n",
    "#     except ValueError:\n",
    "        \n",
    "#         slope = np.array([np.nan])\n",
    "#         intercept = np.array([np.nan])\n",
    "#         score = np.array([np.nan])\n",
    "            \n",
    "#     filter_scoop_index = get_scoop_filter_index(row['time_s'],row['chamber_product_concs'])\n",
    "#     try:\n",
    "#         filter_slope, filter_intercept, filter_score = compute_initial_reaction_slope(row['time_s'][filter_scoop_index:],row['chamber_product_concs'][filter_scoop_index:])\n",
    "#     except ValueError:\n",
    "        \n",
    "#         filter_slope = np.array([np.nan])\n",
    "#         filter_intercept = np.array([np.nan])\n",
    "#         filter_score = np.array([np.nan])\n",
    "#     return slope.item(), intercept.item(), score.item()**2, filter_slope.item(), filter_intercept.item(), filter_score.item()**2\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([-0.02997331,  0.01114055,  0.02856477,  0.03158389,  0.09724754,\n",
       "         0.14304928,  0.15472734,  0.16410932,  0.14304091,  0.17576695,\n",
       "         0.21392022]),\n",
       " array([0.96344253, 0.91953421, 0.98671611, 0.99617796, 0.99884375,\n",
       "        0.99848552, 0.99947381, 0.9986284 , 0.99871565, 0.99954329,\n",
       "        0.99928136]))"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"utf-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       "  \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<svg xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"616.0056pt\" height=\"617.2227pt\" viewBox=\"0 0 616.0056 617.2227\" xmlns=\"http://www.w3.org/2000/svg\" version=\"1.1\">\n",
       " <metadata>\n",
       "  <rdf:RDF xmlns:dc=\"http://purl.org/dc/elements/1.1/\" xmlns:cc=\"http://creativecommons.org/ns#\" xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\">\n",
       "   <cc:Work>\n",
       "    <dc:type rdf:resource=\"http://purl.org/dc/dcmitype/StillImage\"/>\n",
       "    <dc:date>2023-11-16T16:04:15.667887</dc:date>\n",
       "    <dc:format>image/svg+xml</dc:format>\n",
       "    <dc:creator>\n",
       "     <cc:Agent>\n",
       "      <dc:title>Matplotlib v3.8.1, https://matplotlib.org/</dc:title>\n",
       "     </cc:Agent>\n",
       "    </dc:creator>\n",
       "   </cc:Work>\n",
       "  </rdf:RDF>\n",
       " </metadata>\n",
       " <defs>\n",
       "  <style type=\"text/css\">*{stroke-linejoin: round; stroke-linecap: butt}</style>\n",
       " </defs>\n",
       " <g id=\"figure_1\">\n",
       "  <g id=\"patch_1\">\n",
       "   <path d=\"M 0 617.2227 \n",
       "L 616.0056 617.2227 \n",
       "L 616.0056 0 \n",
       "L 0 0 \n",
       "z\n",
       "\" style=\"fill: #ffffff\"/>\n",
       "  </g>\n",
       "  <g id=\"axes_1\">\n",
       "   <g id=\"patch_2\">\n",
       "    <path d=\"M 50.8056 576.3534 \n",
       "L 608.8056 576.3534 \n",
       "L 608.8056 21.9534 \n",
       "L 50.8056 21.9534 \n",
       "z\n",
       "\" style=\"fill: #ffffff\"/>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_1\">\n",
       "    <g id=\"xtick_1\">\n",
       "     <g id=\"line2d_1\">\n",
       "      <defs>\n",
       "       <path id=\"m2ee08342b4\" d=\"M 0 0 \n",
       "L 0 4.8 \n",
       "\" style=\"stroke: #262626\"/>\n",
       "      </defs>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ee08342b4\" x=\"76.169236\" y=\"576.3534\" style=\"fill: #262626; stroke: #262626\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_1\">\n",
       "      <!-- 0 -->\n",
       "      <g style=\"fill: #262626\" transform=\"translate(72.809836 592.67735) scale(0.1056 -0.1056)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-30\" d=\"M 2034 4250 \n",
       "Q 1547 4250 1301 3770 \n",
       "Q 1056 3291 1056 2328 \n",
       "Q 1056 1369 1301 889 \n",
       "Q 1547 409 2034 409 \n",
       "Q 2525 409 2770 889 \n",
       "Q 3016 1369 3016 2328 \n",
       "Q 3016 3291 2770 3770 \n",
       "Q 2525 4250 2034 4250 \n",
       "z\n",
       "M 2034 4750 \n",
       "Q 2819 4750 3233 4129 \n",
       "Q 3647 3509 3647 2328 \n",
       "Q 3647 1150 3233 529 \n",
       "Q 2819 -91 2034 -91 \n",
       "Q 1250 -91 836 529 \n",
       "Q 422 1150 422 2328 \n",
       "Q 422 3509 836 4129 \n",
       "Q 1250 4750 2034 4750 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_2\">\n",
       "     <g id=\"line2d_2\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ee08342b4\" x=\"144.461203\" y=\"576.3534\" style=\"fill: #262626; stroke: #262626\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_2\">\n",
       "      <!-- 250 -->\n",
       "      <g style=\"fill: #262626\" transform=\"translate(134.383003 592.67735) scale(0.1056 -0.1056)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-32\" d=\"M 1228 531 \n",
       "L 3431 531 \n",
       "L 3431 0 \n",
       "L 469 0 \n",
       "L 469 531 \n",
       "Q 828 903 1448 1529 \n",
       "Q 2069 2156 2228 2338 \n",
       "Q 2531 2678 2651 2914 \n",
       "Q 2772 3150 2772 3378 \n",
       "Q 2772 3750 2511 3984 \n",
       "Q 2250 4219 1831 4219 \n",
       "Q 1534 4219 1204 4116 \n",
       "Q 875 4013 500 3803 \n",
       "L 500 4441 \n",
       "Q 881 4594 1212 4672 \n",
       "Q 1544 4750 1819 4750 \n",
       "Q 2544 4750 2975 4387 \n",
       "Q 3406 4025 3406 3419 \n",
       "Q 3406 3131 3298 2873 \n",
       "Q 3191 2616 2906 2266 \n",
       "Q 2828 2175 2409 1742 \n",
       "Q 1991 1309 1228 531 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"DejaVuSans-35\" d=\"M 691 4666 \n",
       "L 3169 4666 \n",
       "L 3169 4134 \n",
       "L 1269 4134 \n",
       "L 1269 2991 \n",
       "Q 1406 3038 1543 3061 \n",
       "Q 1681 3084 1819 3084 \n",
       "Q 2600 3084 3056 2656 \n",
       "Q 3513 2228 3513 1497 \n",
       "Q 3513 744 3044 326 \n",
       "Q 2575 -91 1722 -91 \n",
       "Q 1428 -91 1123 -41 \n",
       "Q 819 9 494 109 \n",
       "L 494 744 \n",
       "Q 775 591 1075 516 \n",
       "Q 1375 441 1709 441 \n",
       "Q 2250 441 2565 725 \n",
       "Q 2881 1009 2881 1497 \n",
       "Q 2881 1984 2565 2268 \n",
       "Q 2250 2553 1709 2553 \n",
       "Q 1456 2553 1204 2497 \n",
       "Q 953 2441 691 2322 \n",
       "L 691 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-32\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-35\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_3\">\n",
       "     <g id=\"line2d_3\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ee08342b4\" x=\"212.753169\" y=\"576.3534\" style=\"fill: #262626; stroke: #262626\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_3\">\n",
       "      <!-- 500 -->\n",
       "      <g style=\"fill: #262626\" transform=\"translate(202.674969 592.67735) scale(0.1056 -0.1056)\">\n",
       "       <use xlink:href=\"#DejaVuSans-35\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_4\">\n",
       "     <g id=\"line2d_4\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ee08342b4\" x=\"281.045136\" y=\"576.3534\" style=\"fill: #262626; stroke: #262626\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_4\">\n",
       "      <!-- 750 -->\n",
       "      <g style=\"fill: #262626\" transform=\"translate(270.966936 592.67735) scale(0.1056 -0.1056)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-37\" d=\"M 525 4666 \n",
       "L 3525 4666 \n",
       "L 3525 4397 \n",
       "L 1831 0 \n",
       "L 1172 0 \n",
       "L 2766 4134 \n",
       "L 525 4134 \n",
       "L 525 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-37\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-35\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_5\">\n",
       "     <g id=\"line2d_5\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ee08342b4\" x=\"349.337102\" y=\"576.3534\" style=\"fill: #262626; stroke: #262626\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_5\">\n",
       "      <!-- 1000 -->\n",
       "      <g style=\"fill: #262626\" transform=\"translate(335.899502 592.67735) scale(0.1056 -0.1056)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-31\" d=\"M 794 531 \n",
       "L 1825 531 \n",
       "L 1825 4091 \n",
       "L 703 3866 \n",
       "L 703 4441 \n",
       "L 1819 4666 \n",
       "L 2450 4666 \n",
       "L 2450 531 \n",
       "L 3481 531 \n",
       "L 3481 0 \n",
       "L 794 0 \n",
       "L 794 531 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"190.869141\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_6\">\n",
       "     <g id=\"line2d_6\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ee08342b4\" x=\"417.629069\" y=\"576.3534\" style=\"fill: #262626; stroke: #262626\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_6\">\n",
       "      <!-- 1250 -->\n",
       "      <g style=\"fill: #262626\" transform=\"translate(404.191469 592.67735) scale(0.1056 -0.1056)\">\n",
       "       <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-32\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-35\" x=\"127.246094\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"190.869141\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_7\">\n",
       "     <g id=\"line2d_7\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ee08342b4\" x=\"485.921035\" y=\"576.3534\" style=\"fill: #262626; stroke: #262626\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_7\">\n",
       "      <!-- 1500 -->\n",
       "      <g style=\"fill: #262626\" transform=\"translate(472.483435 592.67735) scale(0.1056 -0.1056)\">\n",
       "       <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-35\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"190.869141\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_8\">\n",
       "     <g id=\"line2d_8\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ee08342b4\" x=\"554.213002\" y=\"576.3534\" style=\"fill: #262626; stroke: #262626\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_8\">\n",
       "      <!-- 1750 -->\n",
       "      <g style=\"fill: #262626\" transform=\"translate(540.775402 592.67735) scale(0.1056 -0.1056)\">\n",
       "       <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-37\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-35\" x=\"127.246094\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"190.869141\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"text_9\">\n",
       "     <!-- Time (seconds) -->\n",
       "     <g style=\"fill: #262626\" transform=\"translate(285.8433 607.6269) scale(0.1152 -0.1152)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-54\" d=\"M -19 4666 \n",
       "L 3928 4666 \n",
       "L 3928 4134 \n",
       "L 2272 4134 \n",
       "L 2272 0 \n",
       "L 1638 0 \n",
       "L 1638 4134 \n",
       "L -19 4134 \n",
       "L -19 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-69\" d=\"M 603 3500 \n",
       "L 1178 3500 \n",
       "L 1178 0 \n",
       "L 603 0 \n",
       "L 603 3500 \n",
       "z\n",
       "M 603 4863 \n",
       "L 1178 4863 \n",
       "L 1178 4134 \n",
       "L 603 4134 \n",
       "L 603 4863 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-6d\" d=\"M 3328 2828 \n",
       "Q 3544 3216 3844 3400 \n",
       "Q 4144 3584 4550 3584 \n",
       "Q 5097 3584 5394 3201 \n",
       "Q 5691 2819 5691 2113 \n",
       "L 5691 0 \n",
       "L 5113 0 \n",
       "L 5113 2094 \n",
       "Q 5113 2597 4934 2840 \n",
       "Q 4756 3084 4391 3084 \n",
       "Q 3944 3084 3684 2787 \n",
       "Q 3425 2491 3425 1978 \n",
       "L 3425 0 \n",
       "L 2847 0 \n",
       "L 2847 2094 \n",
       "Q 2847 2600 2669 2842 \n",
       "Q 2491 3084 2119 3084 \n",
       "Q 1678 3084 1418 2786 \n",
       "Q 1159 2488 1159 1978 \n",
       "L 1159 0 \n",
       "L 581 0 \n",
       "L 581 3500 \n",
       "L 1159 3500 \n",
       "L 1159 2956 \n",
       "Q 1356 3278 1631 3431 \n",
       "Q 1906 3584 2284 3584 \n",
       "Q 2666 3584 2933 3390 \n",
       "Q 3200 3197 3328 2828 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-65\" d=\"M 3597 1894 \n",
       "L 3597 1613 \n",
       "L 953 1613 \n",
       "Q 991 1019 1311 708 \n",
       "Q 1631 397 2203 397 \n",
       "Q 2534 397 2845 478 \n",
       "Q 3156 559 3463 722 \n",
       "L 3463 178 \n",
       "Q 3153 47 2828 -22 \n",
       "Q 2503 -91 2169 -91 \n",
       "Q 1331 -91 842 396 \n",
       "Q 353 884 353 1716 \n",
       "Q 353 2575 817 3079 \n",
       "Q 1281 3584 2069 3584 \n",
       "Q 2775 3584 3186 3129 \n",
       "Q 3597 2675 3597 1894 \n",
       "z\n",
       "M 3022 2063 \n",
       "Q 3016 2534 2758 2815 \n",
       "Q 2500 3097 2075 3097 \n",
       "Q 1594 3097 1305 2825 \n",
       "Q 1016 2553 972 2059 \n",
       "L 3022 2063 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-20\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-28\" d=\"M 1984 4856 \n",
       "Q 1566 4138 1362 3434 \n",
       "Q 1159 2731 1159 2009 \n",
       "Q 1159 1288 1364 580 \n",
       "Q 1569 -128 1984 -844 \n",
       "L 1484 -844 \n",
       "Q 1016 -109 783 600 \n",
       "Q 550 1309 550 2009 \n",
       "Q 550 2706 781 3412 \n",
       "Q 1013 4119 1484 4856 \n",
       "L 1984 4856 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-73\" d=\"M 2834 3397 \n",
       "L 2834 2853 \n",
       "Q 2591 2978 2328 3040 \n",
       "Q 2066 3103 1784 3103 \n",
       "Q 1356 3103 1142 2972 \n",
       "Q 928 2841 928 2578 \n",
       "Q 928 2378 1081 2264 \n",
       "Q 1234 2150 1697 2047 \n",
       "L 1894 2003 \n",
       "Q 2506 1872 2764 1633 \n",
       "Q 3022 1394 3022 966 \n",
       "Q 3022 478 2636 193 \n",
       "Q 2250 -91 1575 -91 \n",
       "Q 1294 -91 989 -36 \n",
       "Q 684 19 347 128 \n",
       "L 347 722 \n",
       "Q 666 556 975 473 \n",
       "Q 1284 391 1588 391 \n",
       "Q 1994 391 2212 530 \n",
       "Q 2431 669 2431 922 \n",
       "Q 2431 1156 2273 1281 \n",
       "Q 2116 1406 1581 1522 \n",
       "L 1381 1569 \n",
       "Q 847 1681 609 1914 \n",
       "Q 372 2147 372 2553 \n",
       "Q 372 3047 722 3315 \n",
       "Q 1072 3584 1716 3584 \n",
       "Q 2034 3584 2315 3537 \n",
       "Q 2597 3491 2834 3397 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-63\" d=\"M 3122 3366 \n",
       "L 3122 2828 \n",
       "Q 2878 2963 2633 3030 \n",
       "Q 2388 3097 2138 3097 \n",
       "Q 1578 3097 1268 2742 \n",
       "Q 959 2388 959 1747 \n",
       "Q 959 1106 1268 751 \n",
       "Q 1578 397 2138 397 \n",
       "Q 2388 397 2633 464 \n",
       "Q 2878 531 3122 666 \n",
       "L 3122 134 \n",
       "Q 2881 22 2623 -34 \n",
       "Q 2366 -91 2075 -91 \n",
       "Q 1284 -91 818 406 \n",
       "Q 353 903 353 1747 \n",
       "Q 353 2603 823 3093 \n",
       "Q 1294 3584 2113 3584 \n",
       "Q 2378 3584 2631 3529 \n",
       "Q 2884 3475 3122 3366 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-6f\" d=\"M 1959 3097 \n",
       "Q 1497 3097 1228 2736 \n",
       "Q 959 2375 959 1747 \n",
       "Q 959 1119 1226 758 \n",
       "Q 1494 397 1959 397 \n",
       "Q 2419 397 2687 759 \n",
       "Q 2956 1122 2956 1747 \n",
       "Q 2956 2369 2687 2733 \n",
       "Q 2419 3097 1959 3097 \n",
       "z\n",
       "M 1959 3584 \n",
       "Q 2709 3584 3137 3096 \n",
       "Q 3566 2609 3566 1747 \n",
       "Q 3566 888 3137 398 \n",
       "Q 2709 -91 1959 -91 \n",
       "Q 1206 -91 779 398 \n",
       "Q 353 888 353 1747 \n",
       "Q 353 2609 779 3096 \n",
       "Q 1206 3584 1959 3584 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-6e\" d=\"M 3513 2113 \n",
       "L 3513 0 \n",
       "L 2938 0 \n",
       "L 2938 2094 \n",
       "Q 2938 2591 2744 2837 \n",
       "Q 2550 3084 2163 3084 \n",
       "Q 1697 3084 1428 2787 \n",
       "Q 1159 2491 1159 1978 \n",
       "L 1159 0 \n",
       "L 581 0 \n",
       "L 581 3500 \n",
       "L 1159 3500 \n",
       "L 1159 2956 \n",
       "Q 1366 3272 1645 3428 \n",
       "Q 1925 3584 2291 3584 \n",
       "Q 2894 3584 3203 3211 \n",
       "Q 3513 2838 3513 2113 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-64\" d=\"M 2906 2969 \n",
       "L 2906 4863 \n",
       "L 3481 4863 \n",
       "L 3481 0 \n",
       "L 2906 0 \n",
       "L 2906 525 \n",
       "Q 2725 213 2448 61 \n",
       "Q 2172 -91 1784 -91 \n",
       "Q 1150 -91 751 415 \n",
       "Q 353 922 353 1747 \n",
       "Q 353 2572 751 3078 \n",
       "Q 1150 3584 1784 3584 \n",
       "Q 2172 3584 2448 3432 \n",
       "Q 2725 3281 2906 2969 \n",
       "z\n",
       "M 947 1747 \n",
       "Q 947 1113 1208 752 \n",
       "Q 1469 391 1925 391 \n",
       "Q 2381 391 2643 752 \n",
       "Q 2906 1113 2906 1747 \n",
       "Q 2906 2381 2643 2742 \n",
       "Q 2381 3103 1925 3103 \n",
       "Q 1469 3103 1208 2742 \n",
       "Q 947 2381 947 1747 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-29\" d=\"M 513 4856 \n",
       "L 1013 4856 \n",
       "Q 1481 4119 1714 3412 \n",
       "Q 1947 2706 1947 2009 \n",
       "Q 1947 1309 1714 600 \n",
       "Q 1481 -109 1013 -844 \n",
       "L 513 -844 \n",
       "Q 928 -128 1133 580 \n",
       "Q 1338 1288 1338 2009 \n",
       "Q 1338 2731 1133 3434 \n",
       "Q 928 4138 513 4856 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-54\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-69\" x=\"57.958984\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6d\" x=\"85.742188\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-65\" x=\"183.154297\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"244.677734\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-28\" x=\"276.464844\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-73\" x=\"315.478516\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-65\" x=\"367.578125\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-63\" x=\"429.101562\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6f\" x=\"484.082031\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6e\" x=\"545.263672\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-64\" x=\"608.642578\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-73\" x=\"672.119141\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-29\" x=\"724.21875\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_2\">\n",
       "    <g id=\"ytick_1\">\n",
       "     <g id=\"line2d_9\">\n",
       "      <defs>\n",
       "       <path id=\"m6c1675f894\" d=\"M 0 0 \n",
       "L -4.8 0 \n",
       "\" style=\"stroke: #262626\"/>\n",
       "      </defs>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m6c1675f894\" x=\"50.8056\" y=\"521.00105\" style=\"fill: #262626; stroke: #262626\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_10\">\n",
       "      <!-- 0 -->\n",
       "      <g style=\"fill: #262626\" transform=\"translate(35.7868 525.013025) scale(0.1056 -0.1056)\">\n",
       "       <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_2\">\n",
       "     <g id=\"line2d_10\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m6c1675f894\" x=\"50.8056\" y=\"439.179307\" style=\"fill: #262626; stroke: #262626\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_11\">\n",
       "      <!-- 100 -->\n",
       "      <g style=\"fill: #262626\" transform=\"translate(22.3492 443.191282) scale(0.1056 -0.1056)\">\n",
       "       <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_3\">\n",
       "     <g id=\"line2d_11\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m6c1675f894\" x=\"50.8056\" y=\"357.357563\" style=\"fill: #262626; stroke: #262626\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_12\">\n",
       "      <!-- 200 -->\n",
       "      <g style=\"fill: #262626\" transform=\"translate(22.3492 361.369538) scale(0.1056 -0.1056)\">\n",
       "       <use xlink:href=\"#DejaVuSans-32\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_4\">\n",
       "     <g id=\"line2d_12\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m6c1675f894\" x=\"50.8056\" y=\"275.535819\" style=\"fill: #262626; stroke: #262626\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_13\">\n",
       "      <!-- 300 -->\n",
       "      <g style=\"fill: #262626\" transform=\"translate(22.3492 279.547794) scale(0.1056 -0.1056)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-33\" d=\"M 2597 2516 \n",
       "Q 3050 2419 3304 2112 \n",
       "Q 3559 1806 3559 1356 \n",
       "Q 3559 666 3084 287 \n",
       "Q 2609 -91 1734 -91 \n",
       "Q 1441 -91 1130 -33 \n",
       "Q 819 25 488 141 \n",
       "L 488 750 \n",
       "Q 750 597 1062 519 \n",
       "Q 1375 441 1716 441 \n",
       "Q 2309 441 2620 675 \n",
       "Q 2931 909 2931 1356 \n",
       "Q 2931 1769 2642 2001 \n",
       "Q 2353 2234 1838 2234 \n",
       "L 1294 2234 \n",
       "L 1294 2753 \n",
       "L 1863 2753 \n",
       "Q 2328 2753 2575 2939 \n",
       "Q 2822 3125 2822 3475 \n",
       "Q 2822 3834 2567 4026 \n",
       "Q 2313 4219 1838 4219 \n",
       "Q 1578 4219 1281 4162 \n",
       "Q 984 4106 628 3988 \n",
       "L 628 4550 \n",
       "Q 988 4650 1302 4700 \n",
       "Q 1616 4750 1894 4750 \n",
       "Q 2613 4750 3031 4423 \n",
       "Q 3450 4097 3450 3541 \n",
       "Q 3450 3153 3228 2886 \n",
       "Q 3006 2619 2597 2516 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-33\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_5\">\n",
       "     <g id=\"line2d_13\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m6c1675f894\" x=\"50.8056\" y=\"193.714076\" style=\"fill: #262626; stroke: #262626\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_14\">\n",
       "      <!-- 400 -->\n",
       "      <g style=\"fill: #262626\" transform=\"translate(22.3492 197.726051) scale(0.1056 -0.1056)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-34\" d=\"M 2419 4116 \n",
       "L 825 1625 \n",
       "L 2419 1625 \n",
       "L 2419 4116 \n",
       "z\n",
       "M 2253 4666 \n",
       "L 3047 4666 \n",
       "L 3047 1625 \n",
       "L 3713 1625 \n",
       "L 3713 1100 \n",
       "L 3047 1100 \n",
       "L 3047 0 \n",
       "L 2419 0 \n",
       "L 2419 1100 \n",
       "L 313 1100 \n",
       "L 313 1709 \n",
       "L 2253 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-34\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_6\">\n",
       "     <g id=\"line2d_14\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m6c1675f894\" x=\"50.8056\" y=\"111.892332\" style=\"fill: #262626; stroke: #262626\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_15\">\n",
       "      <!-- 500 -->\n",
       "      <g style=\"fill: #262626\" transform=\"translate(22.3492 115.904307) scale(0.1056 -0.1056)\">\n",
       "       <use xlink:href=\"#DejaVuSans-35\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_7\">\n",
       "     <g id=\"line2d_15\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m6c1675f894\" x=\"50.8056\" y=\"30.070588\" style=\"fill: #262626; stroke: #262626\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_16\">\n",
       "      <!-- 600 -->\n",
       "      <g style=\"fill: #262626\" transform=\"translate(22.3492 34.082563) scale(0.1056 -0.1056)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-36\" d=\"M 2113 2584 \n",
       "Q 1688 2584 1439 2293 \n",
       "Q 1191 2003 1191 1497 \n",
       "Q 1191 994 1439 701 \n",
       "Q 1688 409 2113 409 \n",
       "Q 2538 409 2786 701 \n",
       "Q 3034 994 3034 1497 \n",
       "Q 3034 2003 2786 2293 \n",
       "Q 2538 2584 2113 2584 \n",
       "z\n",
       "M 3366 4563 \n",
       "L 3366 3988 \n",
       "Q 3128 4100 2886 4159 \n",
       "Q 2644 4219 2406 4219 \n",
       "Q 1781 4219 1451 3797 \n",
       "Q 1122 3375 1075 2522 \n",
       "Q 1259 2794 1537 2939 \n",
       "Q 1816 3084 2150 3084 \n",
       "Q 2853 3084 3261 2657 \n",
       "Q 3669 2231 3669 1497 \n",
       "Q 3669 778 3244 343 \n",
       "Q 2819 -91 2113 -91 \n",
       "Q 1303 -91 875 529 \n",
       "Q 447 1150 447 2328 \n",
       "Q 447 3434 972 4092 \n",
       "Q 1497 4750 2381 4750 \n",
       "Q 2619 4750 2861 4703 \n",
       "Q 3103 4656 3366 4563 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-36\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"text_17\">\n",
       "     <!-- µM -->\n",
       "     <g style=\"fill: #262626\" transform=\"translate(15.9534 307.788) rotate(-90) scale(0.1152 -0.1152)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-b5\" d=\"M 544 -1331 \n",
       "L 544 3500 \n",
       "L 1119 3500 \n",
       "L 1119 1325 \n",
       "Q 1119 872 1334 640 \n",
       "Q 1550 409 1972 409 \n",
       "Q 2434 409 2667 671 \n",
       "Q 2900 934 2900 1459 \n",
       "L 2900 3500 \n",
       "L 3475 3500 \n",
       "L 3475 806 \n",
       "Q 3475 619 3529 530 \n",
       "Q 3584 441 3700 441 \n",
       "Q 3728 441 3778 458 \n",
       "Q 3828 475 3916 513 \n",
       "L 3916 50 \n",
       "Q 3788 -22 3673 -56 \n",
       "Q 3559 -91 3450 -91 \n",
       "Q 3234 -91 3106 31 \n",
       "Q 2978 153 2931 403 \n",
       "Q 2775 156 2548 32 \n",
       "Q 2322 -91 2016 -91 \n",
       "Q 1697 -91 1473 31 \n",
       "Q 1250 153 1119 397 \n",
       "L 1119 -1331 \n",
       "L 544 -1331 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-4d\" d=\"M 628 4666 \n",
       "L 1569 4666 \n",
       "L 2759 1491 \n",
       "L 3956 4666 \n",
       "L 4897 4666 \n",
       "L 4897 0 \n",
       "L 4281 0 \n",
       "L 4281 4097 \n",
       "L 3078 897 \n",
       "L 2444 897 \n",
       "L 1241 4097 \n",
       "L 1241 0 \n",
       "L 628 0 \n",
       "L 628 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-b5\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-4d\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"PathCollection_1\">\n",
       "    <defs>\n",
       "     <path id=\"mced943a113\" d=\"M 0 2.4 \n",
       "C 0.636487 2.4 1.246992 2.147121 1.697056 1.697056 \n",
       "C 2.147121 1.246992 2.4 0.636487 2.4 0 \n",
       "C 2.4 -0.636487 2.147121 -1.246992 1.697056 -1.697056 \n",
       "C 1.246992 -2.147121 0.636487 -2.4 0 -2.4 \n",
       "C -0.636487 -2.4 -1.246992 -2.147121 -1.697056 -1.697056 \n",
       "C -2.147121 -1.246992 -2.4 -0.636487 -2.4 0 \n",
       "C -2.4 0.636487 -2.147121 1.246992 -1.697056 1.697056 \n",
       "C -1.246992 2.147121 -0.636487 2.4 0 2.4 \n",
       "z\n",
       "\" style=\"stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "    </defs>\n",
       "    <g clip-path=\"url(#p823529a869)\">\n",
       "     <use xlink:href=\"#mced943a113\" x=\"76.169236\" y=\"505.707591\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"86.00328\" y=\"506.298115\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"95.564155\" y=\"507.451602\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"105.398198\" y=\"507.490228\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"115.232241\" y=\"507.023833\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"125.066284\" y=\"507.289383\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"134.62716\" y=\"507.555185\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"144.461203\" y=\"507.539823\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"154.295246\" y=\"507.586194\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"164.129289\" y=\"507.379659\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"173.690165\" y=\"508.096431\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"190.080237\" y=\"507.41503\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"206.470308\" y=\"506.443107\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"222.86038\" y=\"505.865321\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"239.250452\" y=\"506.236825\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"255.640524\" y=\"504.93193\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"337.590884\" y=\"499.117623\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"419.541244\" y=\"496.47496\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"501.491604\" y=\"494.88615\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"583.441964\" y=\"493.211806\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"PathCollection_2\">\n",
       "    <defs>\n",
       "     <path id=\"m5cccf7cbb5\" d=\"M 0 2.4 \n",
       "C 0.636487 2.4 1.246992 2.147121 1.697056 1.697056 \n",
       "C 2.147121 1.246992 2.4 0.636487 2.4 0 \n",
       "C 2.4 -0.636487 2.147121 -1.246992 1.697056 -1.697056 \n",
       "C 1.246992 -2.147121 0.636487 -2.4 0 -2.4 \n",
       "C -0.636487 -2.4 -1.246992 -2.147121 -1.697056 -1.697056 \n",
       "C -2.147121 -1.246992 -2.4 -0.636487 -2.4 0 \n",
       "C -2.4 0.636487 -2.147121 1.246992 -1.697056 1.697056 \n",
       "C -1.246992 2.147121 -0.636487 2.4 0 2.4 \n",
       "z\n",
       "\" style=\"stroke: #dd8452; stroke-width: 0.8\"/>\n",
       "    </defs>\n",
       "    <g clip-path=\"url(#p823529a869)\">\n",
       "     <use xlink:href=\"#m5cccf7cbb5\" x=\"76.169236\" y=\"498.882797\" style=\"fill: #dd8452; stroke: #dd8452; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m5cccf7cbb5\" x=\"86.00328\" y=\"500.929752\" style=\"fill: #dd8452; stroke: #dd8452; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m5cccf7cbb5\" x=\"95.564155\" y=\"501.272144\" style=\"fill: #dd8452; stroke: #dd8452; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m5cccf7cbb5\" x=\"105.398198\" y=\"501.090359\" style=\"fill: #dd8452; stroke: #dd8452; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m5cccf7cbb5\" x=\"115.232241\" y=\"500.445181\" style=\"fill: #dd8452; stroke: #dd8452; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m5cccf7cbb5\" x=\"125.066284\" y=\"500.655731\" style=\"fill: #dd8452; stroke: #dd8452; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m5cccf7cbb5\" x=\"134.62716\" y=\"500.126117\" style=\"fill: #dd8452; stroke: #dd8452; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m5cccf7cbb5\" x=\"144.461203\" y=\"499.496018\" style=\"fill: #dd8452; stroke: #dd8452; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m5cccf7cbb5\" x=\"154.295246\" y=\"499.93839\" style=\"fill: #dd8452; stroke: #dd8452; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m5cccf7cbb5\" x=\"164.129289\" y=\"499.812174\" style=\"fill: #dd8452; stroke: #dd8452; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m5cccf7cbb5\" x=\"173.690165\" y=\"500.846841\" style=\"fill: #dd8452; stroke: #dd8452; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m5cccf7cbb5\" x=\"190.080237\" y=\"498.530796\" style=\"fill: #dd8452; stroke: #dd8452; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m5cccf7cbb5\" x=\"206.470308\" y=\"499.452365\" style=\"fill: #dd8452; stroke: #dd8452; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m5cccf7cbb5\" x=\"222.86038\" y=\"498.366902\" style=\"fill: #dd8452; stroke: #dd8452; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m5cccf7cbb5\" x=\"239.250452\" y=\"498.257281\" style=\"fill: #dd8452; stroke: #dd8452; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m5cccf7cbb5\" x=\"255.640524\" y=\"497.032864\" style=\"fill: #dd8452; stroke: #dd8452; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m5cccf7cbb5\" x=\"337.590884\" y=\"493.521767\" style=\"fill: #dd8452; stroke: #dd8452; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m5cccf7cbb5\" x=\"419.541244\" y=\"487.361685\" style=\"fill: #dd8452; stroke: #dd8452; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m5cccf7cbb5\" x=\"501.491604\" y=\"486.857391\" style=\"fill: #dd8452; stroke: #dd8452; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m5cccf7cbb5\" x=\"583.441964\" y=\"486.283334\" style=\"fill: #dd8452; stroke: #dd8452; stroke-width: 0.8\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"PathCollection_3\">\n",
       "    <defs>\n",
       "     <path id=\"m42342a45fb\" d=\"M 0 2.4 \n",
       "C 0.636487 2.4 1.246992 2.147121 1.697056 1.697056 \n",
       "C 2.147121 1.246992 2.4 0.636487 2.4 0 \n",
       "C 2.4 -0.636487 2.147121 -1.246992 1.697056 -1.697056 \n",
       "C 1.246992 -2.147121 0.636487 -2.4 0 -2.4 \n",
       "C -0.636487 -2.4 -1.246992 -2.147121 -1.697056 -1.697056 \n",
       "C -2.147121 -1.246992 -2.4 -0.636487 -2.4 0 \n",
       "C -2.4 0.636487 -2.147121 1.246992 -1.697056 1.697056 \n",
       "C -1.246992 2.147121 -0.636487 2.4 0 2.4 \n",
       "z\n",
       "\" style=\"stroke: #55a868; stroke-width: 0.8\"/>\n",
       "    </defs>\n",
       "    <g clip-path=\"url(#p823529a869)\">\n",
       "     <use xlink:href=\"#m42342a45fb\" x=\"76.169236\" y=\"500.295006\" style=\"fill: #55a868; stroke: #55a868; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m42342a45fb\" x=\"86.00328\" y=\"501.505104\" style=\"fill: #55a868; stroke: #55a868; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m42342a45fb\" x=\"95.564155\" y=\"499.885286\" style=\"fill: #55a868; stroke: #55a868; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m42342a45fb\" x=\"105.398198\" y=\"497.425704\" style=\"fill: #55a868; stroke: #55a868; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m42342a45fb\" x=\"115.232241\" y=\"497.839059\" style=\"fill: #55a868; stroke: #55a868; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m42342a45fb\" x=\"125.066284\" y=\"496.720564\" style=\"fill: #55a868; stroke: #55a868; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m42342a45fb\" x=\"134.62716\" y=\"495.857504\" style=\"fill: #55a868; stroke: #55a868; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m42342a45fb\" x=\"144.461203\" y=\"495.388643\" style=\"fill: #55a868; stroke: #55a868; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m42342a45fb\" x=\"154.295246\" y=\"494.584187\" style=\"fill: #55a868; stroke: #55a868; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m42342a45fb\" x=\"164.129289\" y=\"494.681512\" style=\"fill: #55a868; stroke: #55a868; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m42342a45fb\" x=\"173.690165\" y=\"493.961705\" style=\"fill: #55a868; stroke: #55a868; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m42342a45fb\" x=\"190.080237\" y=\"492.631112\" style=\"fill: #55a868; stroke: #55a868; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m42342a45fb\" x=\"206.470308\" y=\"490.19922\" style=\"fill: #55a868; stroke: #55a868; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m42342a45fb\" x=\"222.86038\" y=\"487.427496\" style=\"fill: #55a868; stroke: #55a868; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m42342a45fb\" x=\"239.250452\" y=\"486.137426\" style=\"fill: #55a868; stroke: #55a868; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m42342a45fb\" x=\"255.640524\" y=\"484.504711\" style=\"fill: #55a868; stroke: #55a868; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m42342a45fb\" x=\"337.590884\" y=\"478.21127\" style=\"fill: #55a868; stroke: #55a868; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m42342a45fb\" x=\"419.541244\" y=\"472.831685\" style=\"fill: #55a868; stroke: #55a868; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m42342a45fb\" x=\"501.491604\" y=\"469.989125\" style=\"fill: #55a868; stroke: #55a868; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m42342a45fb\" x=\"583.441964\" y=\"467.380157\" style=\"fill: #55a868; stroke: #55a868; stroke-width: 0.8\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"PathCollection_4\">\n",
       "    <defs>\n",
       "     <path id=\"m77f1940be6\" d=\"M 0 2.4 \n",
       "C 0.636487 2.4 1.246992 2.147121 1.697056 1.697056 \n",
       "C 2.147121 1.246992 2.4 0.636487 2.4 0 \n",
       "C 2.4 -0.636487 2.147121 -1.246992 1.697056 -1.697056 \n",
       "C 1.246992 -2.147121 0.636487 -2.4 0 -2.4 \n",
       "C -0.636487 -2.4 -1.246992 -2.147121 -1.697056 -1.697056 \n",
       "C -2.147121 -1.246992 -2.4 -0.636487 -2.4 0 \n",
       "C -2.4 0.636487 -2.147121 1.246992 -1.697056 1.697056 \n",
       "C -1.246992 2.147121 -0.636487 2.4 0 2.4 \n",
       "z\n",
       "\" style=\"stroke: #c44e52; stroke-width: 0.8\"/>\n",
       "    </defs>\n",
       "    <g clip-path=\"url(#p823529a869)\">\n",
       "     <use xlink:href=\"#m77f1940be6\" x=\"76.169236\" y=\"499.090281\" style=\"fill: #c44e52; stroke: #c44e52; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m77f1940be6\" x=\"86.00328\" y=\"498.257913\" style=\"fill: #c44e52; stroke: #c44e52; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m77f1940be6\" x=\"95.564155\" y=\"497.25454\" style=\"fill: #c44e52; stroke: #c44e52; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m77f1940be6\" x=\"105.398198\" y=\"495.666394\" style=\"fill: #c44e52; stroke: #c44e52; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m77f1940be6\" x=\"115.232241\" y=\"493.500431\" style=\"fill: #c44e52; stroke: #c44e52; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m77f1940be6\" x=\"125.066284\" y=\"492.274306\" style=\"fill: #c44e52; stroke: #c44e52; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m77f1940be6\" x=\"134.62716\" y=\"490.96565\" style=\"fill: #c44e52; stroke: #c44e52; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m77f1940be6\" x=\"144.461203\" y=\"489.881673\" style=\"fill: #c44e52; stroke: #c44e52; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m77f1940be6\" x=\"154.295246\" y=\"487.515054\" style=\"fill: #c44e52; stroke: #c44e52; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m77f1940be6\" x=\"164.129289\" y=\"486.61795\" style=\"fill: #c44e52; stroke: #c44e52; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m77f1940be6\" x=\"173.690165\" y=\"484.685548\" style=\"fill: #c44e52; stroke: #c44e52; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m77f1940be6\" x=\"190.080237\" y=\"480.009807\" style=\"fill: #c44e52; stroke: #c44e52; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m77f1940be6\" x=\"206.470308\" y=\"476.530193\" style=\"fill: #c44e52; stroke: #c44e52; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m77f1940be6\" x=\"222.86038\" y=\"472.857984\" style=\"fill: #c44e52; stroke: #c44e52; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m77f1940be6\" x=\"239.250452\" y=\"468.150981\" style=\"fill: #c44e52; stroke: #c44e52; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m77f1940be6\" x=\"255.640524\" y=\"465.265622\" style=\"fill: #c44e52; stroke: #c44e52; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m77f1940be6\" x=\"337.590884\" y=\"451.80253\" style=\"fill: #c44e52; stroke: #c44e52; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m77f1940be6\" x=\"419.541244\" y=\"445.515695\" style=\"fill: #c44e52; stroke: #c44e52; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m77f1940be6\" x=\"501.491604\" y=\"441.865391\" style=\"fill: #c44e52; stroke: #c44e52; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m77f1940be6\" x=\"583.441964\" y=\"431.30242\" style=\"fill: #c44e52; stroke: #c44e52; stroke-width: 0.8\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"PathCollection_5\">\n",
       "    <defs>\n",
       "     <path id=\"mb5228dadd8\" d=\"M 0 2.4 \n",
       "C 0.636487 2.4 1.246992 2.147121 1.697056 1.697056 \n",
       "C 2.147121 1.246992 2.4 0.636487 2.4 0 \n",
       "C 2.4 -0.636487 2.147121 -1.246992 1.697056 -1.697056 \n",
       "C 1.246992 -2.147121 0.636487 -2.4 0 -2.4 \n",
       "C -0.636487 -2.4 -1.246992 -2.147121 -1.697056 -1.697056 \n",
       "C -2.147121 -1.246992 -2.4 -0.636487 -2.4 0 \n",
       "C -2.4 0.636487 -2.147121 1.246992 -1.697056 1.697056 \n",
       "C -1.246992 2.147121 -0.636487 2.4 0 2.4 \n",
       "z\n",
       "\" style=\"stroke: #8172b3; stroke-width: 0.8\"/>\n",
       "    </defs>\n",
       "    <g clip-path=\"url(#p823529a869)\">\n",
       "     <use xlink:href=\"#mb5228dadd8\" x=\"76.169236\" y=\"494.283804\" style=\"fill: #8172b3; stroke: #8172b3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb5228dadd8\" x=\"86.00328\" y=\"492.834297\" style=\"fill: #8172b3; stroke: #8172b3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb5228dadd8\" x=\"95.564155\" y=\"489.88436\" style=\"fill: #8172b3; stroke: #8172b3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb5228dadd8\" x=\"105.398198\" y=\"487.531712\" style=\"fill: #8172b3; stroke: #8172b3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb5228dadd8\" x=\"115.232241\" y=\"484.590783\" style=\"fill: #8172b3; stroke: #8172b3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb5228dadd8\" x=\"125.066284\" y=\"481.502651\" style=\"fill: #8172b3; stroke: #8172b3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb5228dadd8\" x=\"134.62716\" y=\"479.149054\" style=\"fill: #8172b3; stroke: #8172b3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb5228dadd8\" x=\"144.461203\" y=\"476.818596\" style=\"fill: #8172b3; stroke: #8172b3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb5228dadd8\" x=\"154.295246\" y=\"474.10693\" style=\"fill: #8172b3; stroke: #8172b3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb5228dadd8\" x=\"164.129289\" y=\"470.759001\" style=\"fill: #8172b3; stroke: #8172b3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb5228dadd8\" x=\"173.690165\" y=\"468.531968\" style=\"fill: #8172b3; stroke: #8172b3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb5228dadd8\" x=\"190.080237\" y=\"462.614235\" style=\"fill: #8172b3; stroke: #8172b3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb5228dadd8\" x=\"206.470308\" y=\"457.126736\" style=\"fill: #8172b3; stroke: #8172b3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb5228dadd8\" x=\"222.86038\" y=\"452.360339\" style=\"fill: #8172b3; stroke: #8172b3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb5228dadd8\" x=\"239.250452\" y=\"447.926315\" style=\"fill: #8172b3; stroke: #8172b3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb5228dadd8\" x=\"255.640524\" y=\"443.96349\" style=\"fill: #8172b3; stroke: #8172b3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb5228dadd8\" x=\"337.590884\" y=\"419.371778\" style=\"fill: #8172b3; stroke: #8172b3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb5228dadd8\" x=\"419.541244\" y=\"401.953763\" style=\"fill: #8172b3; stroke: #8172b3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb5228dadd8\" x=\"501.491604\" y=\"385.629177\" style=\"fill: #8172b3; stroke: #8172b3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb5228dadd8\" x=\"583.441964\" y=\"375.673452\" style=\"fill: #8172b3; stroke: #8172b3; stroke-width: 0.8\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"PathCollection_6\">\n",
       "    <defs>\n",
       "     <path id=\"m6ae6c68937\" d=\"M 0 2.4 \n",
       "C 0.636487 2.4 1.246992 2.147121 1.697056 1.697056 \n",
       "C 2.147121 1.246992 2.4 0.636487 2.4 0 \n",
       "C 2.4 -0.636487 2.147121 -1.246992 1.697056 -1.697056 \n",
       "C 1.246992 -2.147121 0.636487 -2.4 0 -2.4 \n",
       "C -0.636487 -2.4 -1.246992 -2.147121 -1.697056 -1.697056 \n",
       "C -2.147121 -1.246992 -2.4 -0.636487 -2.4 0 \n",
       "C -2.4 0.636487 -2.147121 1.246992 -1.697056 1.697056 \n",
       "C -1.246992 2.147121 -0.636487 2.4 0 2.4 \n",
       "z\n",
       "\" style=\"stroke: #937860; stroke-width: 0.8\"/>\n",
       "    </defs>\n",
       "    <g clip-path=\"url(#p823529a869)\">\n",
       "     <use xlink:href=\"#m6ae6c68937\" x=\"76.169236\" y=\"485.198567\" style=\"fill: #937860; stroke: #937860; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m6ae6c68937\" x=\"86.00328\" y=\"482.514464\" style=\"fill: #937860; stroke: #937860; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m6ae6c68937\" x=\"95.564155\" y=\"478.727639\" style=\"fill: #937860; stroke: #937860; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m6ae6c68937\" x=\"105.398198\" y=\"475.097881\" style=\"fill: #937860; stroke: #937860; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m6ae6c68937\" x=\"115.232241\" y=\"469.745765\" style=\"fill: #937860; stroke: #937860; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m6ae6c68937\" x=\"125.066284\" y=\"465.793244\" style=\"fill: #937860; stroke: #937860; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m6ae6c68937\" x=\"134.62716\" y=\"463.03407\" style=\"fill: #937860; stroke: #937860; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m6ae6c68937\" x=\"144.461203\" y=\"458.006549\" style=\"fill: #937860; stroke: #937860; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m6ae6c68937\" x=\"154.295246\" y=\"453.545498\" style=\"fill: #937860; stroke: #937860; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m6ae6c68937\" x=\"164.129289\" y=\"448.926401\" style=\"fill: #937860; stroke: #937860; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m6ae6c68937\" x=\"173.690165\" y=\"445.036088\" style=\"fill: #937860; stroke: #937860; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m6ae6c68937\" x=\"190.080237\" y=\"436.990575\" style=\"fill: #937860; stroke: #937860; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m6ae6c68937\" x=\"206.470308\" y=\"430.037353\" style=\"fill: #937860; stroke: #937860; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m6ae6c68937\" x=\"222.86038\" y=\"424.431698\" style=\"fill: #937860; stroke: #937860; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m6ae6c68937\" x=\"239.250452\" y=\"419.12257\" style=\"fill: #937860; stroke: #937860; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m6ae6c68937\" x=\"255.640524\" y=\"413.949015\" style=\"fill: #937860; stroke: #937860; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m6ae6c68937\" x=\"337.590884\" y=\"380.656657\" style=\"fill: #937860; stroke: #937860; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m6ae6c68937\" x=\"419.541244\" y=\"360.456234\" style=\"fill: #937860; stroke: #937860; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m6ae6c68937\" x=\"501.491604\" y=\"337.354711\" style=\"fill: #937860; stroke: #937860; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#m6ae6c68937\" x=\"583.441964\" y=\"319.916688\" style=\"fill: #937860; stroke: #937860; stroke-width: 0.8\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"PathCollection_7\">\n",
       "    <defs>\n",
       "     <path id=\"mb569ce7744\" d=\"M 0 2.4 \n",
       "C 0.636487 2.4 1.246992 2.147121 1.697056 1.697056 \n",
       "C 2.147121 1.246992 2.4 0.636487 2.4 0 \n",
       "C 2.4 -0.636487 2.147121 -1.246992 1.697056 -1.697056 \n",
       "C 1.246992 -2.147121 0.636487 -2.4 0 -2.4 \n",
       "C -0.636487 -2.4 -1.246992 -2.147121 -1.697056 -1.697056 \n",
       "C -2.147121 -1.246992 -2.4 -0.636487 -2.4 0 \n",
       "C -2.4 0.636487 -2.147121 1.246992 -1.697056 1.697056 \n",
       "C -1.246992 2.147121 -0.636487 2.4 0 2.4 \n",
       "z\n",
       "\" style=\"stroke: #da8bc3; stroke-width: 0.8\"/>\n",
       "    </defs>\n",
       "    <g clip-path=\"url(#p823529a869)\">\n",
       "     <use xlink:href=\"#mb569ce7744\" x=\"76.169236\" y=\"478.679973\" style=\"fill: #da8bc3; stroke: #da8bc3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb569ce7744\" x=\"86.00328\" y=\"475.043103\" style=\"fill: #da8bc3; stroke: #da8bc3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb569ce7744\" x=\"95.564155\" y=\"470.907596\" style=\"fill: #da8bc3; stroke: #da8bc3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb569ce7744\" x=\"105.398198\" y=\"466.116766\" style=\"fill: #da8bc3; stroke: #da8bc3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb569ce7744\" x=\"115.232241\" y=\"460.707246\" style=\"fill: #da8bc3; stroke: #da8bc3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb569ce7744\" x=\"125.066284\" y=\"455.803192\" style=\"fill: #da8bc3; stroke: #da8bc3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb569ce7744\" x=\"134.62716\" y=\"452.207224\" style=\"fill: #da8bc3; stroke: #da8bc3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb569ce7744\" x=\"144.461203\" y=\"448.404753\" style=\"fill: #da8bc3; stroke: #da8bc3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb569ce7744\" x=\"154.295246\" y=\"444.334583\" style=\"fill: #da8bc3; stroke: #da8bc3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb569ce7744\" x=\"164.129289\" y=\"438.29803\" style=\"fill: #da8bc3; stroke: #da8bc3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb569ce7744\" x=\"173.690165\" y=\"434.178676\" style=\"fill: #da8bc3; stroke: #da8bc3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb569ce7744\" x=\"190.080237\" y=\"426.02585\" style=\"fill: #da8bc3; stroke: #da8bc3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb569ce7744\" x=\"206.470308\" y=\"418.774015\" style=\"fill: #da8bc3; stroke: #da8bc3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb569ce7744\" x=\"222.86038\" y=\"410.739724\" style=\"fill: #da8bc3; stroke: #da8bc3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb569ce7744\" x=\"239.250452\" y=\"403.495475\" style=\"fill: #da8bc3; stroke: #da8bc3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb569ce7744\" x=\"255.640524\" y=\"396.845007\" style=\"fill: #da8bc3; stroke: #da8bc3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb569ce7744\" x=\"337.590884\" y=\"363.583436\" style=\"fill: #da8bc3; stroke: #da8bc3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb569ce7744\" x=\"419.541244\" y=\"333.057839\" style=\"fill: #da8bc3; stroke: #da8bc3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb569ce7744\" x=\"501.491604\" y=\"305.665829\" style=\"fill: #da8bc3; stroke: #da8bc3; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb569ce7744\" x=\"583.441964\" y=\"281.003471\" style=\"fill: #da8bc3; stroke: #da8bc3; stroke-width: 0.8\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"PathCollection_8\">\n",
       "    <defs>\n",
       "     <path id=\"ma2f2878a3f\" d=\"M 0 2.4 \n",
       "C 0.636487 2.4 1.246992 2.147121 1.697056 1.697056 \n",
       "C 2.147121 1.246992 2.4 0.636487 2.4 0 \n",
       "C 2.4 -0.636487 2.147121 -1.246992 1.697056 -1.697056 \n",
       "C 1.246992 -2.147121 0.636487 -2.4 0 -2.4 \n",
       "C -0.636487 -2.4 -1.246992 -2.147121 -1.697056 -1.697056 \n",
       "C -2.147121 -1.246992 -2.4 -0.636487 -2.4 0 \n",
       "C -2.4 0.636487 -2.147121 1.246992 -1.697056 1.697056 \n",
       "C -1.246992 2.147121 -0.636487 2.4 0 2.4 \n",
       "z\n",
       "\" style=\"stroke: #8c8c8c; stroke-width: 0.8\"/>\n",
       "    </defs>\n",
       "    <g clip-path=\"url(#p823529a869)\">\n",
       "     <use xlink:href=\"#ma2f2878a3f\" x=\"76.169236\" y=\"472.175223\" style=\"fill: #8c8c8c; stroke: #8c8c8c; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#ma2f2878a3f\" x=\"86.00328\" y=\"468.553494\" style=\"fill: #8c8c8c; stroke: #8c8c8c; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#ma2f2878a3f\" x=\"95.564155\" y=\"465.49283\" style=\"fill: #8c8c8c; stroke: #8c8c8c; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#ma2f2878a3f\" x=\"105.398198\" y=\"460.329357\" style=\"fill: #8c8c8c; stroke: #8c8c8c; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#ma2f2878a3f\" x=\"115.232241\" y=\"455.540202\" style=\"fill: #8c8c8c; stroke: #8c8c8c; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#ma2f2878a3f\" x=\"125.066284\" y=\"450.858361\" style=\"fill: #8c8c8c; stroke: #8c8c8c; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#ma2f2878a3f\" x=\"134.62716\" y=\"445.846739\" style=\"fill: #8c8c8c; stroke: #8c8c8c; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#ma2f2878a3f\" x=\"144.461203\" y=\"441.667328\" style=\"fill: #8c8c8c; stroke: #8c8c8c; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#ma2f2878a3f\" x=\"154.295246\" y=\"437.292791\" style=\"fill: #8c8c8c; stroke: #8c8c8c; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#ma2f2878a3f\" x=\"164.129289\" y=\"432.834712\" style=\"fill: #8c8c8c; stroke: #8c8c8c; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#ma2f2878a3f\" x=\"173.690165\" y=\"428.462167\" style=\"fill: #8c8c8c; stroke: #8c8c8c; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#ma2f2878a3f\" x=\"190.080237\" y=\"417.782021\" style=\"fill: #8c8c8c; stroke: #8c8c8c; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#ma2f2878a3f\" x=\"206.470308\" y=\"408.415176\" style=\"fill: #8c8c8c; stroke: #8c8c8c; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#ma2f2878a3f\" x=\"222.86038\" y=\"399.99462\" style=\"fill: #8c8c8c; stroke: #8c8c8c; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#ma2f2878a3f\" x=\"239.250452\" y=\"391.31787\" style=\"fill: #8c8c8c; stroke: #8c8c8c; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#ma2f2878a3f\" x=\"255.640524\" y=\"384.303958\" style=\"fill: #8c8c8c; stroke: #8c8c8c; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#ma2f2878a3f\" x=\"337.590884\" y=\"344.336887\" style=\"fill: #8c8c8c; stroke: #8c8c8c; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#ma2f2878a3f\" x=\"419.541244\" y=\"308.218311\" style=\"fill: #8c8c8c; stroke: #8c8c8c; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#ma2f2878a3f\" x=\"501.491604\" y=\"274.969036\" style=\"fill: #8c8c8c; stroke: #8c8c8c; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#ma2f2878a3f\" x=\"583.441964\" y=\"247.876777\" style=\"fill: #8c8c8c; stroke: #8c8c8c; stroke-width: 0.8\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"PathCollection_9\">\n",
       "    <defs>\n",
       "     <path id=\"mb15471775e\" d=\"M 0 2.4 \n",
       "C 0.636487 2.4 1.246992 2.147121 1.697056 1.697056 \n",
       "C 2.147121 1.246992 2.4 0.636487 2.4 0 \n",
       "C 2.4 -0.636487 2.147121 -1.246992 1.697056 -1.697056 \n",
       "C 1.246992 -2.147121 0.636487 -2.4 0 -2.4 \n",
       "C -0.636487 -2.4 -1.246992 -2.147121 -1.697056 -1.697056 \n",
       "C -2.147121 -1.246992 -2.4 -0.636487 -2.4 0 \n",
       "C -2.4 0.636487 -2.147121 1.246992 -1.697056 1.697056 \n",
       "C -1.246992 2.147121 -0.636487 2.4 0 2.4 \n",
       "z\n",
       "\" style=\"stroke: #ccb974; stroke-width: 0.8\"/>\n",
       "    </defs>\n",
       "    <g clip-path=\"url(#p823529a869)\">\n",
       "     <use xlink:href=\"#mb15471775e\" x=\"76.169236\" y=\"454.790051\" style=\"fill: #ccb974; stroke: #ccb974; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb15471775e\" x=\"86.00328\" y=\"451.996295\" style=\"fill: #ccb974; stroke: #ccb974; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb15471775e\" x=\"95.564155\" y=\"448.13345\" style=\"fill: #ccb974; stroke: #ccb974; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb15471775e\" x=\"105.398198\" y=\"444.3367\" style=\"fill: #ccb974; stroke: #ccb974; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb15471775e\" x=\"115.232241\" y=\"442.989607\" style=\"fill: #ccb974; stroke: #ccb974; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb15471775e\" x=\"125.066284\" y=\"436.242889\" style=\"fill: #ccb974; stroke: #ccb974; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb15471775e\" x=\"134.62716\" y=\"432.944491\" style=\"fill: #ccb974; stroke: #ccb974; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb15471775e\" x=\"144.461203\" y=\"428.479679\" style=\"fill: #ccb974; stroke: #ccb974; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb15471775e\" x=\"154.295246\" y=\"424.484581\" style=\"fill: #ccb974; stroke: #ccb974; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb15471775e\" x=\"164.129289\" y=\"420.287942\" style=\"fill: #ccb974; stroke: #ccb974; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb15471775e\" x=\"173.690165\" y=\"416.042593\" style=\"fill: #ccb974; stroke: #ccb974; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb15471775e\" x=\"190.080237\" y=\"408.875281\" style=\"fill: #ccb974; stroke: #ccb974; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb15471775e\" x=\"206.470308\" y=\"399.77326\" style=\"fill: #ccb974; stroke: #ccb974; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb15471775e\" x=\"222.86038\" y=\"393.1587\" style=\"fill: #ccb974; stroke: #ccb974; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb15471775e\" x=\"239.250452\" y=\"385.608663\" style=\"fill: #ccb974; stroke: #ccb974; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb15471775e\" x=\"255.640524\" y=\"378.425863\" style=\"fill: #ccb974; stroke: #ccb974; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb15471775e\" x=\"337.590884\" y=\"340.336921\" style=\"fill: #ccb974; stroke: #ccb974; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb15471775e\" x=\"419.541244\" y=\"309.787017\" style=\"fill: #ccb974; stroke: #ccb974; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb15471775e\" x=\"501.491604\" y=\"277.684274\" style=\"fill: #ccb974; stroke: #ccb974; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mb15471775e\" x=\"583.441964\" y=\"247.069729\" style=\"fill: #ccb974; stroke: #ccb974; stroke-width: 0.8\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"PathCollection_10\">\n",
       "    <defs>\n",
       "     <path id=\"mf2d7f2c1c7\" d=\"M 0 2.4 \n",
       "C 0.636487 2.4 1.246992 2.147121 1.697056 1.697056 \n",
       "C 2.147121 1.246992 2.4 0.636487 2.4 0 \n",
       "C 2.4 -0.636487 2.147121 -1.246992 1.697056 -1.697056 \n",
       "C 1.246992 -2.147121 0.636487 -2.4 0 -2.4 \n",
       "C -0.636487 -2.4 -1.246992 -2.147121 -1.697056 -1.697056 \n",
       "C -2.147121 -1.246992 -2.4 -0.636487 -2.4 0 \n",
       "C -2.4 0.636487 -2.147121 1.246992 -1.697056 1.697056 \n",
       "C -1.246992 2.147121 -0.636487 2.4 0 2.4 \n",
       "z\n",
       "\" style=\"stroke: #64b5cd; stroke-width: 0.8\"/>\n",
       "    </defs>\n",
       "    <g clip-path=\"url(#p823529a869)\">\n",
       "     <use xlink:href=\"#mf2d7f2c1c7\" x=\"76.169236\" y=\"433.189842\" style=\"fill: #64b5cd; stroke: #64b5cd; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mf2d7f2c1c7\" x=\"86.00328\" y=\"429.974103\" style=\"fill: #64b5cd; stroke: #64b5cd; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mf2d7f2c1c7\" x=\"95.564155\" y=\"425.518331\" style=\"fill: #64b5cd; stroke: #64b5cd; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mf2d7f2c1c7\" x=\"105.398198\" y=\"420.265594\" style=\"fill: #64b5cd; stroke: #64b5cd; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mf2d7f2c1c7\" x=\"115.232241\" y=\"414.415726\" style=\"fill: #64b5cd; stroke: #64b5cd; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mf2d7f2c1c7\" x=\"125.066284\" y=\"409.343921\" style=\"fill: #64b5cd; stroke: #64b5cd; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mf2d7f2c1c7\" x=\"134.62716\" y=\"403.838056\" style=\"fill: #64b5cd; stroke: #64b5cd; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mf2d7f2c1c7\" x=\"144.461203\" y=\"400.034194\" style=\"fill: #64b5cd; stroke: #64b5cd; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mf2d7f2c1c7\" x=\"154.295246\" y=\"394.889561\" style=\"fill: #64b5cd; stroke: #64b5cd; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mf2d7f2c1c7\" x=\"164.129289\" y=\"390.660302\" style=\"fill: #64b5cd; stroke: #64b5cd; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mf2d7f2c1c7\" x=\"173.690165\" y=\"384.541944\" style=\"fill: #64b5cd; stroke: #64b5cd; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mf2d7f2c1c7\" x=\"190.080237\" y=\"374.739998\" style=\"fill: #64b5cd; stroke: #64b5cd; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mf2d7f2c1c7\" x=\"206.470308\" y=\"365.07688\" style=\"fill: #64b5cd; stroke: #64b5cd; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mf2d7f2c1c7\" x=\"222.86038\" y=\"357.919968\" style=\"fill: #64b5cd; stroke: #64b5cd; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mf2d7f2c1c7\" x=\"239.250452\" y=\"350.155968\" style=\"fill: #64b5cd; stroke: #64b5cd; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mf2d7f2c1c7\" x=\"255.640524\" y=\"341.670897\" style=\"fill: #64b5cd; stroke: #64b5cd; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mf2d7f2c1c7\" x=\"337.590884\" y=\"295.755527\" style=\"fill: #64b5cd; stroke: #64b5cd; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mf2d7f2c1c7\" x=\"419.541244\" y=\"254.893249\" style=\"fill: #64b5cd; stroke: #64b5cd; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mf2d7f2c1c7\" x=\"501.491604\" y=\"218.757888\" style=\"fill: #64b5cd; stroke: #64b5cd; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mf2d7f2c1c7\" x=\"583.441964\" y=\"180.067264\" style=\"fill: #64b5cd; stroke: #64b5cd; stroke-width: 0.8\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"PathCollection_11\">\n",
       "    <g clip-path=\"url(#p823529a869)\">\n",
       "     <use xlink:href=\"#mced943a113\" x=\"76.169236\" y=\"368.625971\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"86.00328\" y=\"364.275585\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"95.564155\" y=\"358.605953\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"105.398198\" y=\"352.728964\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"115.232241\" y=\"348.106674\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"125.066284\" y=\"341.707026\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"134.62716\" y=\"335.9505\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"144.461203\" y=\"329.01991\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"154.295246\" y=\"323.810889\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"164.129289\" y=\"318.535046\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"173.690165\" y=\"311.497206\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"190.080237\" y=\"300.201974\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"206.470308\" y=\"288.645902\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"222.86038\" y=\"277.777775\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"239.250452\" y=\"267.012093\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"255.640524\" y=\"256.63628\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"337.590884\" y=\"202.506762\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "     <use xlink:href=\"#mced943a113\" x=\"419.541244\" y=\"152.350026\" style=\"fill: #4c72b0; stroke: #4c72b0; stroke-width: 0.8\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"line2d_16\">\n",
       "    <path d=\"M 76.169236 505.611055 \n",
       "L 583.441964 551.1534 \n",
       "L 583.441964 551.1534 \n",
       "\" clip-path=\"url(#p823529a869)\" style=\"fill: none; stroke: #4c72b0; stroke-width: 3; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"line2d_17\">\n",
       "    <path d=\"M 76.169236 502.142395 \n",
       "L 583.441964 485.215109 \n",
       "L 583.441964 485.215109 \n",
       "\" clip-path=\"url(#p823529a869)\" style=\"fill: none; stroke: #dd8452; stroke-width: 3; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"line2d_18\">\n",
       "    <path d=\"M 76.169236 501.139664 \n",
       "L 583.441964 457.737496 \n",
       "L 583.441964 457.737496 \n",
       "\" clip-path=\"url(#p823529a869)\" style=\"fill: none; stroke: #55a868; stroke-width: 3; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"line2d_19\">\n",
       "    <path d=\"M 76.169236 499.122627 \n",
       "L 583.441964 451.133119 \n",
       "L 583.441964 451.133119 \n",
       "\" clip-path=\"url(#p823529a869)\" style=\"fill: none; stroke: #c44e52; stroke-width: 3; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"line2d_20\">\n",
       "    <path d=\"M 76.169236 495.853116 \n",
       "L 583.441964 348.092311 \n",
       "L 583.441964 348.092311 \n",
       "\" clip-path=\"url(#p823529a869)\" style=\"fill: none; stroke: #8172b3; stroke-width: 3; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"line2d_21\">\n",
       "    <path d=\"M 76.169236 486.755145 \n",
       "L 583.441964 269.401807 \n",
       "L 583.441964 269.401807 \n",
       "\" clip-path=\"url(#p823529a869)\" style=\"fill: none; stroke: #937860; stroke-width: 3; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"line2d_22\">\n",
       "    <path d=\"M 76.169236 479.328377 \n",
       "L 583.441964 244.23105 \n",
       "L 583.441964 244.23105 \n",
       "\" clip-path=\"url(#p823529a869)\" style=\"fill: none; stroke: #da8bc3; stroke-width: 3; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"line2d_23\">\n",
       "    <path d=\"M 76.169236 474.172227 \n",
       "L 583.441964 224.819646 \n",
       "L 583.441964 224.819646 \n",
       "\" clip-path=\"url(#p823529a869)\" style=\"fill: none; stroke: #8c8c8c; stroke-width: 3; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"line2d_24\">\n",
       "    <path d=\"M 76.169236 456.840002 \n",
       "L 583.441964 239.499382 \n",
       "L 583.441964 239.499382 \n",
       "\" clip-path=\"url(#p823529a869)\" style=\"fill: none; stroke: #ccb974; stroke-width: 3; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"line2d_25\">\n",
       "    <path d=\"M 76.169236 435.224123 \n",
       "L 583.441964 168.158587 \n",
       "L 583.441964 168.158587 \n",
       "\" clip-path=\"url(#p823529a869)\" style=\"fill: none; stroke: #64b5cd; stroke-width: 3; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"line2d_26\">\n",
       "    <path d=\"M 76.169236 372.190156 \n",
       "L 583.441964 47.1534 \n",
       "L 583.441964 47.1534 \n",
       "\" clip-path=\"url(#p823529a869)\" style=\"fill: none; stroke: #4c72b0; stroke-width: 3; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_3\">\n",
       "    <path d=\"M 50.8056 576.3534 \n",
       "L 50.8056 21.9534 \n",
       "\" style=\"fill: none; stroke: #262626; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_4\">\n",
       "    <path d=\"M 608.8056 576.3534 \n",
       "L 608.8056 21.9534 \n",
       "\" style=\"fill: none; stroke: #262626; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_5\">\n",
       "    <path d=\"M 50.8056 576.3534 \n",
       "L 608.8056 576.3534 \n",
       "\" style=\"fill: none; stroke: #262626; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_6\">\n",
       "    <path d=\"M 50.8056 21.9534 \n",
       "L 608.8056 21.9534 \n",
       "\" style=\"fill: none; stroke: #262626; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"text_18\">\n",
       "    <!-- butt -->\n",
       "    <g style=\"fill: #262626\" transform=\"translate(317.9823 15.9534) scale(0.1152 -0.1152)\">\n",
       "     <defs>\n",
       "      <path id=\"DejaVuSans-62\" d=\"M 3116 1747 \n",
       "Q 3116 2381 2855 2742 \n",
       "Q 2594 3103 2138 3103 \n",
       "Q 1681 3103 1420 2742 \n",
       "Q 1159 2381 1159 1747 \n",
       "Q 1159 1113 1420 752 \n",
       "Q 1681 391 2138 391 \n",
       "Q 2594 391 2855 752 \n",
       "Q 3116 1113 3116 1747 \n",
       "z\n",
       "M 1159 2969 \n",
       "Q 1341 3281 1617 3432 \n",
       "Q 1894 3584 2278 3584 \n",
       "Q 2916 3584 3314 3078 \n",
       "Q 3713 2572 3713 1747 \n",
       "Q 3713 922 3314 415 \n",
       "Q 2916 -91 2278 -91 \n",
       "Q 1894 -91 1617 61 \n",
       "Q 1341 213 1159 525 \n",
       "L 1159 0 \n",
       "L 581 0 \n",
       "L 581 4863 \n",
       "L 1159 4863 \n",
       "L 1159 2969 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"DejaVuSans-75\" d=\"M 544 1381 \n",
       "L 544 3500 \n",
       "L 1119 3500 \n",
       "L 1119 1403 \n",
       "Q 1119 906 1312 657 \n",
       "Q 1506 409 1894 409 \n",
       "Q 2359 409 2629 706 \n",
       "Q 2900 1003 2900 1516 \n",
       "L 2900 3500 \n",
       "L 3475 3500 \n",
       "L 3475 0 \n",
       "L 2900 0 \n",
       "L 2900 538 \n",
       "Q 2691 219 2414 64 \n",
       "Q 2138 -91 1772 -91 \n",
       "Q 1169 -91 856 284 \n",
       "Q 544 659 544 1381 \n",
       "z\n",
       "M 1991 3584 \n",
       "L 1991 3584 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"DejaVuSans-74\" d=\"M 1172 4494 \n",
       "L 1172 3500 \n",
       "L 2356 3500 \n",
       "L 2356 3053 \n",
       "L 1172 3053 \n",
       "L 1172 1153 \n",
       "Q 1172 725 1289 603 \n",
       "Q 1406 481 1766 481 \n",
       "L 2356 481 \n",
       "L 2356 0 \n",
       "L 1766 0 \n",
       "Q 1100 0 847 248 \n",
       "Q 594 497 594 1153 \n",
       "L 594 3053 \n",
       "L 172 3053 \n",
       "L 172 3500 \n",
       "L 594 3500 \n",
       "L 594 4494 \n",
       "L 1172 4494 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "     </defs>\n",
       "     <use xlink:href=\"#DejaVuSans-62\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-75\" x=\"63.476562\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-74\" x=\"126.855469\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-74\" x=\"166.064453\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"legend_1\">\n",
       "    <g id=\"patch_7\">\n",
       "     <path d=\"M 58.1976 200.9025 \n",
       "L 154.8447 200.9025 \n",
       "Q 156.9567 200.9025 156.9567 198.7905 \n",
       "L 156.9567 29.3454 \n",
       "Q 156.9567 27.2334 154.8447 27.2334 \n",
       "L 58.1976 27.2334 \n",
       "Q 56.0856 27.2334 56.0856 29.3454 \n",
       "L 56.0856 198.7905 \n",
       "Q 56.0856 200.9025 58.1976 200.9025 \n",
       "z\n",
       "\" style=\"fill: #ffffff; opacity: 0.8; stroke: #cccccc; stroke-width: 0.8; stroke-linejoin: miter\"/>\n",
       "    </g>\n",
       "    <g id=\"line2d_27\">\n",
       "     <path d=\"M 60.3096 35.78535 \n",
       "L 70.8696 35.78535 \n",
       "L 81.4296 35.78535 \n",
       "\" style=\"fill: none; stroke: #4c72b0; stroke-width: 3; stroke-linecap: round\"/>\n",
       "    </g>\n",
       "    <g id=\"text_19\">\n",
       "     <!-- 7.81 µM -->\n",
       "     <g style=\"fill: #262626\" transform=\"translate(89.8776 39.48135) scale(0.1056 -0.1056)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-2e\" d=\"M 684 794 \n",
       "L 1344 794 \n",
       "L 1344 0 \n",
       "L 684 0 \n",
       "L 684 794 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-38\" d=\"M 2034 2216 \n",
       "Q 1584 2216 1326 1975 \n",
       "Q 1069 1734 1069 1313 \n",
       "Q 1069 891 1326 650 \n",
       "Q 1584 409 2034 409 \n",
       "Q 2484 409 2743 651 \n",
       "Q 3003 894 3003 1313 \n",
       "Q 3003 1734 2745 1975 \n",
       "Q 2488 2216 2034 2216 \n",
       "z\n",
       "M 1403 2484 \n",
       "Q 997 2584 770 2862 \n",
       "Q 544 3141 544 3541 \n",
       "Q 544 4100 942 4425 \n",
       "Q 1341 4750 2034 4750 \n",
       "Q 2731 4750 3128 4425 \n",
       "Q 3525 4100 3525 3541 \n",
       "Q 3525 3141 3298 2862 \n",
       "Q 3072 2584 2669 2484 \n",
       "Q 3125 2378 3379 2068 \n",
       "Q 3634 1759 3634 1313 \n",
       "Q 3634 634 3220 271 \n",
       "Q 2806 -91 2034 -91 \n",
       "Q 1263 -91 848 271 \n",
       "Q 434 634 434 1313 \n",
       "Q 434 1759 690 2068 \n",
       "Q 947 2378 1403 2484 \n",
       "z\n",
       "M 1172 3481 \n",
       "Q 1172 3119 1398 2916 \n",
       "Q 1625 2713 2034 2713 \n",
       "Q 2441 2713 2670 2916 \n",
       "Q 2900 3119 2900 3481 \n",
       "Q 2900 3844 2670 4047 \n",
       "Q 2441 4250 2034 4250 \n",
       "Q 1625 4250 1398 4047 \n",
       "Q 1172 3844 1172 3481 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-37\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-38\" x=\"95.410156\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-31\" x=\"159.033203\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"222.65625\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-b5\" x=\"254.443359\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-4d\" x=\"318.066406\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"line2d_28\">\n",
       "     <path d=\"M 60.3096 51.28545 \n",
       "L 70.8696 51.28545 \n",
       "L 81.4296 51.28545 \n",
       "\" style=\"fill: none; stroke: #dd8452; stroke-width: 3; stroke-linecap: round\"/>\n",
       "    </g>\n",
       "    <g id=\"text_20\">\n",
       "     <!-- 15.62 µM -->\n",
       "     <g style=\"fill: #262626\" transform=\"translate(89.8776 54.98145) scale(0.1056 -0.1056)\">\n",
       "      <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-35\" x=\"63.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-2e\" x=\"127.246094\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-36\" x=\"159.033203\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-32\" x=\"222.65625\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"286.279297\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-b5\" x=\"318.066406\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-4d\" x=\"381.689453\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"line2d_29\">\n",
       "     <path d=\"M 60.3096 66.78555 \n",
       "L 70.8696 66.78555 \n",
       "L 81.4296 66.78555 \n",
       "\" style=\"fill: none; stroke: #55a868; stroke-width: 3; stroke-linecap: round\"/>\n",
       "    </g>\n",
       "    <g id=\"text_21\">\n",
       "     <!-- 31.25 µM -->\n",
       "     <g style=\"fill: #262626\" transform=\"translate(89.8776 70.48155) scale(0.1056 -0.1056)\">\n",
       "      <use xlink:href=\"#DejaVuSans-33\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-31\" x=\"63.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-2e\" x=\"127.246094\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-32\" x=\"159.033203\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-35\" x=\"222.65625\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"286.279297\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-b5\" x=\"318.066406\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-4d\" x=\"381.689453\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"line2d_30\">\n",
       "     <path d=\"M 60.3096 82.28565 \n",
       "L 70.8696 82.28565 \n",
       "L 81.4296 82.28565 \n",
       "\" style=\"fill: none; stroke: #c44e52; stroke-width: 3; stroke-linecap: round\"/>\n",
       "    </g>\n",
       "    <g id=\"text_22\">\n",
       "     <!-- 62.50 µM -->\n",
       "     <g style=\"fill: #262626\" transform=\"translate(89.8776 85.98165) scale(0.1056 -0.1056)\">\n",
       "      <use xlink:href=\"#DejaVuSans-36\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-32\" x=\"63.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-2e\" x=\"127.246094\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-35\" x=\"159.033203\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"222.65625\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"286.279297\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-b5\" x=\"318.066406\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-4d\" x=\"381.689453\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"line2d_31\">\n",
       "     <path d=\"M 60.3096 97.78575 \n",
       "L 70.8696 97.78575 \n",
       "L 81.4296 97.78575 \n",
       "\" style=\"fill: none; stroke: #8172b3; stroke-width: 3; stroke-linecap: round\"/>\n",
       "    </g>\n",
       "    <g id=\"text_23\">\n",
       "     <!-- 125.00 µM -->\n",
       "     <g style=\"fill: #262626\" transform=\"translate(89.8776 101.48175) scale(0.1056 -0.1056)\">\n",
       "      <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-32\" x=\"63.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-35\" x=\"127.246094\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-2e\" x=\"190.869141\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"222.65625\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"286.279297\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"349.902344\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-b5\" x=\"381.689453\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-4d\" x=\"445.3125\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"line2d_32\">\n",
       "     <path d=\"M 60.3096 113.28585 \n",
       "L 70.8696 113.28585 \n",
       "L 81.4296 113.28585 \n",
       "\" style=\"fill: none; stroke: #937860; stroke-width: 3; stroke-linecap: round\"/>\n",
       "    </g>\n",
       "    <g id=\"text_24\">\n",
       "     <!-- 250.00 µM -->\n",
       "     <g style=\"fill: #262626\" transform=\"translate(89.8776 116.98185) scale(0.1056 -0.1056)\">\n",
       "      <use xlink:href=\"#DejaVuSans-32\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-35\" x=\"63.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-2e\" x=\"190.869141\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"222.65625\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"286.279297\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"349.902344\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-b5\" x=\"381.689453\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-4d\" x=\"445.3125\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"line2d_33\">\n",
       "     <path d=\"M 60.3096 128.78595 \n",
       "L 70.8696 128.78595 \n",
       "L 81.4296 128.78595 \n",
       "\" style=\"fill: none; stroke: #da8bc3; stroke-width: 3; stroke-linecap: round\"/>\n",
       "    </g>\n",
       "    <g id=\"text_25\">\n",
       "     <!-- 500.00 µM -->\n",
       "     <g style=\"fill: #262626\" transform=\"translate(89.8776 132.48195) scale(0.1056 -0.1056)\">\n",
       "      <use xlink:href=\"#DejaVuSans-35\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-2e\" x=\"190.869141\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"222.65625\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"286.279297\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"349.902344\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-b5\" x=\"381.689453\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-4d\" x=\"445.3125\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"line2d_34\">\n",
       "     <path d=\"M 60.3096 144.28605 \n",
       "L 70.8696 144.28605 \n",
       "L 81.4296 144.28605 \n",
       "\" style=\"fill: none; stroke: #8c8c8c; stroke-width: 3; stroke-linecap: round\"/>\n",
       "    </g>\n",
       "    <g id=\"text_26\">\n",
       "     <!-- 1000.00 µM -->\n",
       "     <g style=\"fill: #262626\" transform=\"translate(89.8776 147.98205) scale(0.1056 -0.1056)\">\n",
       "      <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"190.869141\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-2e\" x=\"254.492188\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"286.279297\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"349.902344\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"413.525391\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-b5\" x=\"445.3125\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-4d\" x=\"508.935547\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"line2d_35\">\n",
       "     <path d=\"M 60.3096 159.78615 \n",
       "L 70.8696 159.78615 \n",
       "L 81.4296 159.78615 \n",
       "\" style=\"fill: none; stroke: #ccb974; stroke-width: 3; stroke-linecap: round\"/>\n",
       "    </g>\n",
       "    <g id=\"text_27\">\n",
       "     <!-- 2000.00 µM -->\n",
       "     <g style=\"fill: #262626\" transform=\"translate(89.8776 163.48215) scale(0.1056 -0.1056)\">\n",
       "      <use xlink:href=\"#DejaVuSans-32\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"190.869141\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-2e\" x=\"254.492188\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"286.279297\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"349.902344\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"413.525391\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-b5\" x=\"445.3125\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-4d\" x=\"508.935547\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"line2d_36\">\n",
       "     <path d=\"M 60.3096 175.28625 \n",
       "L 70.8696 175.28625 \n",
       "L 81.4296 175.28625 \n",
       "\" style=\"fill: none; stroke: #64b5cd; stroke-width: 3; stroke-linecap: round\"/>\n",
       "    </g>\n",
       "    <g id=\"text_28\">\n",
       "     <!-- 4000.00 µM -->\n",
       "     <g style=\"fill: #262626\" transform=\"translate(89.8776 178.98225) scale(0.1056 -0.1056)\">\n",
       "      <use xlink:href=\"#DejaVuSans-34\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"190.869141\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-2e\" x=\"254.492188\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"286.279297\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"349.902344\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"413.525391\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-b5\" x=\"445.3125\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-4d\" x=\"508.935547\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"line2d_37\">\n",
       "     <path d=\"M 60.3096 190.78635 \n",
       "L 70.8696 190.78635 \n",
       "L 81.4296 190.78635 \n",
       "\" style=\"fill: none; stroke: #4c72b0; stroke-width: 3; stroke-linecap: round\"/>\n",
       "    </g>\n",
       "    <g id=\"text_29\">\n",
       "     <!-- 8000.00 µM -->\n",
       "     <g style=\"fill: #262626\" transform=\"translate(89.8776 194.48235) scale(0.1056 -0.1056)\">\n",
       "      <use xlink:href=\"#DejaVuSans-38\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"190.869141\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-2e\" x=\"254.492188\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"286.279297\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"349.902344\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"413.525391\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-b5\" x=\"445.3125\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-4d\" x=\"508.935547\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "  </g>\n",
       " </g>\n",
       " <defs>\n",
       "  <clipPath id=\"p823529a869\">\n",
       "   <rect x=\"50.8056\" y=\"21.9534\" width=\"558\" height=\"554.4\"/>\n",
       "  </clipPath>\n",
       " </defs>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<Figure size 1000x1000 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Sanity check:\n",
    "kinetics.get_initial_slopes(time_data[:,0], product_concentration[:,1750,:].T, substrate_concs=conc_data, title='butt', plot=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/nfreitas/workspace/pinney/kinetic_analysis/kinetics/functions.py:52: RuntimeWarning:\n",
      "\n",
      "invalid value encountered in scalar divide\n",
      "\n",
      "/home/nfreitas/workspace/pinney/kinetic_analysis/kinetics/functions.py:57: RuntimeWarning:\n",
      "\n",
      "invalid value encountered in scalar divide\n",
      "\n",
      "/home/nfreitas/workspace/pinney/kinetic_analysis/kinetics/functions.py:58: RuntimeWarning:\n",
      "\n",
      "invalid value encountered in scalar divide\n",
      "\n",
      "/home/nfreitas/workspace/pinney/kinetic_analysis/kinetics/functions.py:89: RuntimeWarning:\n",
      "\n",
      "invalid value encountered in scalar divide\n",
      "\n",
      "/home/nfreitas/workspace/pinney/kinetic_analysis/kinetics/functions.py:92: RuntimeWarning:\n",
      "\n",
      "invalid value encountered in scalar divide\n",
      "\n",
      "/home/nfreitas/workspace/pinney/kinetic_analysis/kinetics/functions.py:93: RuntimeWarning:\n",
      "\n",
      "invalid value encountered in scalar divide\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 7.8125 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 15.625 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 31.25 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 62.5 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 125.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 250.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 500.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 1000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 2000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 4000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n",
      "WARNING: slope for 8000.0 µM is NaN.\n",
      " This is likely due to scipy failing to fit a linear model to the data.\n"
     ]
    }
   ],
   "source": [
    "###TODO:\n",
    "# 1) ~~Is this working properly? Plots look weird.\n",
    "#    Might be a weird dataset + line fitting errors, or a bug~~\n",
    "#    Fixed! Was a bug in how time values were stored.\n",
    "# 2) ~~I think addings nans kills the line fitting. Fix at some point. (in line fitting code, so we can keep nans?)~~\n",
    "#    Fixed? Appears like it's working properly.\n",
    "# 3) We're using the time values for the first chamber for all chambers. \n",
    "#    In practice, they're sometimes off by 1 second, which should not affect results much. Perhaps change in the future.\n",
    "# 4) Kinetics fails without substrate_conc, even though it says it's optional. Also prints an ugly array which we should disable.\n",
    "\n",
    "#make an array of initial slopes for each chamber: should be (#chambers , #concentrations) = (1792 x 11)\n",
    "initial_slopes = None\n",
    "initial_slopes_R2 = None\n",
    "for i, chamber_idx in enumerate(chamber_idxs):\n",
    "\n",
    "    current_chamber_slopes, current_chamber_R2 = kinetics.get_initial_slopes(time_data[:,0], product_concentration[:,i,:].T, substrate_concs=conc_data)\n",
    "\n",
    "    #add a dimension at the end:\n",
    "    current_chamber_slopes = np.expand_dims(current_chamber_slopes, axis=0)\n",
    "    current_chamber_R2 = np.expand_dims(current_chamber_R2, axis=0)\n",
    "    \n",
    "    #add to our complete array:\n",
    "    if initial_slopes is None:\n",
    "        initial_slopes = current_chamber_slopes\n",
    "        initial_slopes_R2 = current_chamber_R2\n",
    "    else:\n",
    "        initial_slopes = np.concatenate([initial_slopes, current_chamber_slopes], axis=0)\n",
    "        initial_slopes_R2 = np.concatenate([initial_slopes_R2, current_chamber_R2], axis=0)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Filter initial rates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Filters:\n",
    "We will filter by masking our product_concentration array with NaN values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "### TODO:\n",
    "# Currently I haven't implemented the manually flagged wells, and positive initial slope filters.\n",
    "# I also have the R^2 filter using the min of all assays for chamber, instead of looking at each assay/chamber individually.\n",
    "# I will need to change how I'm applying these filters to make better sense soon."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Filter threshholds ###\n",
    "\n",
    "#standard curve r^2 threshhold:\n",
    "r2_threshold = 0.999\n",
    "\n",
    "#enzyme expression threshhold:\n",
    "expression_threshhold = 1.5\n",
    "expression_threshhold_units = 'nM'\n",
    "\n",
    "#initial rate fit R^2 threshhold:\n",
    "intitial_rate_R2_threshhold = 0.8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19712"
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(initial_slopes.flatten())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1792,)"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enzyme_concentration.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pearson r^2 filter: 1761/1792 chambers pass\n",
      "Enzyme expression filter: 1197/1792 chambers pass\n",
      "Initial Rate R^2 filter: 1792/1792 chambers pass with 10 or more slopes.\n",
      "Positive Initial Slope filter: 1548/1792 chambers pass with 10 or more slopes.\n"
     ]
    }
   ],
   "source": [
    "### Make filters ###\n",
    "\n",
    "# STANDARD CURVE FILTER #\n",
    "# overwrite all chambers (rows) with r^2 values below the threshold with NaNs:\n",
    "filter_r2 = np.ones_like(initial_slopes)\n",
    "r2_values = np.array([db_conn._json_dict['runs']['standard_0']['analyses']['linear_regression']['chambers'][chamber_idx]['r2'] for chamber_idx in chamber_idxs])\n",
    "_count = 0\n",
    "for i, chamber_idx in enumerate(chamber_idxs):\n",
    "    if r2_values[i] < r2_threshold:\n",
    "        _count +=1\n",
    "        filter_r2[i, :] = np.nan\n",
    "print('Pearson r^2 filter: {}/{} chambers pass'.format(len(chamber_idxs)-_count, len(chamber_idxs)))\n",
    "\n",
    "# ENZYME EXPRESSION FILTER #\n",
    "# overwrite all chambers (rows) with enzyme expression below the threshold with NaNs:\n",
    "#Double check the expression units match the EGFP units:\n",
    "assert expression_threshhold_units == EGFP_SLOPE_CONC_UNITS, 'Error, enzyme expression and EGFP standard curve units do not match!'\n",
    "filter_enzyme_expression = np.ones_like(product_concentration)\n",
    "_count = 0\n",
    "for i, chamber_idx in enumerate(chamber_idxs):\n",
    "    if enzyme_concentration[i] < expression_threshhold:\n",
    "        _count +=1\n",
    "        filter_enzyme_expression[:,i,:] = np.nan\n",
    "print('Enzyme expression filter: {}/{} chambers pass'.format(len(chamber_idxs)-_count, len(chamber_idxs)))\n",
    "\n",
    "# INITIAL RATE FIT FILTER #\n",
    "# overwrite just the assays per chamber (single values) with initial rate fit R^2 values below the threshold with NaNs:\n",
    "filter_initial_rate_R2 = np.ones_like(initial_slopes_R2)\n",
    "_count = 0\n",
    "for i, chamber_idx in enumerate(chamber_idxs):\n",
    "    _chamber_count = 0\n",
    "    for j in range(len(conc_data)):\n",
    "        if initial_slopes_R2[i,j] < intitial_rate_R2_threshhold:\n",
    "            _chamber_count +=1\n",
    "            filter_initial_rate_R2[i,j] = np.nan\n",
    "    if len(conc_data) - _chamber_count < 10:\n",
    "        _count +=1\n",
    "print('Initial Rate R^2 filter: {}/{} chambers pass with 10 or more slopes.'.format(len(chamber_idxs)-_count, len(chamber_idxs)))\n",
    "\n",
    "# POSITIVE INITIAL SLOPE FILTER #\n",
    "# overwrite just the assays per chamber (single values) with initial slopes below zero with NaNs:\n",
    "filter_positive_initial_slope = np.ones_like(initial_slopes)\n",
    "_count = 0\n",
    "for i, chamber_idx in enumerate(chamber_idxs):\n",
    "    _chamber_count = 0\n",
    "    for j in range(len(conc_data)):\n",
    "        if initial_slopes[i,j] < 0:\n",
    "            _chamber_count +=1\n",
    "            filter_positive_initial_slope[i,j] = np.nan\n",
    "    if len(conc_data) - _chamber_count < 10:\n",
    "        _count +=1\n",
    "print('Positive Initial Slope filter: {}/{} chambers pass with 10 or more slopes.'.format(len(chamber_idxs)-_count, len(chamber_idxs)))\n",
    "\n",
    "\n",
    "### manually flagged wells ###\n",
    "#TODO: implement\n",
    "\n",
    "### TODO: make visualization!\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Here, we should take our filters from above, and ignore all wells that don't pass\n",
    "# # pearson r2 value filter\n",
    "# # slope < 0 filter\n",
    "# # low enz expression\n",
    "# # manually flagged wells\n",
    "# # poor initial rate fits\n",
    "\n",
    "\n",
    "# # NOTE: Need to decided on some standard filtering criteria for experiments\n",
    "# # drop rows with initial rates =< 0\n",
    "# cleaned_kinetics_df_filtered = cleaned_kinetics_df[cleaned_kinetics_df['initial_rate'] > 0]\n",
    "\n",
    "# #drop rows with poor standard curves\n",
    "# cleaned_kinetics_df_filtered = cleaned_kinetics_df_filtered[cleaned_kinetics_df_filtered['std_r2'] > 0.98]\n",
    "\n",
    "# #drop rows with poor expression\n",
    "# cleaned_kinetics_df_filtered = cleaned_kinetics_df_filtered[cleaned_kinetics_df_filtered['enzyme_conc_nM'] > 1.5]\n",
    "\n",
    "# #drop flagged rows\n",
    "# cleaned_kinetics_df_filtered = cleaned_kinetics_df_filtered[cleaned_kinetics_df_filtered['mm_tofit_mask'] == True]\n",
    "\n",
    "# #drop rows with poor initial rate fits\n",
    "# cleaned_kinetics_df_filtered = cleaned_kinetics_df_filtered[cleaned_kinetics_df_filtered['filter_initial_rate_r2'] > 0.8]\n",
    "\n",
    "# cleaned_kinetics_df_filtered =cleaned_kinetics_df_filtered.dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #NF: Remove this, replace with visualization + flags.\n",
    "\n",
    "# #manually inspect button quant image and exclude those with issues (positions are inverted from whats on the button summary image)\n",
    "# to_drop = [(8,26),\n",
    "#           (11,23),\n",
    "#           (25,5),\n",
    "#           (24,4),\n",
    "#           (25,4),\n",
    "#           (31,1),\n",
    "#           (28,1),\n",
    "#           (22,1),\n",
    "#           (21,1),\n",
    "#           (20,1),\n",
    "#           (19,1),\n",
    "#           (17,1),\n",
    "#           (12,1),\n",
    "#           (11,1),\n",
    "#           (9,1),\n",
    "#           (4,1),\n",
    "#           (3,1),\n",
    "#           (3,32),\n",
    "#           (51,32),\n",
    "#           (53,32),\n",
    "#           (55,32),\n",
    "#           (56,32),\n",
    "#           ]\n",
    "\n",
    "# cleaned_kinetics_df_filtered = cleaned_kinetics_df_filtered.loc[~cleaned_kinetics_df_filtered['position'].isin(to_drop)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # save fit filtered initial rates to a .csv\n",
    "# cleaned_kinetics_df_filtered.to_csv('5_filtered_initial_rates.csv')\n",
    "# cleaned_kinetics_df_filtered.head(2)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Fit Michalis-Menten curves and visualize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# takes a DataFrame of initial rates at different substrate concentrations as input and compresses these values \n",
    "# into a single row pr chamber by returning a series where initial rate values and substrate concentrations are each a list\n",
    "def squeeze_initial_rates(df):\n",
    "    # Create a copy of the input DataFrame and sort it based on 'substrate_conc' column.\n",
    "    dfcopy = df.copy().sort_values('substrate_conc')\n",
    "    \n",
    "    # Extract relevant data from the first row of the sorted DataFrame.\n",
    "    # These values are assumed to be the same for all rows in the group.\n",
    "    chamber_id = dfcopy.iloc[0]['id']\n",
    "    position = dfcopy.iloc[0]['position']\n",
    "    times = dfcopy.iloc[0]['time_s']\n",
    "    sum_chamber = dfcopy.iloc[0]['chamber_sum']\n",
    "    std_slope = dfcopy.iloc[0]['std_slope']\n",
    "    std_r2 = dfcopy.iloc[0]['std_r2']\n",
    "    substrate_name = dfcopy.iloc[0]['substrate_name']\n",
    "    assay_type = dfcopy.iloc[0]['assay_type']\n",
    "    chamber_product_concs = dfcopy.iloc[0]['chamber_product_concs']\n",
    "    summed_button = dfcopy.iloc[0]['summed_button']\n",
    "    summed_button_BGsub = dfcopy.iloc[0]['summed_button_BGsub']\n",
    "    enzyme_conc_nM = dfcopy.iloc[0]['enzyme_conc_nM']\n",
    "    \n",
    "    # Insert zeros at the beginning of initial rate and substrate_conc columns to anchor our michaelis menten curves at 0,0.\n",
    "    substrate_conc = np.insert(dfcopy.substrate_conc.values.tolist(),0,0,axis = 0)\n",
    "    initial_rate = np.insert(dfcopy.initial_rate.values.tolist(),0,0,axis =0)\n",
    "    intercepts = np.insert(dfcopy.intercepts.values.tolist(), 0,0,axis = 0)\n",
    "    initial_rate_r2 = np.insert(dfcopy.initial_rate_r2.values.tolist(),0,0,axis = 0)\n",
    "\n",
    "    filter_initial_rate = np.insert(dfcopy.filter_initial_rate.values.tolist(),0,0,axis=0)\n",
    "    filter_intercepts = np.insert(dfcopy.filter_intercepts.values.tolist(), 0,0,axis = 0)\n",
    "    filter_initial_rate_r2 = np.insert(dfcopy.filter_initial_rate_r2.values.tolist(),0,0,axis = 0)\n",
    "    \n",
    "    # Create a new pandas Series to store the extracted data.\n",
    "    squeezed_series = pd.Series(data = [chamber_id, times, sum_chamber, std_slope, std_r2, substrate_name, assay_type,chamber_product_concs,\n",
    "                                        summed_button,summed_button_BGsub, substrate_conc, initial_rate, intercepts, initial_rate_r2, \n",
    "                                        filter_initial_rate, filter_intercepts, filter_initial_rate_r2, enzyme_conc_nM],\n",
    "                               index = ['id', 'time_s', 'sum_chamber', 'std_slope', 'std_r2', 'substrate_name', 'assay_type', 'chamber_product_concs', \n",
    "                                        'summed_button', 'summed_button_BGsub', 'substrate_conc', 'initial_rate', 'intercept', 'initial_rate_r2', \n",
    "                                        'filter_initial_rate', 'filter_intercepts', 'filter_initial_rate_r2','enzyme_conc_nM'])\n",
    "    return squeezed_series\n",
    "\n",
    "\n",
    "# Group the 'cleaned_kinetics_df_filtered' DataFrame by 'position' and apply 'squeeze_initial_rates' function to each group.\n",
    "squeezed_cleaned_fltered_kinetic_data = cleaned_kinetics_df_filtered.groupby('position').apply(squeeze_initial_rates).reset_index()\n",
    "squeezed_cleaned_fltered_kinetic_data.head(5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fit_MM_by_chamber(row):\n",
    "    #define the minimum number of initial rates needed to fit a michaelis menten curve. I usually expect 5 or 6.\n",
    "    if len(row['initial_rate']) >5:\n",
    "        try:\n",
    "            params, _ = curve_fit(mm_model, np.insert(row['substrate_conc'], 0, 0), np.insert(row['initial_rate'], 0, 0) , p0=[np.median(row['substrate_conc']), np.max(row['initial_rate'])])\n",
    "            vmax = params[0]\n",
    "            km = params[1]\n",
    "        except TypeError: \n",
    "            print(\"Could not fit\")\n",
    "            vmax = np.nan\n",
    "            km = np.nan\n",
    "    \n",
    "\n",
    "        try:\n",
    "            params, _ = curve_fit(mm_model, np.insert(row['substrate_conc'],0,0), np.insert(row['filter_initial_rate'],0,0) , p0=[np.median(row['substrate_conc']), np.max(row['filter_initial_rate'])])\n",
    "            vmax_filter = params[0]\n",
    "            km_filter = params[1]\n",
    "        except TypeError: \n",
    "            print(\"Could not fit\")\n",
    "            vmax_filter = np.nan\n",
    "            km_filter = np.nan\n",
    "    \n",
    "    \n",
    "    else:\n",
    "        vmax = np.nan\n",
    "        km = np.nan\n",
    "        vmax_filter = np.nan\n",
    "        km_filter = np.nan\n",
    "        \n",
    "    return vmax, km, vmax_filter, km_filter\n",
    "\n",
    "squeezed_cleaned_fltered_kinetic_data[['vmax', 'km', 'vmax_filter', 'km_filter']] = squeezed_cleaned_fltered_kinetic_data.apply(fit_MM_by_chamber, axis=1, result_type='expand')\n",
    "squeezed_cleaned_fltered_kinetic_data.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Idea: Once we have initial fit for KM, have a view that flags each well by substrate \"coverage\":\n",
    "#that is, whether our experiment had 1/10th and 10x substrate concentration than Vmax"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#extract kcat from vmax\n",
    "squeezed_cleaned_fltered_kinetic_data['kcat'] = squeezed_cleaned_fltered_kinetic_data['vmax'].div((squeezed_cleaned_fltered_kinetic_data['enzyme_conc_nM']/1000))\n",
    "squeezed_cleaned_fltered_kinetic_data['kcat_filtered'] = squeezed_cleaned_fltered_kinetic_data['vmax_filter'].div((squeezed_cleaned_fltered_kinetic_data['enzyme_conc_nM']/1000))\n",
    "squeezed_cleaned_fltered_kinetic_data.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save michaelis menten fits to .csv\n",
    "squeezed_cleaned_fltered_kinetic_data.to_csv('6_mm_fits_unfiltered.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter out outliers for each variant\n",
    "MM_outliers_removed = squeezed_cleaned_fltered_kinetic_data.copy()\n",
    "\n",
    "def filter_outliers(group, z_threshold = 1.5):\n",
    "    z_scores = (group - group.mean()) / group.std()\n",
    "    return group[abs(z_scores) <= z_threshold]\n",
    "\n",
    "# Define the columns for filtering\n",
    "columns = ['enzyme_conc_nM', 'kcat', 'km'] #I also filtered on kcat and Km to exclude some bizarre fits that were unbounded\n",
    "\n",
    "for column in columns:\n",
    "    MM_outliers_removed[column + '_z_filtered'] = MM_outliers_removed.groupby('id')[column].transform(filter_outliers)\n",
    "    \n",
    "MM_outliers_removed = MM_outliers_removed.dropna()\n",
    "MM_outliers_removed.head(5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save filtered michaelis menten fits to .csv\n",
    "MM_outliers_removed.to_csv('7_mm_fits_filtered.csv')\n",
    "MM_outliers_removed.shape[0]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Note: plot chip by R2 of exponential fit, then show individual graphs\n",
    "#Or, do same just by linear fit.\n",
    "\n",
    "#plot michaelis menten curves by variant/id\n",
    "\n",
    "# Group the data by 'id' column\n",
    "groups = MM_outliers_removed.groupby('id')\n",
    "\n",
    "# Calculate the number of rows and columns\n",
    "num_groups = len(groups)\n",
    "num_columns = 4\n",
    "num_rows, remaining = divmod(num_groups, num_columns)\n",
    "if remaining > 0:\n",
    "    num_rows += 1\n",
    "\n",
    "#print(groups)\n",
    "\n",
    "# Create a figure and subplots\n",
    "fig, axs = plt.subplots(num_rows, num_columns, figsize=(15, 4*num_rows), sharex=True)\n",
    "\n",
    "# Iterate over the groups and plot the data on separate subplots\n",
    "for i, (ortholog_name, ortholog_data) in enumerate(groups):\n",
    "    # Get the corresponding subplot for the current group\n",
    "    ax = axs.flat[i]\n",
    "    vmax_norm_ave = []\n",
    "    # Plot the data for the current group\n",
    "    for _, row in ortholog_data.iterrows():\n",
    "        ax.scatter(row['substrate_conc'], [(v/(row['enzyme_conc_nM']/1000)) for v in row['initial_rate']], label=row['position'])\n",
    "        x = np.linspace(0, 16000, 1000)\n",
    "        ax.plot(x, mm_model(x, row['kcat'], row['km']))\n",
    "    \n",
    "    # Set labels and title for the subplot\n",
    "    ax.set_xlabel('Substrate Concentration (µM)')\n",
    "    ax.set_ylabel('Initial Rate/[E] ($s^{-1}$)')\n",
    "    ax.set_title(f'{ortholog_name}')\n",
    "    \n",
    "    # Show the legend for the subplot\n",
    "    ax.legend(fontsize='small')\n",
    "    ax.legend()\n",
    "\n",
    "# Adjust spacing between subplots\n",
    "fig.tight_layout()\n",
    "plt.rcParams['pdf.fonttype'] = 42\n",
    "\n",
    "# plt.savefig('{}.pdf'.format('MM_curves_unfiltered_initial_rate'), format='pdf')\n",
    "plt.clf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# takes a DataFrame of michaelis menten parameters for different chambers as input and compresses these values \n",
    "# into a single row per variant by returning a series where michaelis menten parameters are each a list\n",
    "\n",
    "def squeeze_replicates(df):\n",
    "    # Create a copy of the input DataFrame and sort it based on 'id' column.\n",
    "    dfcopy = df.copy().sort_values('id')\n",
    "    \n",
    "    # Extract relevant information that is assumed to be the same for all rows within the group.\n",
    "    substrate_name = dfcopy.iloc[0]['substrate_name']\n",
    "    assay_type = dfcopy.iloc[0]['assay_type']\n",
    "\n",
    "    # Calculate the number of replicates in the group.\n",
    "    replicates = dfcopy.shape[0]\n",
    "\n",
    "    # Extract specific column data as lists for each replicate within the group.\n",
    "    summed_button = dfcopy.summed_button.values.tolist()\n",
    "    summed_button_BGsub = dfcopy.summed_button_BGsub.values.tolist()\n",
    "    positions = dfcopy.position.values.tolist()\n",
    "    substrate_conc = dfcopy.substrate_conc.values.tolist()\n",
    "    initial_rate = dfcopy.initial_rate.values.tolist()\n",
    "    initial_rate_r2 = dfcopy.initial_rate_r2.values.tolist()\n",
    "    vmax = dfcopy.vmax.values.tolist()\n",
    "    km = dfcopy.km.values.tolist()\n",
    "    vmax_filter = dfcopy.vmax_filter.values.tolist()\n",
    "    km_filter = dfcopy.km_filter.values.tolist()\n",
    "    kcat = dfcopy.kcat.values.tolist()\n",
    "    kcat_filtered = dfcopy.kcat_filtered.values.tolist()\n",
    "    enzyme_conc_nM = dfcopy.enzyme_conc_nM.values.tolist()\n",
    "\n",
    "    # Create a new pandas Series to store the extracted data for all replicates within the group.\n",
    "    squeezed_series = pd.Series(data = [substrate_name, assay_type, summed_button, summed_button_BGsub, positions, substrate_conc, initial_rate, initial_rate_r2, vmax, km, vmax_filter, km_filter, kcat, kcat_filtered,enzyme_conc_nM, replicates],\n",
    "                               index = ['substrate_name', 'assay_type', 'summed_button', 'summed_button_BGsub', 'positions', 'substrate_conc', 'initial_rate', 'initial_rate_r2', 'vmax', 'km', 'vmax_filter', 'km_filter', 'kcat', 'kcat_filtered','enzyme_conc_nM', 'replicates'])\n",
    "    return squeezed_series\n",
    "\n",
    "# Group the DataFrame by 'id' and apply 'squeeze_replicates' function to each group of replicates.\n",
    "squeezed_MM_data = MM_outliers_removed.groupby('id').apply(squeeze_replicates).reset_index()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to calculate various statistics related to Michaelis-Menten (MM) kinetic parameters.\n",
    "# The function takes a row (replicate) from the DataFrame as input and returns calculated statistics.\n",
    "\n",
    "def MM_stats(row):\n",
    "    \n",
    "    # Function to calculate the interquartile range (IQR) of a given data array.\n",
    "    def IQR(data):\n",
    "        q1 = np.percentile(data, 25)\n",
    "        q3 = np.percentile(data, 75)\n",
    "    \n",
    "        # Calculate the interquartile range (IQR)\n",
    "        iqr = q3 - q1\n",
    "        return iqr\n",
    "\n",
    "\n",
    "        \n",
    "    # Extract relevant data for calculations from the row (set of replicates).\n",
    "    summed_button = row['summed_button']\n",
    "    vmax = row['vmax']\n",
    "    vmax_filter = row['vmax_filter']\n",
    "    \n",
    "    kcats = row['kcat']\n",
    "    kcats_filtered = row['kcat_filtered']\n",
    "    vmaxs = row['vmax']\n",
    "    vmaxs_filter = row['vmax_filter']\n",
    "\n",
    "    # Calculate various statistics for kcat, Km, and vmax parameters and enzyme concentration.\n",
    "    kcat_mean = np.mean(kcats)\n",
    "    kcat_stdev = np.std(kcats)\n",
    "    kcat_median = np.median(kcats)\n",
    "    kcat_iqr = IQR(kcats)\n",
    "\n",
    "    kcat_mean_filtered = np.mean(kcats_filtered)\n",
    "    kcat_stdev_filtered = np.std(kcats_filtered)\n",
    "    kcat_median_filtered = np.median(kcats_filtered)\n",
    "    kcat_iqr_filtered = IQR(kcats_filtered)\n",
    "\n",
    "    Km_mean = np.mean(row['km'])\n",
    "    Km_stdev = np.std(row['km'])\n",
    "    Km_median = np.median(row['km'])\n",
    "    Km_iqr = IQR(row['km'])\n",
    "\n",
    "    Km_mean_filtered = np.mean(row['km_filter'])\n",
    "    Km_stdev_filtered = np.std(row['km_filter'])\n",
    "    Km_median_filtered = np.median(row['km_filter'])\n",
    "    Km_iqr_filtered = IQR(row['km_filter'])\n",
    "\n",
    "    vmax_mean = np.mean(vmaxs)\n",
    "    vmax_stdev = np.std(vmaxs)\n",
    "    vmax_median = np.median(vmaxs)\n",
    "    vmax_iqr = IQR(vmaxs)\n",
    "\n",
    "    vmax_mean_filtered = np.mean(vmaxs_filter)\n",
    "    vmax_stdev_filtered = np.std(vmaxs_filter)\n",
    "    vmax_median_filtered = np.median(vmaxs_filter)\n",
    "    vmax_iqr_filtered = IQR(vmaxs_filter)\n",
    "\n",
    "    enzyme_conc_nM_mean = np.mean(row['enzyme_conc_nM'])\n",
    "    enzyme_conc_nM_stdev = np.std(row['enzyme_conc_nM'])\n",
    "    \n",
    "    return kcat_mean, kcat_stdev, kcat_median, kcat_iqr, Km_mean, Km_stdev, Km_median, Km_iqr, kcat_mean_filtered, kcat_stdev_filtered, kcat_median_filtered, kcat_iqr_filtered, Km_mean_filtered, Km_stdev_filtered, Km_median_filtered, Km_iqr_filtered, enzyme_conc_nM_mean, enzyme_conc_nM_stdev,vmax_mean, vmax_stdev, vmax_median, vmax_iqr, vmax_mean_filtered, vmax_stdev_filtered, vmax_median_filtered, vmax_iqr_filtered\n",
    "    \n",
    "\n",
    "squeezed_MM_data[['kcat_mean', 'kcat_stdev', 'kcat_median', 'kcat_iqr', 'Km_mean', 'Km_stdev', 'Km_median', 'Km_iqr', 'kcat_mean_filtered', \n",
    "                  'kcat_stdev_filtered', 'kcat_median_filtered', 'kcat_iqr_filtered', 'Km_mean_filtered', 'Km_stdev_filtered', 'Km_median_filtered', \n",
    "                  'Km_iqr_filtered', 'enzyme_conc_nM_mean','enzyme_conc_nM_stdev', 'vmax_mean', 'vmax_stdev', 'vmax_median', 'vmax_iqr', 'vmax_mean_filtered', \n",
    "                 'vmax_stdev_filtered', 'vmax_median_filtered', 'vmax_iqr_filtered']] = squeezed_MM_data.apply(MM_stats, axis=1, result_type='expand')\n",
    "squeezed_MM_data.head(10)         "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#clean up dataframe\n",
    "cleaned_MM = squeezed_MM_data.drop(\n",
    "    columns=[\n",
    "        \"summed_button\",\n",
    "        \"summed_button_BGsub\",\n",
    "        \"positions\",\n",
    "        \"initial_rate\",\n",
    "        \"initial_rate_r2\",\n",
    "        \"vmax\", \n",
    "        \"km\", \n",
    "        \"substrate_conc\",\n",
    "        \"vmax_filter\",\n",
    "        \"km_filter\",\n",
    "        \"kcat\",\n",
    "        \"kcat_filtered\",\n",
    "        \"enzyme_conc_nM\"\n",
    "    ]\n",
    ")\n",
    "\n",
    "cleaned_MM.sort_values(by = 'kcat_median')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add organism name to each row\n",
    "def org_name(row):\n",
    "    #with open(\"/Users/margauxpinney/code/lab-notebook-pinney/20230629/adkLibrary_original.fasta\", 'r') as lib_file:\n",
    "    with open(root+'/adkLibrary_original.fasta', 'r') as lib_file:\n",
    "        orth_dict = {}\n",
    "        for line in lib_file.readlines():\n",
    "            if line[0] == \">\":\n",
    "            \n",
    "                org_name = line.split(\"|\")[2]\n",
    "        \n",
    "                abbrev = f\"{org_name[0].upper()}_{org_name.split('_')[-1][:4]}_ADK\"\n",
    "                orth_dict[abbrev] = org_name\n",
    "    orth_dict['bsADK'] = 'bacillus_subtilis'\n",
    "    orth_dict['gsADK'] = 'geobacillus_stearothermophilus'\n",
    "    orth_dict['bmADK'] = 'bacillus_marinus'\n",
    "    orth_dict['Sym_ther_ADK'] = 'symbiobacterium_thermophilum'\n",
    "    orth_dict['mpADK'] = 'mesotoga_prima'\n",
    "    orth_dict['miADK'] = 'mesotoga_infera'\n",
    "    orth_dict['ecADK_HTA'] = 'escherichia_coli_heat_adapted'\n",
    "    \n",
    "    \n",
    "    \n",
    "    try:\n",
    "        name = orth_dict[row['id']]\n",
    "    except:\n",
    "        name = '-'\n",
    "        \n",
    "    return name\n",
    "    \n",
    "    \n",
    "cleaned_MM['name'] = cleaned_MM.apply(org_name, axis=1)\n",
    "\n",
    "cleaned_MM.sort_values(by = 'kcat_median')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#add the optimal growth temperature to each row\n",
    "temperature_table = pd.read_table(\"temperature_data_full.tsv\", header=0)\n",
    "temperature_table = temperature_table[[\"organism\", \"temperature\"]]\n",
    "\n",
    "MM_final = pd.merge(cleaned_MM, temperature_table, left_on = 'name', right_on = 'organism', how = 'left').drop(columns = 'name').sort_values(by='kcat_mean')\n",
    "MM_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save the final table of michaelis menten parameters to a .csv\n",
    "MM_final.to_csv('8_Final_MM_Data.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "MM_final.shape[0]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "MM",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
